// @azure-tools/typespec-ts/dist/src/modular/buildRootIndex.js
import { NameType, normalizeName } from "@azure-tools/rlc-common";
import { getClassicalClientName } from "./helpers/namingHelpers.js";
import { resolveReference } from "../framework/reference.js";
import { CloudSettingHelpers, MultipartHelpers, PagingHelpers } from "./static-helpers-metadata.js";
import { getModularClientOptions } from "../utils/clientUtils.js";
import { getMethodHierarchiesMap } from "../utils/operationUtil.js";
import { join } from "path/posix";
import { useContext } from "../contextManager.js";
import { reportDiagnostic } from "../lib.js";
import { NoTarget } from "@typespec/compiler";
import { isLroOnlyOperation } from "./helpers/operationHelpers.js";
import { partitionAndEmitExports } from "./buildSubpathIndex.js";
export function buildRootIndex(context, emitterOptions, rootIndexFile, clientMap) {
    if (!clientMap) {
        // we still need to export the models if no client is provided
        exportModels(emitterOptions, rootIndexFile);
        return;
    }
    const project = useContext("outputProject");
    const [_, client] = clientMap;
    const srcPath = emitterOptions.modularOptions.sourceRoot;
    const { subfolder } = getModularClientOptions(clientMap);
    const clientName = `${getClassicalClientName(client)}`;
    const clientFile = project.getSourceFile(`${srcPath}/${subfolder && subfolder !== "" ? subfolder + "/" : ""}${normalizeName(clientName, NameType.File)}.ts`);
    if (!clientFile) {
        reportDiagnostic(context.program, {
            code: "client-file-not-found",
            format: {
                filePath: `${srcPath}/${normalizeName(clientName, NameType.File)}.ts`
            },
            target: NoTarget
        });
        return; // Skip exporting this client but continue with others
    }
    exportClassicalClient(client, rootIndexFile, subfolder !== null && subfolder !== void 0 ? subfolder : "");
    exportSimplePollerLike(context, clientMap, rootIndexFile, project, srcPath, subfolder);
    exportRestoreHelpers(rootIndexFile, project, srcPath, clientName, subfolder, true);
    exportModels(emitterOptions, rootIndexFile, clientName);
    exportModules(rootIndexFile, project, srcPath, clientName, "api", {
        subfolder,
        interfaceOnly: true,
        isTopLevel: true
    });
    exportModules(rootIndexFile, project, srcPath, clientName, "classic", {
        subfolder,
        isTopLevel: true
    });
    exportPagingTypes(context, rootIndexFile);
    exportFileContentsType(context, rootIndexFile);
    exportAzureCloudTypes(context, rootIndexFile);
}
function exportModels(emitterOptions, rootIndexFile, clientName = "") {
    var _a;
    // export models index file if not exists
    const project = useContext("outputProject");
    const srcPath = emitterOptions.modularOptions.sourceRoot;
    const modelsExportsIndex = (_a = rootIndexFile
        .getExportDeclarations()) === null || _a === void 0 ? void 0 : _a.find((i) => {
        var _a;
        return (_a = i.getModuleSpecifierValue()) === null || _a === void 0 ? void 0 : _a.startsWith(`./models/`);
    });
    if (!modelsExportsIndex) {
        exportModules(rootIndexFile, project, srcPath, clientName, "models", {
            isTopLevel: true,
            recursive: true
        });
    }
}
function exportAzureCloudTypes(context, rootIndexFile) {
    if (context.arm) {
        // AzureClouds is an enum (runtime value), AzureSupportedClouds is a type alias
        addExportsToRootIndexFile(rootIndexFile, [
            resolveReference(CloudSettingHelpers.AzureClouds)
        ]);
        addExportsToRootIndexFile(rootIndexFile, [resolveReference(CloudSettingHelpers.AzureSupportedClouds)], true);
    }
}
/**
 * This is a temporary solution for adding paging exports. Eventually we will have the binder generate the exports automatically.
 */
function exportPagingTypes(context, rootIndexFile) {
    if (!hasPaging(context)) {
        return;
    }
    addExportsToRootIndexFile(rootIndexFile, [
        resolveReference(PagingHelpers.PageSettings),
        resolveReference(PagingHelpers.ContinuablePage),
        resolveReference(PagingHelpers.PagedAsyncIterableIterator)
    ], true);
}
function hasPaging(context) {
    return context.sdkPackage.clients.some((client) => {
        const methodMap = getMethodHierarchiesMap(context, client);
        for (const [_, operations] of methodMap) {
            if (operations.some((op) => op.kind === "paging" || op.kind === "lropaging")) {
                return true;
            }
        }
        return false;
    });
}
function exportFileContentsType(context, rootIndexFile) {
    if (context.sdkPackage.models.some((x) => x.properties.some(
    // eslint-disable-next-line
    (y) => { var _a; return y.kind === "property" && ((_a = y.multipartOptions) === null || _a === void 0 ? void 0 : _a.isFilePart); }))) {
        addExportsToRootIndexFile(rootIndexFile, [resolveReference(MultipartHelpers.FileContents)], true);
    }
}
function getExistingExports(rootIndexFile) {
    return new Set(rootIndexFile
        .getExportDeclarations()
        .flatMap((exportDecl) => exportDecl.getNamedExports().map((namedExport) => namedExport.getName())));
}
function getNewNamedExports(namedExports, existingExports) {
    return namedExports.filter((namedExport) => !existingExports.has(namedExport));
}
function addExportsToRootIndexFile(rootIndexFile, namedExports, isTypeOnly = false) {
    const existingExports = getExistingExports(rootIndexFile);
    const newNamedExports = getNewNamedExports(namedExports, existingExports);
    if (newNamedExports.length > 0) {
        rootIndexFile.addExportDeclaration({
            isTypeOnly,
            namedExports: newNamedExports
        });
    }
}
function exportSimplePollerLike(context, clientMap, indexFile, project, srcPath, subfolder = "", isTopLevel = false) {
    var _a;
    const [_, client] = clientMap;
    const methodMap = getMethodHierarchiesMap(context, client);
    const hasLro = Array.from(methodMap.values()).some((operations) => {
        return operations.some(isLroOnlyOperation);
    });
    if (!hasLro || ((_a = context.rlcOptions) === null || _a === void 0 ? void 0 : _a.compatibilityLro) !== true) {
        return;
    }
    const helperFile = project.getSourceFile(`${srcPath}/${subfolder && subfolder !== "" ? subfolder + "/" : ""}static-helpers/simplePollerHelpers.ts`);
    if (!helperFile) {
        return;
    }
    const moduleSpecifier = `./${isTopLevel && subfolder && subfolder !== "" ? subfolder + "/" : ""}static-helpers/simplePollerHelpers.js`;
    indexFile.addExportDeclaration({
        isTypeOnly: true,
        moduleSpecifier,
        namedExports: ["SimplePollerLike"]
    });
}
function exportRestoreHelpers(indexFile, project, srcPath, clientName, subfolder = "", isTopLevel = false) {
    const helperFile = project.getSourceFile(`${srcPath}/${subfolder && subfolder !== "" ? subfolder + "/" : ""}restorePollerHelpers.ts`);
    if (!helperFile) {
        return;
    }
    const exported = new Set(indexFile.getExportedDeclarations().keys());
    const allEntries = [...helperFile.getExportedDeclarations().entries()];
    const moduleSpecifier = `./${isTopLevel && subfolder && subfolder !== "" ? subfolder + "/" : ""}restorePollerHelpers.js`;
    const renamer = (name) => exported.has(name) ? `${name} as ${clientName}${name}` : name;
    partitionAndEmitExports(indexFile, moduleSpecifier, allEntries, renamer);
}
function exportClassicalClient(client, indexFile, subfolder, isSubClient = false) {
    const clientName = client.name;
    indexFile.addExportDeclaration({
        namedExports: [clientName],
        moduleSpecifier: `./${subfolder && subfolder !== "" && !isSubClient ? subfolder + "/" : ""}${normalizeName(clientName, NameType.File)}.js`
    });
}
function exportModules(indexFile, project, srcPath, clientName, moduleName, options = {
    interfaceOnly: false,
    isTopLevel: false,
    subfolder: "",
    recursive: false
}) {
    var _a;
    const subfolder = (_a = options.subfolder) !== null && _a !== void 0 ? _a : "";
    let folders = [];
    if (options.recursive) {
        folders = project
            .getDirectories()
            .filter((dir) => {
            const formattedDir = dir.getPath().replace(/\\/g, "/");
            const targetPath = join(srcPath, subfolder, moduleName).replace(/\\/g, "/");
            return formattedDir.startsWith(targetPath);
        })
            .map((dir) => {
            return dir.getPath().replace(/\\/g, "/");
        });
    }
    else if (options.isTopLevel && moduleName === "api") {
        folders = project
            .getDirectories()
            .filter((dir) => {
            const formattedDir = dir.getPath().replace(/\\/g, "/");
            const targetPath = join(srcPath, subfolder, moduleName).replace(/\\/g, "/");
            return formattedDir.startsWith(targetPath);
        })
            .map((dir) => {
            return dir.getPath().replace(/\\/g, "/");
        });
    }
    else {
        folders = [join(srcPath, subfolder, moduleName).replace(/\\/g, "/")];
    }
    for (const folder of folders) {
        const apiFilePattern = join(folder, "index.ts").replace(/\\/g, "/");
        const modelsFile = project.getSourceFile(apiFilePattern);
        if (!modelsFile) {
            continue;
        }
        const exported = new Set(indexFile.getExportedDeclarations().keys());
        const serializerOrDeserializerRegex = /.*(Serializer|Deserializer)(_\d+)?$/;
        const filteredEntries = [
            ...modelsFile.getExportedDeclarations().entries()
        ].filter((exDeclaration) => {
            if (exDeclaration[0].startsWith("_")) {
                return false;
            }
            return exDeclaration[1].some((ex) => {
                if (options.interfaceOnly &&
                    ex.getKindName() !== "InterfaceDeclaration") {
                    return false;
                }
                if (moduleName === "models" &&
                    ex.getKindName() === "FunctionDeclaration" &&
                    serializerOrDeserializerRegex.test(exDeclaration[0])) {
                    return false;
                }
                if (options.interfaceOnly &&
                    options.isTopLevel &&
                    exDeclaration[0].endsWith("Context")) {
                    return false;
                }
                return true;
            });
        });
        const moduleSpecifier = `.${modelsFile
            .getFilePath()
            .replace(indexFile.getDirectoryPath(), "")
            .replace(/\\/g, "/")
            .replace(".ts", "")}.js`;
        const renamer = (name) => exported.has(name) ? `${name} as ${clientName}${name}` : name;
        partitionAndEmitExports(indexFile, moduleSpecifier, filteredEntries, renamer);
    }
}
export function buildSubClientIndexFile(context, clientMap, emitterOptions) {
    const project = useContext("outputProject");
    const [_, client] = clientMap;
    const { subfolder } = getModularClientOptions(clientMap);
    const srcPath = emitterOptions.modularOptions.sourceRoot;
    const subClientIndexFile = project.createSourceFile(`${srcPath}/${subfolder && subfolder !== "" ? subfolder + "/" : ""}index.ts`, undefined, { overwrite: true });
    const clientName = `${getClassicalClientName(client)}`;
    const clientFilePath = `${srcPath}/${subfolder && subfolder !== "" ? subfolder + "/" : ""}${normalizeName(clientName, NameType.File)}.ts`;
    const clientFile = project.getSourceFile(clientFilePath);
    if (!clientFile) {
        reportDiagnostic(context.program, {
            code: "client-file-not-found",
            format: {
                filePath: clientFilePath
            },
            target: NoTarget
        });
        return; // Skip exporting this client but continue with others
    }
    exportClassicalClient(client, subClientIndexFile, subfolder !== null && subfolder !== void 0 ? subfolder : "", true);
    exportSimplePollerLike(context, clientMap, subClientIndexFile, project, srcPath, subfolder);
    exportRestoreHelpers(subClientIndexFile, project, srcPath, clientName, subfolder);
    exportModules(subClientIndexFile, project, srcPath, clientName, "api", {
        subfolder,
        interfaceOnly: true,
        recursive: true
    });
    exportModules(subClientIndexFile, project, srcPath, clientName, "classic", {
        subfolder
    });
}
