// @azure-tools/rlc-common/dist/test/buildSnippets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildSnippets = buildSnippets;
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore: to fix the handlebars issue
const handlebars_1 = require("handlebars");
const template_js_1 = require("./template.js");
const nameConstructors_js_1 = require("../helpers/nameConstructors.js");
function buildSnippets(model, clientName, azureSdkForJs) {
    var _a, _b, _c, _d, _e, _f;
    const azureSdkForJsInfo = azureSdkForJs
        ? azureSdkForJs
        : (_a = model.options) === null || _a === void 0 ? void 0 : _a.azureSdkForJs;
    // to keep the same config for azure scope in buildReadmeFile.ts
    if ((((_c = (_b = model === null || model === void 0 ? void 0 : model.options) === null || _b === void 0 ? void 0 : _b.packageDetails) === null || _c === void 0 ? void 0 : _c.scopeName) === "azure" ||
        ((_e = (_d = model === null || model === void 0 ? void 0 : model.options) === null || _d === void 0 ? void 0 : _d.packageDetails) === null || _e === void 0 ? void 0 : _e.scopeName) === "azure-rest") &&
        model.options.addCredentials &&
        azureSdkForJsInfo) {
        return {
            path: "test/snippets.spec.ts",
            content: handlebars_1.default.compile(template_js_1.snippetsContent, { noEscape: true })({
                clientClassName: clientName ? clientName : (0, nameConstructors_js_1.getClientName)(model),
                azureArm: (_f = model.options) === null || _f === void 0 ? void 0 : _f.azureArm,
                azureSdkForJs: azureSdkForJsInfo,
                isModularLibrary: model.options.isModularLibrary,
                hasSubscriptionId: model.options.hasSubscriptionId
            })
        };
    }
}
