// @azure-tools/typespec-ts/dist/src/framework/hooks/binder.js
import { StructureKind } from "ts-morph";
import { provideContext, useContext } from "../../contextManager.js";
import { provideDependencies, useDependencies } from "./useDependencies.js";
import { refkey } from "../refkey.js";
import { SourceFileSymbol } from "../load-static-helpers.js";
import path from "path/posix";
import { normalizePath } from "@typespec/compiler";
import { generateLocallyUniqueName } from "../../modular/helpers/namingHelpers.js";
const PLACEHOLDER_PREFIX = "__PLACEHOLDER_";
class BinderImp {
    constructor(project, options = {}) {
        var _a;
        this.declarations = new Map();
        this.references = new Map();
        this.imports = new Map();
        this.symbolsBySourceFile = new Map();
        this.project = project;
        provideDependencies(options.dependencies);
        this.staticHelpers = (_a = options.staticHelpers) !== null && _a !== void 0 ? _a : new Map();
        this.dependencies = useDependencies();
    }
    trackDeclaration(refkey, name, sourceFile) {
        const uniqueName = this.generateLocallyUniqueDeclarationName(name, sourceFile);
        const declarationInfo = { name: uniqueName, sourceFile };
        this.declarations.set(refkey, declarationInfo);
        // Ensure symbolsBySourceFile has an entry for this sourceFile
        if (!this.symbolsBySourceFile.has(sourceFile)) {
            this.symbolsBySourceFile.set(sourceFile, new Set());
        }
        this.symbolsBySourceFile.get(sourceFile).add(uniqueName);
        return uniqueName;
    }
    /**
     * Generates a locally unique name for a declaration within a source file.
     * @param name - The base name of the declaration.
     * @param sourceFile - The source file where the declaration is made.
     * @returns A unique name for the declaration within the file.
     */
    generateLocallyUniqueDeclarationName(name, sourceFile) {
        var _a;
        const existingNamesInFile = (_a = this.symbolsBySourceFile.get(sourceFile)) !== null && _a !== void 0 ? _a : new Set();
        return generateLocallyUniqueName(name, existingNamesInFile);
    }
    /**
     * Generates a locally unique name for an import within a source file.
     * @param name - The base name of the import.
     * @param sourceFile - The source file where the import is made.
     * @returns A unique name for the import within the file.
     */
    generateLocallyUniqueImportName(name, sourceFile) {
        var _a, _b;
        const existingImports = ((_a = this.imports.get(sourceFile)) !== null && _a !== void 0 ? _a : [])
            .flatMap((i) => i.namedImports)
            .map((i) => { var _a; return (_a = i.alias) !== null && _a !== void 0 ? _a : i.name; });
        // Also check actual imports already added to the source file (e.g., manual imports)
        const actualImports = sourceFile
            .getImportDeclarations()
            .flatMap((imp) => imp.getNamedImports())
            .map((namedImp) => { var _a, _b; return (_b = (_a = namedImp.getAliasNode()) === null || _a === void 0 ? void 0 : _a.getText()) !== null && _b !== void 0 ? _b : namedImp.getName(); });
        const existingDeclarations = (_b = this.symbolsBySourceFile.get(sourceFile)) !== null && _b !== void 0 ? _b : new Set();
        return generateLocallyUniqueName(name, new Set([...existingImports, ...actualImports, ...existingDeclarations]));
    }
    /**
     * Resolves a reference to a declaration.
     *  If a declaration is not ready yet, a placeholder is added to the source file.
     *  Placeholders will be resolved when the imports are applied.
     * @param refkey - The reference key for the declaration.
     * @param currentSourceFile - The current source file where the reference is being resolved.
     * @returns The declaration information if resolved, or a placeholder if not found.
     */
    resolveReference(refkey) {
        return this.serializePlaceholder(refkey);
    }
    /**
     * Serializes a placeholder reference key to a string.
     * @param refkey - The reference key.
     * @returns The serialized placeholder string.
     */
    serializePlaceholder(refkey) {
        return `${PLACEHOLDER_PREFIX}${String(refkey)}__`;
    }
    /**
     * Adds an import declaration to a source file.
     * @param fileWhereImportIsAdded - The source file where the import is added.
     * @param fileWhereImportPointsTo - The source file being imported.
     * @param name - The name of the import.
     * @returns The import specifier structure.
     */
    addImport(fileWhereImportIsAdded, fileWhereImportPointsTo, name) {
        const importAlias = this.generateLocallyUniqueImportName(name, fileWhereImportIsAdded);
        let moduleSpecifier = "";
        if (typeof fileWhereImportPointsTo === "string") {
            moduleSpecifier = fileWhereImportPointsTo;
        }
        else {
            const relative = fileWhereImportIsAdded
                .getRelativePathTo(fileWhereImportPointsTo)
                .replace(".ts", ".js");
            moduleSpecifier =
                relative.startsWith(".") || relative.startsWith("/")
                    ? relative
                    : `./${relative}`;
        }
        const importStructures = this.imports.get(fileWhereImportIsAdded) || [];
        let importStructure = importStructures.find((imp) => imp.moduleSpecifier === moduleSpecifier);
        if (!importStructure) {
            importStructure = {
                kind: StructureKind.ImportDeclaration,
                moduleSpecifier,
                namedImports: []
            };
            importStructures.push(importStructure);
        }
        const namedImports = importStructure.namedImports;
        let importSpecifier = namedImports.find((n) => n.name === importAlias);
        if (!importSpecifier) {
            importSpecifier = {
                name: name,
                alias: name === importAlias ? undefined : importAlias,
                kind: StructureKind.ImportSpecifier
            };
            namedImports.push(importSpecifier);
        }
        importStructure.namedImports = namedImports;
        this.imports.set(fileWhereImportIsAdded, importStructures);
        return importSpecifier;
    }
    /**
     * Applies all tracked imports to their respective source files.
     */
    resolveAllReferences(sourceRoot, testRoot) {
        this.project.getSourceFiles().map((file) => {
            this.resolveDeclarationReferences(file);
            this.resolveDependencyReferences(file);
            const importStructures = this.imports.get(file);
            if (importStructures) {
                // Sort imports in place by module specifier to ensure consistent ordering
                importStructures.sort((a, b) => a.moduleSpecifier < b.moduleSpecifier ? -1 : 1);
                for (const importStructure of importStructures) {
                    file.addImportDeclaration(importStructure);
                }
            }
        });
        this.cleanUnreferencedHelpers(sourceRoot, testRoot);
    }
    resolveDependencyReferences(file) {
        var _a;
        if (!hasAnyPlaceholders(file)) {
            return;
        }
        for (const dependency of Object.values(this.dependencies)) {
            const placeholder = this.serializePlaceholder(refkey(dependency));
            const { name, module } = dependency;
            const occurences = countPlaceholderOccurrences(file, placeholder);
            if (occurences > 0) {
                const importDec = this.addImport(file, module, name);
                const uniqueName = (_a = importDec.alias) !== null && _a !== void 0 ? _a : name;
                replacePlaceholder(file, placeholder, uniqueName);
            }
        }
    }
    resolveDeclarationReferences(file) {
        var _a;
        if (!hasAnyPlaceholders(file)) {
            return;
        }
        for (const [declarationKey, declaration] of [
            ...this.declarations,
            ...this.staticHelpers
        ]) {
            const placeholderKey = this.serializePlaceholder(declarationKey);
            const occurences = countPlaceholderOccurrences(file, placeholderKey);
            if (!occurences) {
                continue;
            }
            let name = declaration.name;
            let declarationSourceFile;
            if ("sourceFile" in declaration) {
                declarationSourceFile = declaration.sourceFile;
            }
            else {
                declarationSourceFile = declaration[SourceFileSymbol];
            }
            if (file !== declarationSourceFile) {
                this.trackReference(declarationKey, file);
                const importDec = this.addImport(file, declarationSourceFile, name);
                name = (_a = importDec.alias) !== null && _a !== void 0 ? _a : name;
            }
            replacePlaceholder(file, placeholderKey, name);
        }
    }
    trackReference(refkey, sourceFile) {
        if (!this.references.has(refkey)) {
            this.references.set(refkey, new Set());
        }
        this.references.get(refkey).add(sourceFile);
    }
    cleanUnreferencedHelpers(sourceRoot, testRoot) {
        const usedHelperNames = [];
        for (const helper of this.staticHelpers.values()) {
            const sourceFile = helper[SourceFileSymbol];
            if (!sourceFile) {
                // This should be unreachable
                throw new Error(`Static helper ${helper.name} does not have a source file. Make sure that loadStaticHelpers has been correctly initialized in index.ts`);
            }
            const referencedHelper = this.references.get(refkey(helper));
            if (referencedHelper === null || referencedHelper === void 0 ? void 0 : referencedHelper.size) {
                usedHelperNames.push(sourceFile.getBaseNameWithoutExtension());
            }
        }
        function isFileUnused(file) {
            const name = file.getBaseNameWithoutExtension();
            // If one of the used helpers' name is a prefix of this file, the file likely represents a platform-specific implementation of the helper
            // so it should be marked as used even if the file has no direct references.
            return !usedHelperNames.some((s) => name.startsWith(s));
        }
        this.project
            //normalizae the final path to adapt to different systems
            .getSourceFiles(normalizePath(path.join(sourceRoot, "static-helpers/**/*.*ts")))
            .filter(isFileUnused)
            .forEach((helperFile) => helperFile.delete());
        if (!testRoot) {
            return;
        }
        this.project
            //normalizae the final path to adapt to different systems
            .getSourceFiles(normalizePath(path.join(testRoot, "test/generated/util/**/*.*ts")))
            .filter(isFileUnused)
            .forEach((helperFile) => helperFile.delete());
    }
}
// Provide the binder context to be used globally
export function provideBinder(project, options = {}) {
    const binder = new BinderImp(project, options);
    provideContext("binder", binder);
    return binder;
}
/**
 * Hook to use the binder context.
 * @returns The binder instance.
 */
export function useBinder() {
    return useContext("binder");
}
/**
 * Replaces all instances of a placeholder in a source file with a given value.
 * @param sourceFile - The source file where the replacement occurs.
 * @param placeholder - The placeholder string to replace.
 * @param value - The value to replace the placeholder with.
 */
function replacePlaceholder(sourceFile, placeholder, value) {
    const fileText = sourceFile.getFullText();
    const regex = new RegExp(escapeRegExp(placeholder), "g");
    const updatedText = fileText.replace(regex, value);
    sourceFile.replaceWithText(updatedText);
}
function countPlaceholderOccurrences(sourceFile, placeholder) {
    return sourceFile.getFullText().split(placeholder).length - 1;
}
/**
 * Escapes special characters in a string to be used in a regular expression.
 * @param string - The input string.
 * @returns The escaped string.
 */
function escapeRegExp(string) {
    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function hasAnyPlaceholders(sourceFile) {
    return sourceFile.getFullText().includes(PLACEHOLDER_PREFIX);
}
