// @azure-tools/typespec-ts/dist/src/framework/hooks/sdkTypes.js
import { getNamespaceFullName } from "@typespec/compiler";
import { getClientType } from "@azure-tools/typespec-client-generator-core";
import { provideContext, useContext } from "../../contextManager.js";
import { visitPackageTypes } from "../../modular/emitModels.js";
import { getAllAncestors, getAllProperties } from "../../modular/helpers/operationHelpers.js";
import { normalizeModelPropertyName } from "../../modular/type-expressions/get-type-expression.js";
import { reportDiagnostic } from "../../lib.js";
import { NameType, normalizeName } from "@azure-tools/rlc-common";
export const emitQueue = new Set();
export const flattenPropertyModelMap = new Map();
export function useSdkTypes() {
    const sdkTypesContext = useContext("sdkTypes");
    const { tcgcContext } = useContext("emitContext");
    function getSdkType(type) {
        var _a;
        let sdkType;
        if (type.kind === "Operation") {
            sdkType = sdkTypesContext.operations.get(type);
        }
        else {
            sdkType =
                (_a = sdkTypesContext.types.get(type)) !== null && _a !== void 0 ? _a : getClientType(tcgcContext, type);
        }
        if (!sdkType) {
            throw new Error(`SdkType not found for type: ${type.kind} ${"name" in type && typeof type.name == "string" ? type.name : ""} ${"namespace" in type && type.namespace
                ? ` in ${getNamespaceFullName(type.namespace)}`
                : ""}`);
        }
        return sdkType;
    }
    return getSdkType;
}
export function provideSdkTypes(context) {
    const { sdkPackage } = context;
    const sdkTypesContext = {
        operations: new Map(),
        types: new Map(),
        flattenProperties: new Map()
    };
    visitPackageTypes(context);
    enrichFlattenProperties(context, sdkTypesContext.flattenProperties, flattenPropertyModelMap);
    for (const sdkModel of emitQueue) {
        switch (sdkModel.kind) {
            case "model":
                sdkModel.properties.forEach((prop) => {
                    sdkTypesContext.types.set(prop.type.__raw, prop.type);
                });
                if (sdkModel.discriminatedSubtypes) {
                    Object.values(sdkModel.discriminatedSubtypes).forEach((subtype) => {
                        sdkTypesContext.types.set(subtype.__raw, subtype);
                    });
                }
                break;
            case "enum":
                if (sdkModel.__raw) {
                    sdkTypesContext.types.set(sdkModel.__raw, sdkModel);
                }
                break;
            case "union":
                sdkModel.variantTypes.forEach((v) => {
                    if (v.__raw) {
                        sdkTypesContext.types.set(v.__raw, v);
                    }
                });
                break;
            case "array":
                sdkTypesContext.types.set(sdkModel.valueType.__raw, sdkModel.valueType);
                break;
            case "dict":
                sdkTypesContext.types.set(sdkModel.valueType.__raw, sdkModel.valueType);
                break;
            case "nullable":
                sdkTypesContext.types.set(sdkModel.__raw, sdkModel);
                sdkTypesContext.types.set(sdkModel.type.__raw, sdkModel.type);
                break;
        }
        if (!sdkModel.__raw) {
            continue;
        }
        sdkTypesContext.types.set(sdkModel.__raw, sdkModel);
    }
    for (const client of sdkPackage.clients) {
        for (const method of getAllOperationsFromClient(client)) {
            if (!method.__raw) {
                continue;
            }
            sdkTypesContext.operations.set(method.__raw, method);
            // Visit the parameters of the method to add them to the types map
            method.parameters.forEach((param) => {
                sdkTypesContext.types.set(param.type.__raw, param.type);
            });
        }
    }
    provideContext("sdkTypes", sdkTypesContext);
}
// Enrich flatten properties with their base models and conflict maps
function enrichFlattenProperties(context, propertyContextMap, propertyModelMap) {
    // Build a map of base model to its existing properties excluding flatten properties
    // To check for conflicts later
    const baseModelProperties = new Map();
    propertyModelMap.forEach((baseModel, _) => {
        if (!baseModelProperties.has(baseModel)) {
            const propertiesExcludedFlatten = getAllProperties(context, baseModel, getAllAncestors(baseModel))
                .filter((p) => !(p.flatten && p.type.kind === "model"))
                .map((p) => normalizeModelPropertyName(context, p));
            baseModelProperties.set(baseModel, new Set(propertiesExcludedFlatten));
        }
    });
    for (const [flattenProperty, baseModel] of propertyModelMap) {
        const flattenContext = {
            baseModel: baseModel
        };
        propertyContextMap.set(flattenProperty, flattenContext);
        const existingProperties = baseModelProperties.get(baseModel);
        const conflictMap = new Map();
        const flattenModel = flattenProperty.type;
        if (baseModelProperties.has(flattenModel)) {
            // If the flatten model is also a base model of other flatten properties, which means it has multiple consecutive flatten operations
            // Since we cannot handle the flatten transition, report warning and skip it for now
            reportDiagnostic(context.program, {
                code: "unsupported-flatten-transition",
                format: {
                    propertyName: flattenProperty.name,
                    modelName: baseModel.name
                },
                target: flattenProperty.__raw
            });
        }
        const allFlattenProperties = getAllProperties(context, flattenModel, getAllAncestors(flattenModel));
        for (const flattenChildProperty of allFlattenProperties) {
            let childPropertyName = normalizeModelPropertyName(context, flattenChildProperty);
            if (existingProperties.has(childPropertyName)) {
                childPropertyName = normalizeName(`${childPropertyName}_${flattenProperty.name}_${childPropertyName}`, NameType.Property);
                conflictMap.set(flattenChildProperty, childPropertyName);
            }
            existingProperties.add(childPropertyName);
        }
        if (conflictMap.size === 0) {
            continue;
        }
        flattenContext.conflictMap = conflictMap;
    }
}
export function getAllOperationsFromClient(client) {
    const clientQueue = [client];
    const operations = [];
    while (clientQueue.length > 0) {
        const client = clientQueue.pop();
        if (client.children) {
            clientQueue.push(...client.children);
        }
        operations.push(...client.methods);
    }
    return operations;
}
