// @azure-tools/typespec-ts/dist/src/utils/clientUtils.js
import { InitializedByFlags, listAllServiceNamespaces, listClients } from "@azure-tools/typespec-client-generator-core";
import { getNamespaceFullName, isTemplateDeclaration, isTemplateDeclarationOrInstance } from "@typespec/compiler";
import { NameType, normalizeName } from "@azure-tools/rlc-common";
export function getRLCClients(dpgContext, isModularLibrary) {
    var _a, _b, _c;
    const modular = (_b = isModularLibrary !== null && isModularLibrary !== void 0 ? isModularLibrary : (_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.isModularLibrary) !== null && _b !== void 0 ? _b : false;
    const clients = listClients(dpgContext);
    const rawServiceNamespaces = (_c = dpgContext.allServiceNamespaces) !== null && _c !== void 0 ? _c : listAllServiceNamespaces(dpgContext);
    // For one client in Modular: Return the client from listClients with multi-service support
    if (modular && clients.length === 1) {
        return clients.map((client) => {
            const services = client.services;
            return {
                ...client,
                services: services,
                crossLanguageDefinitionId: `${getNamespaceFullName(services[0])}.${client.name}`
            };
        });
    }
    else {
        // For RLC and multiple clients in Modular:
        // Flatten all services and return one client per service
        const services = new Set();
        clients.forEach((c) => {
            const clientService = c.services;
            clientService.forEach((ns) => services.add(ns));
        });
        if (services.size > 0) {
            return [...services.values()].map((service) => {
                const clientName = service.name + "Client";
                return {
                    kind: "SdkClient",
                    name: clientName,
                    service: service,
                    type: service,
                    services: [service],
                    subClients: [],
                    clientPath: clientName,
                    arm: Boolean(dpgContext.arm),
                    crossLanguageDefinitionId: `${getNamespaceFullName(service)}.${clientName}`,
                    subOperationGroups: []
                };
            });
        }
    }
    // Fallback to raw service namespaces if no clients found
    return rawServiceNamespaces.map((service) => {
        const clientName = service.name + "Client";
        return {
            kind: "SdkClient",
            name: clientName,
            service: service,
            type: service,
            services: [service],
            subClients: [],
            clientPath: clientName,
            arm: Boolean(dpgContext.arm),
            crossLanguageDefinitionId: `${getNamespaceFullName(service)}.${clientName}`,
            subOperationGroups: []
        };
    });
}
export function listOperationsUnderRLCClient(client) {
    const operations = [];
    const serviceArray = client.services;
    const queue = [...serviceArray];
    while (queue.length > 0) {
        const current = queue.shift();
        if (current.decorators.some((d) => {
            var _a, _b;
            return ((_a = d.definition) === null || _a === void 0 ? void 0 : _a.name) === "@client" &&
                getNamespaceFullName((_b = d.definition) === null || _b === void 0 ? void 0 : _b.namespace) ===
                    "Azure.ClientGenerator.Core";
        }) &&
            !serviceArray.includes(current)) {
            continue;
        }
        operations.push(...[...current.operations.values()].filter((op) => isTemplateDeclarationOrInstance(op) === false));
        if (current.kind === "Namespace") {
            queue.push(...current.namespaces.values());
            queue.push(...[...current.interfaces.values()].filter((i) => isTemplateDeclaration(i) === false));
        }
    }
    return operations;
}
export function isRLCMultiEndpoint(dpgContext) {
    return getRLCClients(dpgContext).length > 1;
}
export function getModularClientOptions(clientMap) {
    const [hierarchy, client] = clientMap;
    const clientOptions = {
        rlcClientName: `${client.name.replace(/Client$/, "")}Context`
    };
    clientOptions.subfolder = hierarchy.join("/");
    return clientOptions;
}
export function getClientHierarchyMap(context) {
    var _a;
    const clientMap = [];
    const individualClients = context.sdkPackage.clients.filter((client) => {
        return (client.clientInitialization.initializedBy &
            InitializedByFlags.Individually);
    });
    const clients = individualClients.map((client) => {
        return [
            context.sdkPackage.clients.length > 1
                ? [normalizeName(client.name.replace("Client", ""), NameType.File)]
                : [],
            client
        ];
    });
    for (let i = 0; i < clients.length; i++) {
        const [hierarchy, client] = clients[i];
        clientMap.push([hierarchy, client]);
        const childIndividualClients = (_a = client.children) === null || _a === void 0 ? void 0 : _a.filter((client) => {
            return (client.clientInitialization.initializedBy &
                InitializedByFlags.Individually);
        });
        if (childIndividualClients && childIndividualClients.length > 0) {
            childIndividualClients.forEach((child) => {
                const childHierarchy = [
                    ...hierarchy,
                    normalizeName(child.name.replace("Client", ""), NameType.File)
                ];
                clients.push([childHierarchy, child]);
            });
        }
    }
    return clientMap;
}
