// @azure-tools/typespec-ts/dist/src/utils/credentialUtils.js
import { NoTarget } from "@typespec/compiler";
import { reportDiagnostic } from "../lib.js";
/**
 * Get supported HTTP authentication schemes and filter out unsupported ones.
 * @param program used to report diagnostics
 * @param authorization authentication object
 * @returns supported HTTP authentication schemes
 */
export function getSupportedHttpAuth(program, authorization) {
    const authList = [];
    for (const option of authorization.options) {
        for (const auth of option.schemes) {
            switch (auth.type) {
                case "apiKey":
                    if (isSupportedKeyCredential(auth)) {
                        authList.push(auth);
                    }
                    else {
                        // just skip un-supported authentication and report a warning
                        reportDiagnostic(program, {
                            code: "un-supported-credential",
                            format: {
                                credentialType: `${auth.type} ${auth.in}`
                            },
                            target: NoTarget
                        });
                    }
                    break;
                default:
                    authList.push(auth);
                    break;
            }
        }
    }
    return authList;
}
export function isSupportedKeyCredential(auth) {
    return ((auth.type === "apiKey" && auth.in === "header") || auth.type === "http");
}
export function isSupportedTokenCredential(auth) {
    return auth.type === "oauth2";
}
export function hasKeyCredential(initialization) {
    const authScheme = getAuthScheme(initialization);
    return authScheme.some((auth) => isSupportedKeyCredential(auth));
}
export function hasTokenCredential(initialization) {
    const authScheme = getAuthScheme(initialization);
    return authScheme.some((auth) => isSupportedTokenCredential(auth));
}
function getAuthScheme(initialization) {
    var _a;
    const credentialParams = (_a = initialization.parameters) === null || _a === void 0 ? void 0 : _a.find((param) => param.kind === "credential");
    if (!credentialParams) {
        return [];
    }
    const kind = credentialParams.type.kind;
    const authScheme = [];
    if (kind === "credential") {
        authScheme.push(credentialParams.type.scheme);
    }
    else if (kind === "union") {
        for (const param of credentialParams.type.variantTypes) {
            authScheme.push(param.scheme);
        }
    }
    return authScheme;
}
