// @azure-tools/typespec-ts/dist/src/utils/namespaceUtils.js
import { getAllModels, UsageFlags } from "@azure-tools/typespec-client-generator-core";
import { isGlobalNamespace, isService, NoTarget } from "@typespec/compiler";
import { reportDiagnostic } from "../lib.js";
export function getModelNamespaceName(dpgContext, namespace) {
    const result = [];
    if (namespace &&
        !isGlobalNamespace(dpgContext.program, namespace) &&
        !isService(dpgContext.program, namespace)) {
        result.push(...getModelNamespaceName(dpgContext, namespace.namespace));
        result.push(namespace.name);
    }
    return result;
}
export function getOperationNamespaceInterfaceName(dpgContext, operation) {
    var _a, _b, _c, _d;
    const result = [];
    if (((_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.hierarchyClient) === false &&
        ((_b = dpgContext.rlcOptions) === null || _b === void 0 ? void 0 : _b.enableOperationGroup) !== true) {
        return result;
    }
    if (operation.interface) {
        if (((_c = dpgContext.rlcOptions) === null || _c === void 0 ? void 0 : _c.enableOperationGroup) === true &&
            ((_d = dpgContext.rlcOptions) === null || _d === void 0 ? void 0 : _d.hierarchyClient) === false) {
            result.push(operation.interface.name);
            return result;
        }
        if (operation.interface.namespace &&
            !isGlobalNamespace(dpgContext.program, operation.interface.namespace) &&
            !isService(dpgContext.program, operation.interface.namespace)) {
            result.push(...getModelNamespaceName(dpgContext, operation.interface.namespace));
        }
        result.push(operation.interface.name);
    }
    else if (operation.namespace &&
        !isGlobalNamespace(dpgContext.program, operation.namespace) &&
        !isService(dpgContext.program, operation.namespace)) {
        result.push(...getModelNamespaceName(dpgContext, operation.namespace.namespace));
        result.push(operation.namespace.name);
    }
    return result;
}
export function detectModelConflicts(dpgContext) {
    const allModels = getAllModels(dpgContext);
    const nameSet = new Set();
    const reported = new Set();
    let hasConflict = false;
    for (const model of allModels) {
        if (model.name === "" || model.usage === UsageFlags.ApiVersionEnum) {
            continue;
        }
        if (nameSet.has(model.name) && !reported.has(model.name)) {
            reportDiagnostic(dpgContext.program, {
                code: "detected-model-name-conflict",
                format: {
                    modelName: model.name,
                    namespaces: allModels
                        .filter((m) => m.name === model.name)
                        .map((m) => m.namespace)
                        .join(" and ")
                },
                target: NoTarget
            });
            reported.add(model.name);
            hasConflict = true;
        }
        nameSet.add(model.name);
    }
    return hasConflict;
}
