// @azure-tools/typespec-ts/dist/src/utils/operationUtil.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { getLroLogicalResponseName, getResponseTypeName, NameType, normalizeName, OPERATION_LRO_HIGH_PRIORITY, OPERATION_LRO_LOW_PRIORITY } from "@azure-tools/rlc-common";
import { getLroMetadata } from "@azure-tools/typespec-azure-core";
import { getDisablePageable, getHttpOperationWithCache, getWireName, InitializedByFlags } from "@azure-tools/typespec-client-generator-core";
import { isList, NoTarget } from "@typespec/compiler";
import { KnownMediaType, knownMediaType, isMediaTypeXml, isMediaTypeMultipart } from "./mediaTypes.js";
import { isByteOrByteUnion } from "./modelUtils.js";
import { getOperationNamespaceInterfaceName } from "./namespaceUtils.js";
import { resolveReference } from "../framework/reference.js";
import { SerializationHelpers } from "../modular/static-helpers-metadata.js";
import { listOperationsUnderRLCClient } from "./clientUtils.js";
import { $ } from "@typespec/compiler/typekit";
import { reportDiagnostic } from "../lib.js";
// Sorts the responses by status code
export function sortedOperationResponses(responses) {
    return responses.sort((a, b) => {
        if (a.statusCodes === "*") {
            return 1;
        }
        if (b.statusCodes === "*") {
            return -1;
        }
        const aStatus = typeof a.statusCodes === "number" ? a.statusCodes : a.statusCodes.start;
        const bStatus = typeof b.statusCodes === "number" ? b.statusCodes : b.statusCodes.start;
        return aStatus - bStatus;
    });
}
/**
 * This function computes all the response types error and success
 * an operation can end up returning.
 */
export function getOperationResponseTypes(dpgContext, operation) {
    const returnTypes = {
        error: [],
        success: []
    };
    function getResponseType(responses) {
        return responses.map((r) => {
            const statusCode = getOperationStatuscode(r);
            const responseName = getResponseTypeName(getOperationGroupName(dpgContext, operation), getOperationName(dpgContext, operation.operation), statusCode);
            return responseName;
        });
    }
    if (operation.responses && operation.responses.length) {
        returnTypes.error = getResponseType(operation.responses.filter((r) => isDefaultStatusCode(r.statusCodes)));
        returnTypes.success = getResponseType(operation.responses.filter((r) => isDefinedStatusCode(r.statusCodes)));
    }
    return returnTypes;
}
/**
 * Extracts all success or defined status codes for a give operation
 */
export function getOperationSuccessStatus(operation) {
    var _a;
    const responses = (_a = operation.responses) !== null && _a !== void 0 ? _a : [];
    const status = [];
    for (const response of responses) {
        if (isDefinedStatusCode(response.statusCodes)) {
            status.push(getOperationStatuscode(response));
        }
    }
    return status;
}
export function getOperationStatuscode(response) {
    const statusCodes = response.statusCodes;
    if (statusCodes === "*") {
        return "default";
    }
    else if (typeof statusCodes === "number") {
        return String(statusCodes);
    }
    else {
        // FIXME - this is a hack to get the first status code
        // https://github.com/Azure/autorest.typescript/issues/2063
        return String(statusCodes.start);
    }
}
export function getOperationGroupName(dpgContext, operationOrRoute) {
    var _a;
    if (!((_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.enableOperationGroup) || !operationOrRoute) {
        return "";
    }
    // If this is a HttpOperation
    if (operationOrRoute.kind !== "Operation") {
        operationOrRoute = operationOrRoute.operation;
    }
    const operation = operationOrRoute;
    const namespaceNames = getOperationNamespaceInterfaceName(dpgContext, operation);
    return namespaceNames
        .map((name) => {
        return normalizeName(name, NameType.Interface, true);
    })
        .join("");
}
export function getOperationName(dpgContext, operation) {
    const projectedOperationName = getWireName(dpgContext, operation);
    return normalizeName(projectedOperationName !== null && projectedOperationName !== void 0 ? projectedOperationName : operation.name, NameType.Interface, true);
}
export function isDefaultStatusCode(statusCodes) {
    return statusCodes === "*";
}
export function isDefinedStatusCode(statusCodes) {
    return statusCodes !== "*";
}
export function isBinaryPayload(dpgContext, body, contentType) {
    const knownMediaTypes = (Array.isArray(contentType) ? contentType : [contentType]).map((ct) => knownMediaType(ct));
    for (const type of knownMediaTypes) {
        if (type === KnownMediaType.Binary && isByteOrByteUnion(dpgContext, body)) {
            return true;
        }
    }
    return false;
}
/**
 * Checks if the content type(s) indicate XML payload
 */
export function isXmlPayload(contentType) {
    const contentTypes = Array.isArray(contentType) ? contentType : [contentType];
    return contentTypes.some((ct) => isMediaTypeXml(ct));
}
/**
 * Checks if the content type(s) indicate multipart payload (multipart/mixed, multipart/form-data, etc.)
 */
export function isMultipartPayload(contentType) {
    return isMediaTypeMultipart(contentType);
}
/**
 * Checks if the operation supports multiple content types (e.g., both JSON and XML)
 */
export function hasDualFormatSupport(contentTypes) {
    const hasJson = contentTypes.some((ct) => knownMediaType(ct) === KnownMediaType.Json);
    const hasXml = contentTypes.some((ct) => knownMediaType(ct) === KnownMediaType.Xml);
    return hasJson && hasXml;
}
export function isLongRunningOperation(program, operation) {
    return Boolean(getLroMetadata(program, operation.operation));
}
/**
 * Return if we have a client-level LRO overloading
 * @param pathDictionary
 * @returns
 */
export function getClientLroOverload(pathDictionary) {
    var _a;
    let lroCounts = 0, allowCounts = 0;
    for (const details of Object.values(pathDictionary)) {
        for (const methodDetails of Object.values(details.methods)) {
            for (const detail of methodDetails) {
                const lroDetail = (_a = detail.operationHelperDetail) === null || _a === void 0 ? void 0 : _a.lroDetails;
                if (lroDetail === null || lroDetail === void 0 ? void 0 : lroDetail.isLongRunning) {
                    lroCounts++;
                    if (!lroDetail.operationLroOverload) {
                        return false;
                    }
                    allowCounts++;
                }
            }
        }
    }
    return Boolean(lroCounts > 0 && lroCounts === allowCounts);
}
/**
 * Check if we have an operation-level overloading
 * @param program
 * @param operation The operation detail
 * @param existingResponseTypes auxilary param for current response types
 * @param existingResponses auxilary param for raw response data
 * @returns
 */
export function getOperationLroOverload(program, operation, existingResponseTypes, existingResponses) {
    const metadata = getLroMetadata(program, operation.operation);
    if (!metadata) {
        return false;
    }
    const hasSuccessReturn = existingResponses === null || existingResponses === void 0 ? void 0 : existingResponses.filter((r) => r.statusCode.startsWith("20"));
    if ((existingResponseTypes === null || existingResponseTypes === void 0 ? void 0 : existingResponseTypes.success) || hasSuccessReturn) {
        return true;
    }
    return false;
}
/**
 * Extract the operation LRO details
 * @param program
 * @param operation Operation detail
 * @param responsesTypes Calculated response types
 * @param operationGroupName Operation group name
 * @returns
 */
export function extractOperationLroDetail(dpgContext, operation, responsesTypes, operationGroupName) {
    var _a, _b;
    let logicalResponseTypes;
    let precedence = OPERATION_LRO_LOW_PRIORITY;
    const operationLroOverload = getOperationLroOverload(dpgContext.program, operation, responsesTypes);
    if (operationLroOverload) {
        logicalResponseTypes = {
            error: responsesTypes.error,
            success: [
                getLroLogicalResponseName(operationGroupName, getOperationName(dpgContext, operation.operation))
            ]
        };
        const metadata = getLroMetadata(dpgContext.program, operation.operation);
        precedence =
            (metadata === null || metadata === void 0 ? void 0 : metadata.finalStep) &&
                metadata.finalStep.kind === "pollingSuccessProperty" &&
                (metadata === null || metadata === void 0 ? void 0 : metadata.finalStep.target) &&
                ((_b = (_a = metadata === null || metadata === void 0 ? void 0 : metadata.finalStep) === null || _a === void 0 ? void 0 : _a.target) === null || _b === void 0 ? void 0 : _b.name) === "result"
                ? OPERATION_LRO_HIGH_PRIORITY
                : OPERATION_LRO_LOW_PRIORITY;
    }
    return {
        isLongRunning: Boolean(getLroMetadata(dpgContext.program, operation.operation)),
        logicalResponseTypes,
        operationLroOverload,
        precedence
    };
}
export function hasPollingOperations(client, dpgContext) {
    var _a;
    const program = dpgContext.program;
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        if (isLongRunningOperation(program, route)) {
            return true;
        }
    }
    return false;
}
export function extractPageDetails(program, operation) {
    var _a, _b, _c, _d;
    if (isList(program, operation.operation)) {
        // If the operation is a list, we don't need to extract paging details.
        const metadata = $(program).operation.getPagingMetadata(operation.operation);
        if (metadata === undefined) {
            // would fallback to default paging metadata
            reportDiagnostic(program, {
                code: "no-paging-items-defined",
                format: {
                    operationName: operation.operation.name
                },
                target: NoTarget
            });
        }
        const nextLinkPath = mapFirstSegmentForResultSegments((_a = metadata === null || metadata === void 0 ? void 0 : metadata.output.nextLink) === null || _a === void 0 ? void 0 : _a.path, operation.responses);
        const itemNamePath = mapFirstSegmentForResultSegments((_b = metadata === null || metadata === void 0 ? void 0 : metadata.output.pageItems) === null || _b === void 0 ? void 0 : _b.path, operation.responses);
        if ((nextLinkPath && (nextLinkPath === null || nextLinkPath === void 0 ? void 0 : nextLinkPath.length) > 1) ||
            (itemNamePath && (itemNamePath === null || itemNamePath === void 0 ? void 0 : itemNamePath.length) > 1)) {
            // Any cases with nested nextLink or itemName are not supported
            reportDiagnostic(program, {
                code: "un-supported-paging-cases",
                format: {
                    operationName: operation.operation.name
                },
                target: NoTarget
            });
            // these paging information will be ignored
            // and the operation will be treated as a non-paging operation.
            return undefined;
        }
        const nextLinkNames = (_c = nextLinkPath === null || nextLinkPath === void 0 ? void 0 : nextLinkPath.map((prop) => prop.name).join(".")) !== null && _c !== void 0 ? _c : "nextLink";
        const itemNames = (_d = itemNamePath === null || itemNamePath === void 0 ? void 0 : itemNamePath.map((prop) => prop.name).join(".")) !== null && _d !== void 0 ? _d : "value";
        return {
            nextLinkNames: [nextLinkNames],
            itemNames: [itemNames]
        };
    }
    return undefined;
}
export function isPagingOperation(dpgContext, operation) {
    const { program } = dpgContext;
    if (getDisablePageable(dpgContext, operation.operation)) {
        return false;
    }
    return extractPageDetails(program, operation) !== undefined;
}
function mapFirstSegmentForResultSegments(resultSegments, responses) {
    var _a, _b, _c;
    const pagingBodyType = (_b = (_a = responses.find((r) => r.statusCodes === 200)) === null || _a === void 0 ? void 0 : _a.responses[0]) === null || _b === void 0 ? void 0 : _b.body;
    if (!pagingBodyType || pagingBodyType.bodyKind !== "single")
        return undefined;
    const bodyType = pagingBodyType.type;
    if (resultSegments === undefined || bodyType === undefined)
        return undefined;
    // TCGC use Http response type as the return type
    // For implicit body response, we need to locate the first segment in the response type
    // Several cases:
    // 1. `op test(): {items, nextLink}`
    // 2. `op test(): {items, nextLink} & {a, b, c}`
    // 3. `op test(): {@bodyRoot body: {items, nextLink}}`
    if (resultSegments.length > 0 && bodyType && bodyType.kind === "Model") {
        for (let i = 0; i < resultSegments.length; i++) {
            const segment = resultSegments[i];
            for (const property of (_c = bodyType.properties) !== null && _c !== void 0 ? _c : []) {
                if (property &&
                    segment &&
                    findRootSourceProperty(property[1]) ===
                        findRootSourceProperty(segment)) {
                    return [property[1], ...resultSegments.slice(i + 1)];
                }
            }
        }
    }
    return resultSegments;
}
function findRootSourceProperty(property) {
    while (property.sourceProperty) {
        property = property.sourceProperty;
    }
    return property;
}
export function hasPagingOperations(client, dpgContext) {
    var _a;
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        if (isPagingOperation(dpgContext, route)) {
            return true;
        }
    }
    return false;
}
export function hasCollectionFormatInfo(paramType, paramFormat) {
    return (getHasMultiCollection(paramType, paramFormat, false) ||
        getHasSsvCollection(paramType, paramFormat) ||
        getHasTsvCollection(paramType, paramFormat) ||
        getHasCsvCollection(paramType, paramFormat) ||
        getHasPipeCollection(paramType, paramFormat) ||
        getHasNewlineCollection(paramType, paramFormat));
}
export function getSpecialSerializeInfo(dpgContext, paramType, paramFormat) {
    var _a, _b;
    const hasMultiCollection = getHasMultiCollection(paramType, paramFormat, 
    // Include query multi support in compatibility mode
    (_b = (_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.compatibilityQueryMultiFormat) !== null && _b !== void 0 ? _b : false);
    const hasCsvCollection = getHasCsvCollection(paramType, paramFormat);
    const descriptions = [];
    const collectionInfo = [];
    if (hasMultiCollection) {
        descriptions.push("buildMultiCollection");
        collectionInfo.push("multi");
    }
    if (hasCsvCollection) {
        descriptions.push("buildCsvCollection");
        collectionInfo.push("csv");
    }
    return {
        hasMultiCollection,
        hasCsvCollection,
        descriptions,
        collectionInfo
    };
}
function getHasMultiCollection(paramType, paramFormat, includeQuery = true) {
    return (((includeQuery && paramType === "query") || paramType === "header") &&
        paramFormat === KnownCollectionFormat.Multi);
}
function getHasSsvCollection(paramType, paramFormat) {
    return ((paramType === "query" || paramType === "property") &&
        paramFormat === KnownCollectionFormat.Ssv);
}
function getHasTsvCollection(paramType, paramFormat) {
    return paramType === "query" && paramFormat === KnownCollectionFormat.Tsv;
}
function getHasCsvCollection(paramType, paramFormat) {
    return ((paramType === "header" || paramType === "property") &&
        paramFormat === KnownCollectionFormat.Csv);
}
function getHasPipeCollection(paramType, paramFormat) {
    return ((paramType === "query" || paramType === "property") &&
        paramFormat === KnownCollectionFormat.Pipes);
}
function getHasNewlineCollection(paramType, paramFormat) {
    return (paramType === "property" && paramFormat === KnownCollectionFormat.Newline);
}
export function getCollectionFormatHelper(format) {
    switch (format) {
        case KnownCollectionFormat.Multi:
            return resolveReference(SerializationHelpers.buildMultiCollection);
        case KnownCollectionFormat.Pipes:
            return resolveReference(SerializationHelpers.buildPipeCollection);
        case KnownCollectionFormat.Ssv:
            return resolveReference(SerializationHelpers.buildSsvCollection);
        case KnownCollectionFormat.Tsv:
            return resolveReference(SerializationHelpers.buildTsvCollection);
        case KnownCollectionFormat.Csv:
            return resolveReference(SerializationHelpers.buildCsvCollection);
        case KnownCollectionFormat.Newline:
            return resolveReference(SerializationHelpers.buildNewlineCollection);
        default:
            return undefined;
    }
}
export function getCollectionFormatParseHelper(format) {
    switch (format) {
        case KnownCollectionFormat.Pipes:
            return resolveReference(SerializationHelpers.parsePipeCollection);
        case KnownCollectionFormat.Ssv:
            return resolveReference(SerializationHelpers.parseSsvCollection);
        case KnownCollectionFormat.Csv:
            return resolveReference(SerializationHelpers.parseCsvCollection);
        case KnownCollectionFormat.Newline:
            return resolveReference(SerializationHelpers.parseNewlineCollection);
        default:
            return undefined;
    }
}
export function getCollectionFormatFromArrayEncoding(encoding) {
    switch (encoding) {
        case "pipeDelimited":
            return KnownCollectionFormat.Pipes;
        case "spaceDelimited":
            return KnownCollectionFormat.Ssv;
        case "commaDelimited":
            return KnownCollectionFormat.Csv;
        case "newlineDelimited":
            return KnownCollectionFormat.Newline;
        default:
            return undefined;
    }
}
export var KnownCollectionFormat;
(function (KnownCollectionFormat) {
    KnownCollectionFormat["Csv"] = "csv";
    KnownCollectionFormat["Ssv"] = "ssv";
    KnownCollectionFormat["Tsv"] = "tsv";
    KnownCollectionFormat["Pipes"] = "pipes";
    KnownCollectionFormat["Newline"] = "newline";
    KnownCollectionFormat["Multi"] = "multi";
})(KnownCollectionFormat || (KnownCollectionFormat = {}));
export function getCustomRequestHeaderNameForOperation(route) {
    const params = route.parameters.parameters.filter(isCustomClientRequestIdParam);
    if (params.length > 0) {
        return "client-request-id";
    }
    return undefined;
}
export function isCustomClientRequestIdParam(param) {
    return (param.type === "header" && param.name.toLowerCase() === "client-request-id");
}
export function isIgnoredHeaderParam(param) {
    return (isCustomClientRequestIdParam(param) ||
        (param.type === "header" &&
            ["return-client-request-id", "ocp-date"].includes(param.name.toLowerCase())));
}
export function getMethodHierarchiesMap(context, client) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    const methodQueue = (_b = (_a = client.children) === null || _a === void 0 ? void 0 : _a.filter((p) => {
        return !(p.clientInitialization.initializedBy & InitializedByFlags.Individually);
    }).map((m) => {
        return [[], m];
    })) !== null && _b !== void 0 ? _b : [];
    client.methods.forEach((m) => {
        methodQueue.push([[], m]);
    });
    const operationHierarchiesMap = new Map();
    while (methodQueue.length > 0) {
        const method = methodQueue.pop();
        if (!method) {
            continue;
        }
        const prefixes = ((_c = context.rlcOptions) === null || _c === void 0 ? void 0 : _c.hierarchyClient) === false &&
            ((_d = context.rlcOptions) === null || _d === void 0 ? void 0 : _d.enableOperationGroup) &&
            method[0].length > 0
            ? [method[0][method[0].length - 1]]
            : method[0];
        const operationOrGroup = method[1];
        if (operationOrGroup.kind === "client") {
            operationOrGroup.methods.forEach((m) => methodQueue.push([[...prefixes, operationOrGroup.name], m]));
            const nonIndependentChildren = (_e = operationOrGroup.children) === null || _e === void 0 ? void 0 : _e.filter((child) => {
                return !(child.clientInitialization.initializedBy &
                    InitializedByFlags.Individually);
            });
            if (nonIndependentChildren && nonIndependentChildren.length > 0) {
                nonIndependentChildren.forEach((child) => methodQueue.push([[...prefixes, operationOrGroup.name], child]));
            }
        }
        else {
            const prefixKey = ((_f = context.rlcOptions) === null || _f === void 0 ? void 0 : _f.hierarchyClient) ||
                ((_g = context.rlcOptions) === null || _g === void 0 ? void 0 : _g.enableOperationGroup)
                ? prefixes.join("/")
                : "";
            const groupName = prefixes
                .map((p) => normalizeName(p, NameType.OperationGroup))
                .join("");
            if (((_h = context.rlcOptions) === null || _h === void 0 ? void 0 : _h.hierarchyClient) === false &&
                ((_j = context.rlcOptions) === null || _j === void 0 ? void 0 : _j.enableOperationGroup) &&
                groupName !== "" &&
                !operationOrGroup.name.startsWith(groupName + "_")) {
                operationOrGroup.oriName = operationOrGroup.name;
                operationOrGroup.name = `${groupName}_${operationOrGroup.name}`;
            }
            operationOrGroup.parameters.map((p) => {
                return resolveParameterNameConflict(operationOrGroup, p);
            });
            operationOrGroup.operation.parameters.map((p) => {
                return resolveParameterNameConflict(operationOrGroup, p);
            });
            if (((_k = operationOrGroup.operation.bodyParam) === null || _k === void 0 ? void 0 : _k.name) === operationOrGroup.name) {
                operationOrGroup.operation.bodyParam = resolveParameterNameConflict(operationOrGroup, operationOrGroup.operation.bodyParam);
            }
            const operations = operationHierarchiesMap.get(prefixKey);
            operationHierarchiesMap.set(prefixKey, [
                ...(operations !== null && operations !== void 0 ? operations : []),
                operationOrGroup
            ]);
        }
    }
    return operationHierarchiesMap;
}
export function isTenantLevelOperation(operation, client) {
    var _a, _b;
    // Check if this operation has a subscriptionId path parameter
    const subscriptionIdParam = (_a = operation.operation.parameters) === null || _a === void 0 ? void 0 : _a.find((param) => param.name.toLowerCase() === "subscriptionid" && param.kind === "path");
    if (subscriptionIdParam) {
        // The operation has a client-level subscriptionId parameter, then it's not tenant-level
        if (subscriptionIdParam.onClient) {
            return false;
        }
    }
    else {
        // Skip tenant-level internal ARM APIs
        // Ref: https://armwiki.azurewebsites.net/rpaas/operations_auto_gen.html#special-cases
        const pathLC = operation.operation.path.toLowerCase();
        // Get the provider namespace from the client
        const clientNamespaceLC = client.namespace.toLowerCase();
        if (((_b = operation.crossLanguageDefinitionId) === null || _b === void 0 ? void 0 : _b.toLowerCase()) ===
            "azure.resourcemanager.operations.list" ||
            pathLC.includes(`${clientNamespaceLC}/checknameavailability`)) {
            return false;
        }
    }
    // The operation has no subscriptionId parameter or has method-level subscriptionId parameter
    // Considered as tenant-level
    return true;
}
function resolveParameterNameConflict(operationOrGroup, p) {
    const paramName = normalizeName(p.name, NameType.Parameter, true);
    if (paramName === operationOrGroup.name) {
        p.name = `${paramName}Parameter`;
    }
    else {
        p.name = paramName;
    }
    return p;
}
