// @azure-tools/typespec-ts/dist/src/utils/parameterUtils.js
import { NameType, normalizeName, SchemaContext } from "@azure-tools/rlc-common";
import { getTypeName, isArrayType, isObjectOrDictType } from "./modelUtils.js";
import { reportDiagnostic } from "../lib.js";
import { getEncode, NoTarget } from "@typespec/compiler";
export function getParameterSerializationInfo(dpgContext, parameter, valueSchema, operationGroup = "", operationName = "") {
    var _a;
    const schemaContext = [SchemaContext.Input];
    const retVal = buildSerializationInfo(getTypeName(valueSchema, schemaContext));
    const prefix = `${operationGroup}_${operationName}_${parameter.name}`;
    switch (parameter.type) {
        case "path": {
            if (parameter.allowReserved === true) {
                const wrapperType = buildAllowReserved(normalizeName(`${prefix}_PathParam`, NameType.Interface), valueSchema, parameter.name);
                return buildSerializationInfo(wrapperType);
            }
            return retVal;
        }
        case "query": {
            if (!isArrayType(valueSchema) && !isObjectOrDictType(valueSchema)) {
                // if the value is a primitive type or other types, we will not generate a wrapper type
                return retVal;
            }
            const name = normalizeName(`${prefix}_QueryParam`, NameType.Interface);
            const encode = getEncode(dpgContext.program, parameter.param);
            const format = (encode === null || encode === void 0 ? void 0 : encode.encoding) === "ArrayEncoding.pipeDelimited" ||
                (encode === null || encode === void 0 ? void 0 : encode.encoding) === "ArrayEncoding.spaceDelimited"
                ? encode.encoding.replace("ArrayEncoding.", "")
                : undefined;
            let wrapperType = buildExplodeAndStyle(name, parameter.explode, format !== null && format !== void 0 ? format : "form", valueSchema, parameter.name);
            if (parameter.explode === true) {
                if (format !== undefined) {
                    reportDiagnostic(dpgContext.program, {
                        code: "un-supported-format-cases",
                        format: {
                            paramName: parameter.name,
                            explode: String(parameter.explode),
                            format: format
                        },
                        target: NoTarget
                    });
                }
                if ((_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.compatibilityQueryMultiFormat) {
                    wrapperType = buildUnionType([
                        wrapperType,
                        { type: "string", name: "string" }
                    ]);
                }
                return buildSerializationInfo(wrapperType);
            }
            if (format === undefined) {
                if (isArrayType(valueSchema)) {
                    wrapperType = buildUnionType([valueSchema, wrapperType]);
                }
                return buildSerializationInfo(wrapperType);
            }
            return buildSerializationInfo(wrapperType);
        }
        case "header": {
            return buildSerializationInfo(getHeaderSerializeTypeName(dpgContext.program, valueSchema, schemaContext));
        }
        default:
            return retVal;
    }
}
function buildSerializationInfo(typeOrSchema) {
    if (typeof typeOrSchema === "string") {
        return {
            typeName: typeOrSchema
        };
    }
    return {
        typeName: getTypeName(typeOrSchema),
        wrapperType: typeOrSchema
    };
}
function getHeaderSerializeTypeName(program, schema, usage) {
    var _a;
    const typeName = getTypeName(schema, usage);
    const formattedName = ((_a = schema.alias) !== null && _a !== void 0 ? _a : typeName).replace("Date | string", "string");
    const canSerialize = isSerializable(schema);
    if (canSerialize) {
        return schema.alias ? typeName : formattedName;
    }
    reportDiagnostic(program, {
        code: "unable-serialized-type",
        format: { type: typeName },
        target: NoTarget
    });
    return "string";
    function isSerializable(type) {
        if (type.enum) {
            return type.enum.every((i) => {
                return isSerializable(i) || i.type === "null";
            });
        }
        return (["string", "number", "boolean"].includes(type.type) || type.isConstant);
    }
}
function buildAllowReserved(typeName, valueSchema, parameterName) {
    var _a;
    return {
        type: "object",
        name: typeName,
        description: `This is the wrapper object for the parameter \`${parameterName}\` with allowReserved set to true.`,
        properties: {
            value: {
                ...valueSchema,
                description: (_a = valueSchema.description) !== null && _a !== void 0 ? _a : "Value of the parameter",
                required: true
            },
            allowReserved: {
                type: `true`,
                description: "Whether to allow reserved characters",
                required: true
            }
        }
    };
}
function buildExplodeAndStyle(typeName, explode, style, valueSchema, parameterName) {
    var _a;
    return {
        type: "object",
        name: typeName,
        description: `This is the wrapper object for the parameter \`${parameterName}\` with explode set to ${explode} and style set to ${style}.`,
        properties: {
            value: {
                ...valueSchema,
                description: (_a = valueSchema.description) !== null && _a !== void 0 ? _a : "Value of the parameter",
                required: true
            },
            explode: {
                type: `${explode}`,
                description: "Should we explode the value?",
                required: true
            },
            style: {
                type: `"${style}"`,
                description: "Style of the value",
                required: true
            }
        }
    };
}
function buildUnionType(values) {
    return {
        name: "",
        enum: values,
        type: "union",
        typeName: `${values.map((v) => getTypeName(v)).join(" | ")}`,
        required: true,
        description: undefined
    };
}
