// @azure-tools/typespec-ts/dist/src/framework/load-static-helpers.js
import { readdir, stat, readFile } from "fs/promises";
import * as path from "path";
import { refkey } from "./refkey.js";
import { resolveProjectRoot } from "../utils/resolve-project-root.js";
import { isAzurePackage } from "@azure-tools/rlc-common";
import { NoTarget } from "@typespec/compiler";
import { reportDiagnostic } from "../lib.js";
export const SourceFileSymbol = Symbol("SourceFile");
export function isStaticHelperMetadata(metadata) {
    return Boolean(metadata &&
        metadata.name &&
        metadata.kind &&
        metadata.location &&
        metadata[SourceFileSymbol]);
}
const DEFAULT_SOURCES_STATIC_HELPERS_PATH = "static/static-helpers";
const DEFAULT_SOURCES_TESTING_HELPERS_PATH = "static/test-helpers";
export async function loadStaticHelpers(project, helpers, options = {}) {
    var _a, _b, _c, _d, _e;
    const helpersMap = new Map();
    // Load static helpers used in sources code
    const defaultStaticHelpersPath = path.join(resolveProjectRoot(), DEFAULT_SOURCES_STATIC_HELPERS_PATH);
    const filesInSources = await traverseDirectory((_a = options.helpersAssetDirectory) !== null && _a !== void 0 ? _a : defaultStaticHelpersPath, options.program);
    await loadFiles(filesInSources, (_b = options.sourcesDir) !== null && _b !== void 0 ? _b : "");
    // Load static helpers used in testing code (only when loadTestHelpers is enabled)
    if ((_c = options.loadTestHelpers) !== null && _c !== void 0 ? _c : (((_d = options.options) === null || _d === void 0 ? void 0 : _d.generateTest) &&
        isAzurePackage({ options: options.options }))) {
        const defaultTestingHelpersPath = path.join(resolveProjectRoot(), DEFAULT_SOURCES_TESTING_HELPERS_PATH);
        const filesInTestings = await traverseDirectory(defaultTestingHelpersPath, options.program, [], "", "test/generated/util");
        await loadFiles(filesInTestings, (_e = options.rootDir) !== null && _e !== void 0 ? _e : "");
    }
    return assertAllHelpersLoadedPresent(helpersMap);
    async function loadFiles(files, generateDir) {
        for (const file of files) {
            const targetPath = path.join(generateDir, file.target);
            const contents = await readFile(file.source, "utf-8");
            const addedFile = project.createSourceFile(targetPath, contents, {
                overwrite: true
            });
            addedFile.getImportDeclarations().map((i) => {
                if (!isAzurePackage({ options: options.options })) {
                    if (i
                        .getModuleSpecifier()
                        .getFullText()
                        .includes("@azure/core-rest-pipeline")) {
                        i.setModuleSpecifier("@typespec/ts-http-runtime");
                    }
                    if (i
                        .getModuleSpecifier()
                        .getFullText()
                        .includes("@azure-rest/core-client")) {
                        i.setModuleSpecifier("@typespec/ts-http-runtime");
                    }
                }
            });
            for (const entry of Object.values(helpers)) {
                if (!addedFile.getFilePath().endsWith(entry.location)) {
                    continue;
                }
                const declaration = getDeclarationByMetadata(addedFile, entry);
                if (!declaration) {
                    throw new Error(`Declaration ${entry.name} not found in file ${addedFile.getFilePath()}\n This is an Emitter bug, make sure that the map of static helpers passed to loadStaticHelpers matches what is in the file.`);
                }
                entry[SourceFileSymbol] = addedFile;
                helpersMap.set(refkey(entry), entry);
            }
        }
    }
}
function assertAllHelpersLoadedPresent(helpers) {
    const missingHelpers = [];
    for (const helper of helpers.values()) {
        if (!helper[SourceFileSymbol]) {
            missingHelpers.push(helper);
        }
    }
    if (missingHelpers.length > 0) {
        const missingHelpersString = missingHelpers
            .map((helper) => `${helper.name} - ${helper.location}`)
            .join("\n");
        throw new Error(`The following helpers were not found in the project, make sure they are defined in the expected static helper file: ${missingHelpersString}`);
    }
    return helpers;
}
function getDeclarationByMetadata(file, declaration) {
    switch (declaration.kind) {
        case "class":
            return file.getClass(declaration.name);
        case "function":
            return file.getFunction(declaration.name);
        case "interface":
            return file.getInterface(declaration.name);
        case "typeAlias":
            return file.getTypeAlias(declaration.name);
        case "enum":
            return file.getEnum(declaration.name);
        default:
            throw new Error(`invalid helper kind ${declaration.kind}\nAll helpers provided to loadStaticHelpers are of kind: function, interface, typeAlias, class`);
    }
}
const _targetStaticHelpersBaseDir = "static-helpers";
async function traverseDirectory(directory, program, result = [], relativePath = "", targetBaseDir = _targetStaticHelpersBaseDir) {
    try {
        const files = await readdir(directory);
        await Promise.all(files.map(async (file) => {
            const filePath = path.join(directory, file);
            const fileStat = await stat(filePath);
            if (fileStat.isDirectory()) {
                await traverseDirectory(filePath, program, result, path.join(relativePath, file), targetBaseDir);
            }
            else if (fileStat.isFile() &&
                !file.endsWith(".d.ts") &&
                /.*\..?ts$/.test(file)) {
                const target = path.join(targetBaseDir, relativePath, file);
                result.push({ source: filePath, target });
            }
        }));
        return result;
    }
    catch (error) {
        if (program) {
            reportDiagnostic(program, {
                code: "directory-traversal-error",
                format: { directory, error: String(error) },
                target: NoTarget
            });
        }
        throw error;
    }
}
