// @azure-tools/typespec-ts/dist/src/modular/serialization/serializeUtils.js
import { getAllAncestors } from "../helpers/operationHelpers.js";
import { getDirectSubtypes } from "../helpers/typeHelpers.js";
/**
 * Get all discriminated values for a given model type and given discriminator property.
 * @param type The model type to get discriminated values from.
 * @param discriminatorProperty The discriminator property to use for filtering.
 * @returns An array of all discriminated values.
 */
export function getAllDiscriminatedValues(type, discriminatorProperty) {
    var _a, _b, _c;
    if (!type.discriminatorValue) {
        return [];
    }
    const values = [];
    const children = [type];
    while (children.length > 0) {
        const model = children.shift();
        // skip if no discriminator value
        if (!model.discriminatorValue) {
            continue;
        }
        // Check if this model is a subtype of the given discriminator property
        if (!!discriminatorProperty &&
            ((_b = (_a = model.baseModel) === null || _a === void 0 ? void 0 : _a.discriminatorProperty) === null || _b === void 0 ? void 0 : _b.name) ===
                (discriminatorProperty === null || discriminatorProperty === void 0 ? void 0 : discriminatorProperty.name)) {
            values.push(model.discriminatorValue);
            if (((_c = model.discriminatorProperty) === null || _c === void 0 ? void 0 : _c.name) === discriminatorProperty.name) {
                // Traverse subtypes also if they have the same discriminator property
                children.push(...getDirectSubtypes(model));
            }
        }
    }
    return values;
}
export function isSupportedSerializeType(type) {
    return (type.kind === "model" ||
        type.kind === "dict" ||
        type.kind === "array" ||
        type.kind === "union");
}
/**
 *
 * In general, we have two kinds of basic special union variants.
 * 1. datetime type
 * 2. bytes type
 * If we consider model type, we have the following three type.
 * 3. model with property of datetime type.
 * 4. model with property of binary array type.
 * 5. model that has different property name between rest layer and modular layer.
 * 6. nested model i.e. model with property that is a model with one of the above three conditions.
 * If we consider array type, with all the above 6 types as the element types.
 */
const specialVariantMap = new Map();
export function isSpecialUnionVariant(t, variantStack = []) {
    var _a, _b, _c, _d, _e;
    if (variantStack.length <= 0) {
        variantStack.push(t);
    }
    const ancestors = getAllAncestors(t);
    if (specialVariantMap.has(t)) {
        variantStack.pop();
        return (_a = specialVariantMap.get(t)) !== null && _a !== void 0 ? _a : false;
    }
    if (t.kind === "utcDateTime" ||
        t.kind === "plainDate" ||
        t.kind === "bytes" ||
        (t.kind === "model" &&
            ((_c = (_b = t.properties) === null || _b === void 0 ? void 0 : _b.filter((p) => {
                return !(variantStack.includes(p.type) ||
                    ancestors.includes(p.type) ||
                    (p.type.kind === "array" &&
                        p.type.valueType &&
                        (variantStack.includes(p.type.valueType) ||
                            ancestors.includes(p.type.valueType))));
            })) === null || _c === void 0 ? void 0 : _c.some((p) => 
            // eslint-disable-next-line
            (p.kind === "property" && p.name !== p.serializedName) ||
                isSpecialUnionVariant(p.type, [...variantStack, p.type])))) ||
        isPolymorphicUnion(t) ||
        (t.kind === "array" &&
            t.valueType &&
            !variantStack.includes(t.valueType) &&
            !ancestors.includes(t.valueType) &&
            isSpecialUnionVariant(t.valueType, [...variantStack, t.valueType])) ||
        (t.kind === "union" &&
            ((_e = (_d = t.variantTypes) === null || _d === void 0 ? void 0 : _d.filter((p) => {
                return !(variantStack.includes(p) || ancestors.includes(p));
            })) === null || _e === void 0 ? void 0 : _e.some((p) => {
                return isSpecialUnionVariant(p, [...variantStack, p]);
            }))) ||
        (t.kind === "enum" && !(t.isFixed || t.isNonExhaustive))) {
        specialVariantMap.set(t, true);
        variantStack.pop();
        return true;
    }
    variantStack.pop();
    specialVariantMap.set(t, false);
    return false;
}
export function isNormalUnion(t) {
    var _a, _b;
    return (t.kind === "union" &&
        !((_b = (_a = t.variantTypes) === null || _a === void 0 ? void 0 : _a.some((p) => {
            return isSpecialUnionVariant(p);
        })) !== null && _b !== void 0 ? _b : false));
}
export function isDiscriminatedUnion(type) {
    if (!type) {
        return false;
    }
    return Boolean((type === null || type === void 0 ? void 0 : type.kind) === "model" &&
        type.discriminatorProperty &&
        type.discriminatedSubtypes);
}
export function isSpecialHandledUnion(t) {
    return isDiscriminatedUnion(t) || isPolymorphicUnion(t);
}
const polymorphicUnionMap = new Map();
export function isPolymorphicUnion(t) {
    var _a, _b, _c;
    if (polymorphicUnionMap.has(t)) {
        return (_a = polymorphicUnionMap.get(t)) !== null && _a !== void 0 ? _a : false;
    }
    const ancestors = getAllAncestors(t);
    if (t.kind === "model" &&
        t.discriminatedSubtypes &&
        ((_c = (_b = Object.values(t.discriminatedSubtypes)) === null || _b === void 0 ? void 0 : _b.filter((p) => {
            return !ancestors.includes(p);
        })) === null || _c === void 0 ? void 0 : _c.some((p) => {
            return isSpecialUnionVariant(p);
        }))) {
        polymorphicUnionMap.set(t, true);
        return true;
    }
    polymorphicUnionMap.set(t, false);
    return false;
}
export function getPropertyWithOverrides(property, overrides) {
    var _a;
    if (!overrides) {
        return property;
    }
    let updatedProperty = property;
    if (overrides.allOptional) {
        updatedProperty = { ...updatedProperty, optional: true };
    }
    const renamedClientName = (_a = overrides.propertyRenames) === null || _a === void 0 ? void 0 : _a.get(property);
    if (renamedClientName) {
        updatedProperty = { ...updatedProperty, name: renamedClientName };
    }
    return updatedProperty;
}
