// @azure-tools/rlc-common/dist/metadata/buildApiExtractorConfig.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildApiExtractorConfig = buildApiExtractorConfig;
const ts_morph_1 = require("ts-morph");
function buildApiExtractorConfig(model) {
    var _a, _b;
    const { packageDetails, isModularLibrary, generateTest, azureSdkForJs } = model.options || {};
    const project = new ts_morph_1.Project();
    let mainEntryPointFilePath = "dist/esm/index.d.ts";
    if (((_a = model.options) === null || _a === void 0 ? void 0 : _a.moduleKind) === "cjs") {
        mainEntryPointFilePath = `./types${generateTest || isModularLibrary ? "/src" : ""}/index.d.ts`;
    }
    const config = azureSdkForJs
        ? {
            extends: "../../../api-extractor-base.json"
        }
        : {
            $schema: "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
            mainEntryPointFilePath,
            docModel: {
                enabled: true
            },
            apiReport: {
                enabled: true,
                reportFolder: "./review"
            },
            dtsRollup: {
                enabled: true,
                untrimmedFilePath: "",
                publicTrimmedFilePath: `dist/${(_b = packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.nameWithoutScope) !== null && _b !== void 0 ? _b : packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.name}.d.ts`
            },
            messages: {
                tsdocMessageReporting: {
                    default: {
                        logLevel: "none"
                    }
                },
                extractorMessageReporting: {
                    "ae-missing-release-tag": {
                        logLevel: "none"
                    },
                    "ae-unresolved-link": {
                        logLevel: "none"
                    }
                }
            }
        };
    const filePath = "api-extractor.json";
    const configFile = project.createSourceFile(filePath, JSON.stringify(config), {
        overwrite: true
    });
    return {
        path: filePath,
        content: configFile.getFullText()
    };
}
