// @azure-tools/rlc-common/dist/metadata/buildTestConfig.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTestBrowserTsConfig = buildTestBrowserTsConfig;
exports.buildTestNodeTsConfig = buildTestNodeTsConfig;
exports.buildTestMainTsConfig = buildTestMainTsConfig;
const ts_morph_1 = require("ts-morph");
const utils_js_1 = require("./utils.js");
const highLevelTsTestConfig = {
    references: [
        {
            path: "./tsconfig.test.node.json"
        },
        {
            path: "./tsconfig.browser.config.json"
        }
    ],
    compilerOptions: {
        composite: true
    },
    files: []
};
function shouldGenerateTestConfig(model) {
    var _a, _b, _c, _d;
    const isAzureSdkForJs = (_b = (_a = model.options) === null || _a === void 0 ? void 0 : _a.azureSdkForJs) !== null && _b !== void 0 ? _b : false;
    return !(((_c = model.options) === null || _c === void 0 ? void 0 : _c.generateMetadata) === false ||
        ((_d = model.options) === null || _d === void 0 ? void 0 : _d.generateTest) === false ||
        isAzureSdkForJs !== true);
}
function buildTestBrowserTsConfig(model) {
    if (!shouldGenerateTestConfig(model)) {
        return;
    }
    const browserFilePath = "tsconfig.browser.config.json";
    const project = new ts_morph_1.Project();
    const name = (0, utils_js_1.getPackageName)(model);
    const browserContent = {
        extends: "../../../tsconfig.browser.base.json",
        compilerOptions: {
            paths: {
                [name]: ["./dist/browser/index.d.ts"],
                [`${name}/*`]: ["./dist/browser/*"],
                "$internal/*": ["./dist/browser/*"]
            }
        }
    };
    const browserConfigFile = project.createSourceFile(browserFilePath, JSON.stringify(browserContent, null, 2), {
        overwrite: true
    });
    return {
        path: browserFilePath,
        content: browserConfigFile.getFullText()
    };
}
function buildTestNodeTsConfig(model) {
    if (!shouldGenerateTestConfig(model)) {
        return;
    }
    const testNodeFilePath = "tsconfig.test.node.json";
    const project = new ts_morph_1.Project();
    const name = (0, utils_js_1.getPackageName)(model);
    const testNodeContent = {
        extends: "../../../tsconfig.test.node.base.json",
        compilerOptions: {
            paths: {
                [name]: ["./src/index.ts"],
                [`${name}/*`]: ["./src/*"],
                "$internal/*": ["./src/*"]
            }
        }
    };
    const testNodeConfigFile = project.createSourceFile(testNodeFilePath, JSON.stringify(testNodeContent, null, 2), {
        overwrite: true
    });
    return {
        path: testNodeFilePath,
        content: testNodeConfigFile.getFullText()
    };
}
function buildTestMainTsConfig(model) {
    if (!shouldGenerateTestConfig(model)) {
        return;
    }
    const testFilePath = "tsconfig.test.json";
    return {
        path: testFilePath,
        content: JSON.stringify(highLevelTsTestConfig, null, 2)
    };
}
