// @azure-tools/rlc-common/dist/metadata/buildReadmeFile.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildReadmeFile = buildReadmeFile;
exports.hasClientNameChanged = hasClientNameChanged;
exports.updateReadmeFile = updateReadmeFile;
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore: to fix the handlebars issue
const handlebars_1 = require("handlebars");
const nameUtils_js_1 = require("../helpers/nameUtils.js");
const packageUtil_js_1 = require("../helpers/packageUtil.js");
const nameConstructors_js_1 = require("../helpers/nameConstructors.js");
const fs_1 = require("fs");
const azureReadmeRLCTemplate = `# {{ clientDescriptiveName }} library for JavaScript

{{ description }}

{{#if azureArm}}
**If you are not familiar with our REST client, please spend 5 minutes to take a look at {{#if serviceDocURL}}[the service's documentation]({{ serviceDocURL }}) and {{/if}}our [REST client docs](https://github.com/Azure/azure-sdk-for-js/blob/main/documentation/rest-clients.md) to use this library, the REST client provides a light-weighted & developer friendly way to call azure rest api
{{else}}
**Please rely heavily on {{#if serviceDocURL}}[the service's documentation]({{ serviceDocURL }}) and {{/if}}our [REST client docs](https://github.com/Azure/azure-sdk-for-js/blob/main/documentation/rest-clients.md) to use this library**
{{/if}}

Key links:

{{#if packageSourceURL}}
- [Source code]({{ packageSourceURL }})
{{/if}}
{{#if packageNPMURL}}
- [Package (NPM)]({{ packageNPMURL }})
{{/if}}
{{#if apiRefURL}}
- [API reference documentation]({{ apiRefURL }})
{{/if}}
{{#if serviceDocURL}}
- [Product documentation]({{ serviceDocURL }})
{{/if}}
{{#if samplesURL}}
- [Samples]({{ samplesURL }})
{{/if}}

## Getting started

### Currently supported environments

- LTS versions of Node.js

### Prerequisites

- You must have an [Azure subscription](https://azure.microsoft.com/free/){{#if dependencyLink}} and follow [these]({{ dependencyLink }}) instructions{{/if}} to use this package.

### Install the \`{{ clientPackageName }}\` package

Install the {{ clientDescriptiveName }} REST client library for JavaScript with \`npm\`:

\`\`\`bash
npm install {{ clientPackageName }}
\`\`\`

### Create and authenticate a \`{{ clientClassName }}\`

To use an [Azure Active Directory (AAD) token credential](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/samples/AzureIdentityExamples.md#authenticating-with-a-pre-fetched-access-token),
provide an instance of the desired credential type obtained from the
[@azure/identity](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/identity/identity#credentials) library.

To authenticate with AAD, you must first \`npm\` install [\`@azure/identity\`](https://www.npmjs.com/package/@azure/identity) {{#if dependencyLink}}and
[{{dependencyDescription }}]({{ dependencyLink }}){{/if}}

After setup, you can choose which type of [credential](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/identity/identity#credentials) from \`@azure/identity\` to use.
As an example, [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/identity/identity#defaultazurecredential)
can be used to authenticate the client.

## Troubleshooting

### Logging

Enabling logging may help uncover useful information about failures. In order to see a log of HTTP requests and responses, set the \`AZURE_LOG_LEVEL\` environment variable to \`info\`. Alternatively, logging can be enabled at runtime by calling \`setLogLevel\` in the \`@azure/logger\`:

\`\`\`ts {{#if azureSdkForJs}}{{#if generateTest}}snippet:SetLogLevel{{/if}}{{/if}}
import { setLogLevel } from "@azure/logger";

setLogLevel("info");
\`\`\`

For more detailed instructions on how to enable logs, you can look at the [@azure/logger package docs](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/core/logger).
`;
const azureReadmeModularTemplate = `# {{ clientDescriptiveName }} library for JavaScript

This package contains an isomorphic SDK (runs both in Node.js and in browsers) for {{ clientDescriptiveName }}.

{{ description }}

Key links:

{{#if packageSourceURL}}
- [Source code]({{ packageSourceURL }})
{{/if}}
{{#if packageNPMURL}}
- [Package (NPM)]({{ packageNPMURL }})
{{/if}}
{{#if apiRefURL}}
- [API reference documentation]({{ apiRefURL }})
{{/if}}
{{#if samplesURL}}
- [Samples]({{samplesURL}})
{{/if}}

## Getting started

### Currently supported environments

- [LTS versions of Node.js](https://github.com/nodejs/release#release-schedule)
- Latest versions of Safari, Chrome, Edge and Firefox.

See our [support policy](https://github.com/Azure/azure-sdk-for-js/blob/main/SUPPORT.md) for more details.

{{#if azure}}
### Prerequisites

- An [Azure subscription][azure_sub].
{{/if}}

{{#if isReleasablePackage}}
### Install the \`{{ clientPackageName }}\` package

Install the {{ clientDescriptiveName }} library for JavaScript with \`npm\`:

\`\`\`bash
npm install {{ clientPackageName }}
\`\`\`
{{/if}}

{{#if azure}}
{{#if addCredentials}}
### Create and authenticate a \`{{ clientClassName}}\`

To create a client object to access the {{ serviceName }} API, you will need the \`endpoint\` of your {{ serviceName }} resource and a \`credential\`. The {{ clientDescriptiveName }} can use Azure Active Directory credentials to authenticate.
You can find the endpoint for your {{ serviceName }} resource in the [Azure Portal][azure_portal].

You can authenticate with Azure Active Directory using a credential from the [@azure/identity][azure_identity] library or [an existing AAD Token](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/samples/AzureIdentityExamples.md#authenticating-with-a-pre-fetched-access-token).

To use the [DefaultAzureCredential][defaultazurecredential] provider shown below, or other credential providers provided with the Azure SDK, please install the \`@azure/identity\` package:

\`\`\`bash
npm install @azure/identity
\`\`\`

You will also need to **register a new AAD application and grant access to {{ serviceName}}** by assigning the suitable role to your service principal (note: roles such as \`"Owner"\` will not grant the necessary permissions).

For more information about how to create an Azure AD Application check out [this guide](https://learn.microsoft.com/azure/active-directory/develop/howto-create-service-principal-portal).

{{#if azureArm}}
Using Node.js and Node-like environments, you can use the \`DefaultAzureCredential\` class to authenticate the client.

\`\`\`ts {{#if azureSdkForJs}}{{#if generateTest}}snippet:ReadmeSampleCreateClient_Node{{/if}}{{/if}}
import { {{ clientClassName }} } from "{{ clientPackageName }}";
import { DefaultAzureCredential } from "@azure/identity";

{{#if hasSubscriptionId}}
const subscriptionId = "00000000-0000-0000-0000-000000000000";
const client = new {{ clientClassName }}(new DefaultAzureCredential(), subscriptionId);
{{else}}
const client = new {{ clientClassName }}(new DefaultAzureCredential());
{{/if}}
\`\`\`

For browser environments, use the \`InteractiveBrowserCredential\` from the \`@azure/identity\` package to authenticate.

\`\`\`ts {{#if azureSdkForJs}}{{#if generateTest}}snippet:ReadmeSampleCreateClient_Browser{{/if}}{{/if}}
import { InteractiveBrowserCredential } from "@azure/identity";
import { {{ clientClassName }} } from "{{ clientPackageName }}";

const credential = new InteractiveBrowserCredential({
  tenantId: "<YOUR_TENANT_ID>",
  clientId: "<YOUR_CLIENT_ID>",
 });

{{#if hasSubscriptionId}}
const subscriptionId = "00000000-0000-0000-0000-000000000000";
const client = new {{ clientClassName }}(credential, subscriptionId);
{{else}}
const client = new {{ clientClassName }}(credential);
{{/if}}
\`\`\`
{{else}}
Using Node.js and Node-like environments, you can use the \`DefaultAzureCredential\` class to authenticate the client.

\`\`\`ts {{#if azureSdkForJs}}{{#if generateTest}}snippet:ReadmeSampleCreateClient_Node{{/if}}{{/if}}
import { {{ clientClassName }} } from "{{ clientPackageName }}";
import { DefaultAzureCredential } from "@azure/identity";

const client = new {{ clientClassName }}("<endpoint>", new DefaultAzureCredential());
\`\`\`

For browser environments, use the \`InteractiveBrowserCredential\` from the \`@azure/identity\` package to authenticate.

\`\`\`ts {{#if azureSdkForJs}}{{#if generateTest}}snippet:ReadmeSampleCreateClient_Browser{{/if}}{{/if}}
import { InteractiveBrowserCredential } from "@azure/identity";
import { {{ clientClassName }} } from "{{ clientPackageName }}";

const credential = new InteractiveBrowserCredential({
  tenantId: "<YOUR_TENANT_ID>",
  clientId: "<YOUR_CLIENT_ID>"
 });
const client = new {{ clientClassName }}("<endpoint>", credential);
\`\`\`
{{/if}}
{{/if}}{{/if}}

### JavaScript Bundle
To use this client library in the browser, first you need to use a bundler. For details on how to do this, please refer to our [bundling documentation](https://aka.ms/AzureSDKBundling).

## Key concepts

### {{ clientClassName }}

\`{{ clientClassName }}\` is the primary interface for developers using the {{ clientDescriptiveName }} library. Explore the methods on this client object to understand the different features of the {{ serviceName }} service that you can access.

{{#if azure}}
## Troubleshooting

### Logging

Enabling logging may help uncover useful information about failures. In order to see a log of HTTP requests and responses, set the \`AZURE_LOG_LEVEL\` environment variable to \`info\`. Alternatively, logging can be enabled at runtime by calling \`setLogLevel\` in the \`@azure/logger\`:

\`\`\`ts {{#if azureSdkForJs}}{{#if generateTest}}snippet:SetLogLevel{{/if}}{{/if}}
import { setLogLevel } from "@azure/logger";

setLogLevel("info");
\`\`\`

For more detailed instructions on how to enable logs, you can look at the [@azure/logger package docs]({{ repoURL }}/tree/main/sdk/core/logger).

{{#if samplesURL}}
## Next steps

Please take a look at the [samples]({{ samplesURL }}) directory for detailed examples on how to use this library.
{{/if}}

## Contributing

If you'd like to contribute to this library, please read the [contributing guide]({{ contributingGuideURL }}) to learn more about how to build and test the code.

## Related projects

- [{{ projectName }}]({{ repoURL }})

[azure_sub]: https://azure.microsoft.com/free/
[azure_portal]: https://portal.azure.com
{{#if identityPackageURL}}[azure_identity]: {{ identityPackageURL }}
{{/if}}[defaultazurecredential]: {{ identityPackageURL }}#defaultazurecredential
{{/if}}
`;
const nonBrandedReadmeTemplate = `# {{ clientDescriptiveName }} library for JavaScript

{{ description }}

Key links:

{{#if packageSourceURL}}
- [Source code]({{ packageSourceURL }})
{{/if}}
{{#if packageNPMURL}}
- [Package (NPM)]({{ packageNPMURL }})
{{/if}}
{{#if apiRefURL}}
- [API reference documentation]({{ apiRefURL }})
{{/if}}
{{#if serviceDocURL}}
- [Product documentation]({{ serviceDocURL }})
{{/if}}
{{#if samplesURL}}
- [Samples]({{ samplesURL }})
{{/if}}

## Getting started

### Currently supported environments

- LTS versions of Node.js

### Install the \`{{ clientPackageName }}\` package

Install the {{ clientDescriptiveName }} library for JavaScript with \`npm\`:

\`\`\`bash
npm install {{ clientPackageName }}
\`\`\`
`;
const apiReferenceTemplate = `{{#if apiRefURL}}
- [API reference documentation]({{ apiRefURL }})
{{/if}}
`;
function buildReadmeFile(model) {
    var _a;
    const metadata = (_a = createMetadata(model)) !== null && _a !== void 0 ? _a : {};
    const readmeFileContents = handlebars_1.default.compile(model.options && (0, packageUtil_js_1.isAzurePackage)(model)
        ? model.options.isModularLibrary
            ? azureReadmeModularTemplate
            : azureReadmeRLCTemplate
        : nonBrandedReadmeTemplate, { noEscape: true });
    return {
        path: "README.md",
        content: readmeFileContents(metadata)
    };
}
function hasClientNameChanged(model, existingReadmeFilePath) {
    try {
        const existingContent = (0, fs_1.readFileSync)(existingReadmeFilePath, "utf8");
        const importMatch = existingContent.match(/import\s*\{\s*([A-Za-z0-9_]+)\s*\}\s*from\s*["'][^"']*["']/);
        const existingClientName = importMatch === null || importMatch === void 0 ? void 0 : importMatch[1];
        const newClientName = (0, nameConstructors_js_1.getClientName)(model);
        return !!existingClientName && existingClientName !== newClientName;
    }
    catch (_a) {
        return false;
    }
}
function updateReadmeFile(model, existingReadmeFilePath) {
    var _a;
    try {
        const existingContent = (0, fs_1.readFileSync)(existingReadmeFilePath, "utf8");
        const metadata = (_a = createMetadata(model)) !== null && _a !== void 0 ? _a : {};
        const newApiRefLink = handlebars_1.default
            .compile(apiReferenceTemplate, { noEscape: true })(metadata)
            .trim();
        if (!newApiRefLink) {
            return { path: "README.md", content: existingContent };
        }
        const apiRefRegex = /^- \[API reference documentation\]\(https:\/\/learn\.microsoft\.com\/javascript\/api\/[^)]+\)$/m;
        const updatedContent = existingContent.replace(apiRefRegex, (match) => match ? newApiRefLink : match);
        return { path: "README.md", content: updatedContent };
    }
    catch (_b) {
        return;
    }
}
/**
 * Returns meta data information about the service, the package, and the client.
 * @param codeModel - include the client details
 * @returns inferred metadata about the service, the package, and the client
 */
function createMetadata(model) {
    var _a, _b, _c, _d, _e;
    if (!model.options || !model.options.packageDetails) {
        return;
    }
    // const packageDetails = model.options.packageDetails;
    const { packageDetails, azureOutputDirectory, productDocLink, dependencyInfo, multiClient, batch, serviceInfo, isTypeSpecTest } = model.options;
    const azureHuh = (packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.scopeName) === "azure" ||
        (packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.scopeName) === "azure-rest";
    const repoURL = "https://github.com/Azure/azure-sdk-for-js";
    const relativePackageSourcePath = azureOutputDirectory;
    const packageSourceURL = relativePackageSourcePath &&
        repoURL &&
        `${repoURL}/tree/main/${relativePackageSourcePath}`;
    const clientPackageName = packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.name;
    const clientClassName = (0, nameConstructors_js_1.getClientName)(model);
    const serviceName = getServiceName(model);
    let apiRefUrlQueryParameter = "";
    if (!(packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.isVersionUserProvided) &&
        ((_a = model.apiVersionInfo) === null || _a === void 0 ? void 0 : _a.defaultValue)) {
        if ((_c = (_b = model.apiVersionInfo) === null || _b === void 0 ? void 0 : _b.defaultValue) === null || _c === void 0 ? void 0 : _c.toLowerCase().includes("preview")) {
            apiRefUrlQueryParameter = "?view=azure-node-preview";
        }
    }
    else {
        packageDetails.version = (_d = packageDetails.version) !== null && _d !== void 0 ? _d : "1.0.0-beta.1";
        if (packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.version.includes("beta")) {
            apiRefUrlQueryParameter = "?view=azure-node-preview";
        }
    }
    return {
        serviceName,
        clientClassName,
        clientPackageName: clientPackageName,
        clientDescriptiveName: model.options.isModularLibrary
            ? `${serviceName} client`
            : `${serviceName} REST client`,
        description: (_e = serviceInfo === null || serviceInfo === void 0 ? void 0 : serviceInfo.description) !== null && _e !== void 0 ? _e : packageDetails.description,
        serviceDocURL: productDocLink,
        packageSourceURL: packageSourceURL,
        packageNPMURL: `https://www.npmjs.com/package/${clientPackageName}`,
        samplesURL: model.options.generateSample && packageSourceURL
            ? `${packageSourceURL}/samples`
            : undefined,
        apiRefURL: azureHuh
            ? `https://learn.microsoft.com/javascript/api/${clientPackageName}${apiRefUrlQueryParameter}`
            : undefined,
        dependencyDescription: dependencyInfo === null || dependencyInfo === void 0 ? void 0 : dependencyInfo.description,
        dependencyLink: dependencyInfo === null || dependencyInfo === void 0 ? void 0 : dependencyInfo.link,
        hasMultiClients: multiClient && batch && batch.length > 1,
        azureArm: Boolean(model.options.azureArm),
        azure: azureHuh,
        isReleasablePackage: !isTypeSpecTest,
        repoURL: repoURL,
        projectName: azureHuh ? "Microsoft Azure SDK for JavaScript" : undefined,
        identityPackageURL: repoURL && `${repoURL}/tree/main/sdk/identity/identity`,
        addCredentials: model.options.addCredentials,
        contributingGuideURL: repoURL && `${repoURL}/blob/main/CONTRIBUTING.md`,
        azureSdkForJs: model.options.azureSdkForJs,
        generateTest: model.options.generateTest,
        hasSubscriptionId: model.options.hasSubscriptionId
    };
}
function getServiceName(model) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
    const azureHuh = ((_b = (_a = model === null || model === void 0 ? void 0 : model.options) === null || _a === void 0 ? void 0 : _a.packageDetails) === null || _b === void 0 ? void 0 : _b.scopeName) === "azure" ||
        ((_d = (_c = model === null || model === void 0 ? void 0 : model.options) === null || _c === void 0 ? void 0 : _c.packageDetails) === null || _d === void 0 ? void 0 : _d.scopeName) === "azure-rest";
    const libraryName = model.libraryName;
    const serviceTitle = ((_e = model.options) === null || _e === void 0 ? void 0 : _e.isModularLibrary)
        ? model.libraryName
        : ((_h = (_g = (_f = model.options) === null || _f === void 0 ? void 0 : _f.serviceInfo) === null || _g === void 0 ? void 0 : _g.title) !== null && _h !== void 0 ? _h : model.libraryName);
    const batch = (_j = model === null || model === void 0 ? void 0 : model.options) === null || _j === void 0 ? void 0 : _j.batch, packageDetails = (_k = model === null || model === void 0 ? void 0 : model.options) === null || _k === void 0 ? void 0 : _k.packageDetails;
    let simpleServiceName = batch && batch.length > 1
        ? (0, nameUtils_js_1.normalizeName)((_m = (_l = packageDetails.nameWithoutScope) !== null && _l !== void 0 ? _l : packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.name) !== null && _m !== void 0 ? _m : "", nameUtils_js_1.NameType.Class)
        : (0, nameUtils_js_1.normalizeName)(serviceTitle, nameUtils_js_1.NameType.Class);
    simpleServiceName =
        /**
         * It is a required convention in Azure swaggers for their titles to end with
         * "Client".
         */
        (_v = (_t = (_r = (_p = (_o = serviceTitle.match(/(.*) Client/)) === null || _o === void 0 ? void 0 : _o[1]) !== null && _p !== void 0 ? _p : (_q = serviceTitle.match(/(.*)Client/)) === null || _q === void 0 ? void 0 : _q[1]) !== null && _r !== void 0 ? _r : (_s = libraryName.match(/(.*)Client/)) === null || _s === void 0 ? void 0 : _s[1]) !== null && _t !== void 0 ? _t : (_u = serviceTitle.match(/(.*) Service/)) === null || _u === void 0 ? void 0 : _u[1]) !== null && _v !== void 0 ? _v : simpleServiceName;
    return azureHuh
        ? simpleServiceName.startsWith("Azure")
            ? simpleServiceName
            : `Azure ${simpleServiceName}`
        : simpleServiceName;
}
