// @azure-tools/rlc-common/dist/metadata/buildWarpConfig.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.WarpConfigTemplate = void 0;
exports.buildWarpConfig = buildWarpConfig;
const packageUtil_js_1 = require("../helpers/packageUtil.js");
exports.WarpConfigTemplate = `# warp.config.yml — build configuration

exports:
{{exports}}

targets:
  - name: browser
    tsconfig: "../../../tsconfig.src.browser.json"
    polyfillSuffix: "-browser"

  - name: react-native
    tsconfig: "../../../tsconfig.src.react-native.json"
    polyfillSuffix: "-react-native"

  - name: esm
    condition: import
    tsconfig: "../../../tsconfig.src.esm.json"

  - name: commonjs
    condition: require
    tsconfig: "../../../tsconfig.src.cjs.json"
`;
/**
 * Builds a warp.config.yml file for Azure SDK monorepo packages.
 * Only generated when azureSdkForJs is true.
 */
function buildWarpConfig(model, { exports } = {}) {
    var _a;
    if (!(0, packageUtil_js_1.isAzureMonorepoPackage)(model)) {
        return;
    }
    if (((_a = model.options) === null || _a === void 0 ? void 0 : _a.moduleKind) !== "esm") {
        return;
    }
    const allExports = {
        "./package.json": "./package.json",
        ".": "./src/index.ts",
        ...exports
    };
    const exportsContent = Object.entries(allExports)
        .map(([key, value]) => `  ${JSON.stringify(key)}: ${JSON.stringify(value)}`)
        .join("\n");
    const content = exports.WarpConfigTemplate.replace("{{exports}}", exportsContent);
    return {
        path: "warp.config.yml",
        content
    };
}
