// @azure-tools/rlc-common/dist/metadata/buildPackageFile.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildPackageFile = buildPackageFile;
exports.updatePackageFile = updatePackageFile;
const nameUtils_js_1 = require("../helpers/nameUtils.js");
const operationHelpers_js_1 = require("../helpers/operationHelpers.js");
const packageUtil_js_1 = require("../helpers/packageUtil.js");
const packageCommon_js_1 = require("./packageJson/packageCommon.js");
const ts_morph_1 = require("ts-morph");
const buildAzureMonorepoPackage_js_1 = require("./packageJson/buildAzureMonorepoPackage.js");
const buildAzureStandalonePackage_js_1 = require("./packageJson/buildAzureStandalonePackage.js");
const buildFlavorlessPackage_js_1 = require("./packageJson/buildFlavorlessPackage.js");
const pathUtils_js_1 = require("../helpers/pathUtils.js");
const utils_js_1 = require("./utils.js");
function buildPackageFile(model, { exports, dependencies, clientContextPaths } = {}) {
    var _a, _b, _c, _d, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    const config = {
        description: getDescription(model),
        moduleKind: (_b = (_a = model.options) === null || _a === void 0 ? void 0 : _a.moduleKind) !== null && _b !== void 0 ? _b : "esm",
        name: (0, utils_js_1.getPackageName)(model),
        version: getPackageVersion(model),
        withSamples: ((_c = model.options) === null || _c === void 0 ? void 0 : _c.generateSample) === true,
        withTests: ((_d = model.options) === null || _d === void 0 ? void 0 : _d.generateTest) === true,
        nameWithoutScope: (_g = (_f = model.options) === null || _f === void 0 ? void 0 : _f.packageDetails) === null || _g === void 0 ? void 0 : _g.nameWithoutScope,
        exports,
        azureArm: (_h = model.options) === null || _h === void 0 ? void 0 : _h.azureArm,
        isModularLibrary: (_k = (_j = model.options) === null || _j === void 0 ? void 0 : _j.isModularLibrary) !== null && _k !== void 0 ? _k : false,
        azureSdkForJs: (_l = model.options) === null || _l === void 0 ? void 0 : _l.azureSdkForJs
    };
    let packageInfo = (0, buildFlavorlessPackage_js_1.buildFlavorlessPackage)(config);
    const extendedConfig = {
        ...config,
        clientFilePaths: [getClientFilePath(model)],
        hasLro: (0, operationHelpers_js_1.hasPollingOperations)(model),
        monorepoPackageDirectory: (_m = model.options) === null || _m === void 0 ? void 0 : _m.azureOutputDirectory,
        specSource: (_p = (_o = model.options) === null || _o === void 0 ? void 0 : _o.sourceFrom) !== null && _p !== void 0 ? _p : "TypeSpec",
        dependencies,
        clientContextPaths
    };
    if ((0, packageUtil_js_1.isAzureMonorepoPackage)(model)) {
        packageInfo = (0, buildAzureMonorepoPackage_js_1.buildAzureMonorepoPackage)(extendedConfig);
    }
    if ((0, packageUtil_js_1.isAzureStandalonePackage)(model)) {
        packageInfo = (0, buildAzureStandalonePackage_js_1.buildAzureStandalonePackage)(extendedConfig);
    }
    const project = new ts_morph_1.Project();
    const filePath = "package.json";
    if (!packageInfo) {
        return;
    }
    const packageFile = project.createSourceFile(filePath, JSON.stringify(packageInfo, null, 2), {
        overwrite: true
    });
    return {
        path: filePath,
        content: packageFile.getFullText()
    };
}
/**
 * Automatically updates the package.json with correct paging and LRO dependencies for Azure SDK.
 * Also updates tshy.exports if provided.
 */
function updatePackageFile(model, existingFilePathOrContent, { exports, clientContextPaths } = {}) {
    var _a, _b;
    const hasLro = (0, operationHelpers_js_1.hasPollingOperations)(model);
    const isAzure = (0, packageUtil_js_1.isAzurePackage)(model);
    const needsLroUpdate = isAzure && hasLro;
    const needsExportsUpdate = exports;
    const needsConstantPathsUpdate = clientContextPaths && clientContextPaths.length > 0;
    // Early return if nothing needs to be updated
    if (!needsLroUpdate && !needsExportsUpdate && !needsConstantPathsUpdate) {
        return;
    }
    let packageInfo;
    if (typeof existingFilePathOrContent === "string") {
        let packageFile;
        try {
            const project = new ts_morph_1.Project();
            packageFile = project.addSourceFileAtPath(existingFilePathOrContent);
        }
        catch (_e) {
            // If the file doesn't exist, we don't need to update it.
            return;
        }
        packageInfo = JSON.parse(packageFile.getFullText());
    }
    else {
        packageInfo = existingFilePathOrContent;
    }
    // Update exports based on build system (warp for monorepo, tshy for others)
    if (needsExportsUpdate) {
        if ((_a = model.options) === null || _a === void 0 ? void 0 : _a.azureSdkForJs) {
            // Warp: update resolved exports in package.json
            packageInfo.exports = (0, packageCommon_js_1.resolveWarpExports)(exports);
        }
        else if (packageInfo.tshy) {
            // Tshy: update tshy.exports in package.json
            const newTshy = (0, packageCommon_js_1.getTshyConfig)({
                exports,
                azureSdkForJs: (_b = model.options) === null || _b === void 0 ? void 0 : _b.azureSdkForJs
            });
            packageInfo.tshy.exports = newTshy.exports;
        }
    }
    // Update LRO dependencies for Azure packages
    if (needsLroUpdate) {
        packageInfo.dependencies = {
            ...packageInfo.dependencies,
            "@azure/core-lro": "^3.1.0",
            "@azure/abort-controller": "^2.1.2"
        };
    }
    // Update constantPaths metadata for Azure packages
    if (needsConstantPathsUpdate && isAzure && packageInfo["//metadata"]) {
        const metadata = packageInfo["//metadata"];
        // Filter out existing userAgentInfo entries
        const nonUserAgentPaths = (metadata.constantPaths || []).filter((item) => item.prefix !== "userAgentInfo");
        // Add new userAgentInfo entries from clientContextPaths
        const newUserAgentPaths = clientContextPaths.map((path) => ({
            path: path,
            prefix: "userAgentInfo"
        }));
        metadata.constantPaths = [...nonUserAgentPaths, ...newUserAgentPaths];
    }
    return {
        path: "package.json",
        content: JSON.stringify(packageInfo, null, 2)
    };
}
function getPackageVersion(model) {
    var _a, _b, _c;
    return (_c = (_b = (_a = model.options) === null || _a === void 0 ? void 0 : _a.packageDetails) === null || _b === void 0 ? void 0 : _b.version) !== null && _c !== void 0 ? _c : "1.0.0-beta.1";
}
function getDescription(model) {
    var _a, _b;
    const description = (_b = (_a = model.options) === null || _a === void 0 ? void 0 : _a.packageDetails) === null || _b === void 0 ? void 0 : _b.description;
    if (!description) {
        return `A generated SDK for ${model.libraryName}.`;
    }
    return description;
}
function getClientFilePath(model) {
    const { srcPath } = model;
    const sdkReletivePart = (0, pathUtils_js_1.getRelativePartFromSrcPath)(srcPath);
    const clientFilename = (0, nameUtils_js_1.normalizeName)(model.libraryName, nameUtils_js_1.NameType.File);
    return sdkReletivePart
        ? `src/${sdkReletivePart}/${clientFilename}.ts`
        : `src/${clientFilename}.ts`;
}
