// @azure-tools/rlc-common/dist/metadata/buildTsConfig.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTsConfig = buildTsConfig;
exports.buildTsSrcConfig = buildTsSrcConfig;
exports.buildTsSampleConfig = buildTsSampleConfig;
exports.buildTsTestConfig = buildTsTestConfig;
exports.buildTsSnippetsConfig = buildTsSnippetsConfig;
const ts_morph_1 = require("ts-morph");
const restLevelTsConfigInAzureSdkForJs = function () {
    return {
        references: [
            {
                path: "./tsconfig.src.json"
            }
        ]
    };
};
const tsSrcConfigInAzureSdkForJs = `{
    extends: "../../../tsconfig.lib.json"
  }`;
const tsSampleConfigInAzureSdkForJs = function (clientPackageName) {
    return `{
    extends: "../../../tsconfig.samples.base.json",
    compilerOptions: {
      paths: {
        "${clientPackageName}": ["./dist/esm"]
      }
    }
  }`;
};
const tsTestConfigInAzureSdkForJs = `{
    extends: ["./tsconfig.src.json", "../../../tsconfig.test.base.json"]
  }`;
const restLevelTsConfigNotInAzureSdkForJs = function (model) {
    const { options } = model;
    return {
        compilerOptions: {
            target: "ES2017",
            module: (options === null || options === void 0 ? void 0 : options.moduleKind) === "esm" ? "NodeNext" : "es6",
            lib: [],
            declaration: true,
            declarationMap: true,
            inlineSources: true,
            sourceMap: true,
            importHelpers: true,
            strict: true,
            alwaysStrict: true,
            noUnusedLocals: true,
            noUnusedParameters: true,
            noImplicitReturns: true,
            noFallthroughCasesInSwitch: true,
            forceConsistentCasingInFileNames: true,
            moduleResolution: (options === null || options === void 0 ? void 0 : options.moduleKind) === "esm" ? "NodeNext" : "node",
            allowSyntheticDefaultImports: true,
            esModuleInterop: true,
            outDir: (options === null || options === void 0 ? void 0 : options.moduleKind) === "cjs" ? "./dist-esm" : undefined,
            declarationDir: (options === null || options === void 0 ? void 0 : options.moduleKind) === "cjs" ? "./types" : undefined
        },
        include: ["src/**/*.ts"]
    };
};
function buildTsConfig(model) {
    var _a;
    const { packageDetails, azureSdkForJs } = model.options || {};
    const { generateTest, generateSample } = model.options || {};
    // Take the undefined as true by default
    const clientPackageName = (_a = packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.name) !== null && _a !== void 0 ? _a : "";
    const project = new ts_morph_1.Project();
    const restLevelTsConfig = azureSdkForJs
        ? restLevelTsConfigInAzureSdkForJs()
        : restLevelTsConfigNotInAzureSdkForJs(model);
    if (!azureSdkForJs) {
        if (generateTest) {
            restLevelTsConfig.include.push("test/**/*.ts");
        }
        if (generateSample) {
            restLevelTsConfig.include.push("samples-dev/**/*.ts");
            restLevelTsConfig.compilerOptions["paths"] = {};
            restLevelTsConfig.compilerOptions["paths"][clientPackageName] = [
                "./src/index"
            ];
        }
    }
    else {
        if (generateSample) {
            restLevelTsConfig.references.push({
                path: "./tsconfig.samples.json"
            });
        }
        if (generateTest) {
            restLevelTsConfig.references.push({
                path: "./tsconfig.test.json"
            });
            restLevelTsConfig.references.push({
                path: "./tsconfig.snippets.json"
            });
        }
    }
    const filePath = "tsconfig.json";
    const configFile = project.createSourceFile(filePath, JSON.stringify(restLevelTsConfig, null, 2), {
        overwrite: true
    });
    return {
        path: filePath,
        content: configFile.getFullText()
    };
}
function buildTsSrcConfig() {
    return {
        path: "tsconfig.src.json",
        content: tsSrcConfigInAzureSdkForJs
    };
}
function buildTsSampleConfig(model) {
    var _a;
    const { packageDetails } = model.options || {};
    return {
        path: "tsconfig.samples.json",
        content: tsSampleConfigInAzureSdkForJs((_a = packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.name) !== null && _a !== void 0 ? _a : "")
    };
}
function buildTsTestConfig() {
    return {
        path: "tsconfig.test.json",
        content: tsTestConfigInAzureSdkForJs
    };
}
function buildTsSnippetsConfig() {
    return {
        path: "tsconfig.snippets.json",
        content: JSON.stringify({
            extends: ["../../../tsconfig.snippets.base.json"]
        }, null, 2)
    };
}
