// @azure-tools/rlc-common/dist/helpers/apiVersionUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPathApiVersion = extractPathApiVersion;
exports.extractDefinedPosition = extractDefinedPosition;
/**
 * Extract the path api-version detail from UrlInfo, return undefined if no valid api-version parameter
 * @param urlInfo UrlInfo detail
 * @returns path api-version detail
 */
function extractPathApiVersion(urlInfo) {
    var _a;
    if (!urlInfo) {
        return;
    }
    const param = (_a = urlInfo.urlParameters) === null || _a === void 0 ? void 0 : _a.filter((p) => p.name.toLowerCase() === "api-version" ||
        p.name.toLowerCase() === "apiversion");
    if (!param || (param === null || param === void 0 ? void 0 : param.length) < 1) {
        return;
    }
    const detail = {
        definedPosition: "path",
        isCrossedVersion: Boolean((param === null || param === void 0 ? void 0 : param.length) > 1),
        defaultValue: param.length === 1 ? param[0].value : undefined,
        required: true
    };
    return detail;
}
/**
 * Extract the final position value from api-version in query and path defined.
 * it could be in either the url or the operation level,
 * and in operation level, it could be either path or query.
 * @param operationApiVersion api-version detail in both query and path
 * @param urlVersionDetail api-version detail in parameterized host
 * @returns calculated combined position info
 */
function extractDefinedPosition(operationApiVersion, urlVersionDetail) {
    let pos = "none";
    if (operationApiVersion && urlVersionDetail) {
        pos = "duplicate";
    }
    else if (operationApiVersion === null || operationApiVersion === void 0 ? void 0 : operationApiVersion.definedPosition) {
        pos = operationApiVersion.definedPosition;
    }
    else if (urlVersionDetail) {
        pos = "baseurl";
    }
    return pos;
}
