// @azure-tools/rlc-common/dist/helpers/importsUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRuntimeImports = buildRuntimeImports;
exports.initInternalImports = initInternalImports;
exports.getImportSpecifier = getImportSpecifier;
exports.addImportToSpecifier = addImportToSpecifier;
exports.clearImportSets = clearImportSets;
exports.addImportsToFiles = addImportsToFiles;
/**
 * Build the common imports for generated SDK
 * @param flavor flavor of SDK to generate, if any. When set to "azure", Azure Core packages will be used. When unset, the generic `ts-http-runtime` package will be used.
 * @returns
 */
function buildRuntimeImports(flavor) {
    if (flavor === "azure") {
        return {
            restClient: {
                type: "restClient",
                specifier: "@azure-rest/core-client",
                version: "^2.0.0"
            },
            coreAuth: {
                type: "coreAuth",
                specifier: "@azure/core-auth",
                version: "^1.6.0"
            },
            restPipeline: {
                type: "restPipeline",
                specifier: "@azure/core-rest-pipeline",
                version: "^1.14.0"
            },
            coreUtil: {
                type: "coreUtil",
                specifier: "@azure/core-util",
                version: "^1.4.0"
            },
            coreLogger: {
                type: "coreLogger",
                specifier: "@azure/logger",
                version: "^1.0.4"
            },
            azureEslintPlugin: {
                type: "azureEslintPlugin",
                specifier: "@azure/eslint-plugin-azure-sdk",
                version: "^3.0.0"
            },
            azureTestRecorder: {
                type: "azureTestRecorder",
                specifier: "@azure-tools/test-recorder",
                version: "^3.0.0"
            },
            azureCoreLro: {
                type: "azureCoreLro",
                specifier: "@azure/core-lro"
            }
        };
    }
    else {
        // In non-azure branded scope we only have one dependency that is ts-http-runtime
        return {
            commonFallback: {
                type: "commonFallback",
                specifier: "@typespec/ts-http-runtime",
                version: "0.1.0"
            }
        };
    }
}
/**
 * Initialize the inner imports for parameter and response, the import set would be used for referred models
 * @returns
 */
function initInternalImports() {
    return {
        parameter: {
            type: "parameter",
            importsSet: new Set()
        },
        response: {
            type: "response",
            importsSet: new Set()
        },
        rlcIndex: {
            type: "rlcIndex",
            importsSet: new Set()
        },
        modularModel: {
            type: "modularModel",
            importsSet: new Set()
        },
        rlcClientFactory: {
            type: "rlcClientFactory",
            importsSet: new Set()
        },
        rlcClientDefinition: {
            type: "rlcClientDefinition",
            importsSet: new Set()
        },
        serializerHelpers: {
            type: "serializerHelpers",
            importsSet: new Set()
        }
    };
}
function getImportSpecifier(importType, imports, includeFallback = true) {
    var _a, _b, _c, _d, _e, _f;
    imports = imports !== null && imports !== void 0 ? imports : {};
    const defaultPackageMap = {
        restClient: "@azure-rest/core-client",
        coreAuth: "@azure/core-auth",
        restPipeline: "@azure/core-rest-pipeline",
        coreUtil: "@azure/core-util",
        coreLogger: "@azure/logger",
        azureCoreLro: "@azure/core-lro"
    };
    if (!includeFallback) {
        return (_b = (_a = imports[importType]) === null || _a === void 0 ? void 0 : _a.specifier) !== null && _b !== void 0 ? _b : "";
    }
    return ((_f = (_e = (_d = ((_c = imports[importType]) !== null && _c !== void 0 ? _c : imports.commonFallback)) === null || _d === void 0 ? void 0 : _d.specifier) !== null && _e !== void 0 ? _e : defaultPackageMap[importType]) !== null && _f !== void 0 ? _f : "");
}
function addImportToSpecifier(importType, runtimeImports, importedName) {
    var _a;
    const specifier = getImportSpecifier(importType, runtimeImports);
    const importSet = (_a = runtimeImports[importType]) === null || _a === void 0 ? void 0 : _a.importsSet;
    if (!importSet) {
        runtimeImports[importType] = {
            type: importType,
            specifier,
            importsSet: new Set().add(importedName)
        };
    }
    else {
        importSet.add(importedName);
    }
}
function clearImportSets(runtimeImports) {
    var _a;
    for (const importType of Object.values(runtimeImports)) {
        (_a = importType.importsSet) === null || _a === void 0 ? void 0 : _a.clear();
    }
}
function addImportsToFiles(runtimeImports, file, internalSpecifierMap) {
    Object.values(runtimeImports)
        .filter((importType) => {
        var _a;
        return (_a = importType.importsSet) === null || _a === void 0 ? void 0 : _a.size;
    })
        .forEach((importType) => {
        var _a;
        const specifier = (_a = internalSpecifierMap === null || internalSpecifierMap === void 0 ? void 0 : internalSpecifierMap[importType.type]) !== null && _a !== void 0 ? _a : importType.specifier;
        let hasModifier = false;
        if (!specifier) {
            return;
        }
        file
            .getImportDeclarations()
            .filter((importDeclaration) => {
            return importDeclaration.getModuleSpecifierValue() === specifier;
        })
            .forEach((importDeclaration) => {
            hasModifier = true;
            importDeclaration.addNamedImports([
                ...importType.importsSet.values()
            ]);
        });
        if (!hasModifier) {
            return file.addImportDeclaration({
                isTypeOnly: true,
                moduleSpecifier: specifier,
                namedImports: [...importType.importsSet.values()]
            });
        }
    });
    clearImportSets(runtimeImports);
}
