// @azure-tools/rlc-common/dist/helpers/nameConstructors.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResponseTypeName = getResponseTypeName;
exports.getLroLogicalResponseName = getLroLogicalResponseName;
exports.getResponseBaseName = getResponseBaseName;
exports.getParameterBaseName = getParameterBaseName;
exports.getParameterTypeName = getParameterTypeName;
exports.getImportModuleName = getImportModuleName;
exports.getClientName = getClientName;
exports.getMultipartPartTypeName = getMultipartPartTypeName;
const nameUtils_js_1 = require("./nameUtils.js");
function getResponseTypeName(baseNameOrOperationGroup, operationName, statusCode) {
    if (operationName) {
        baseNameOrOperationGroup = getResponseBaseName(baseNameOrOperationGroup, operationName, statusCode || "");
    }
    return (0, nameUtils_js_1.normalizeName)(`${baseNameOrOperationGroup}Response`, nameUtils_js_1.NameType.Interface);
}
function getLroLogicalResponseName(operationGroup, operationName) {
    return (0, nameUtils_js_1.normalizeName)(`${operationGroup}_${operationName}_Logical_Response`, nameUtils_js_1.NameType.Interface);
}
/**
 * The prefix of all response types
 * @param operationGroup operation group name e.g string_PutEmpty
 * @param operationName operation name D e.g string_PutEmpty
 * @param statusCode 2XX, 4XX, 5XX, default etc.
 * @returns normolized base name e.g StringPutEmpty200
 */
function getResponseBaseName(operationGroup, operationName, statusCode) {
    return (0, nameUtils_js_1.normalizeName)(`${operationGroup}_${(0, nameUtils_js_1.normalizeName)(operationName, nameUtils_js_1.NameType.Interface, true)}_${statusCode}`, nameUtils_js_1.NameType.Interface);
}
/**
 * The prefix of all parameter relevant types
 * @param operationName is composed with operationGroup and operationID e.g string_PutEmpty
 * @returns
 */
function getParameterBaseName(operationGroup, operationName) {
    return (0, nameUtils_js_1.normalizeName)(`${operationGroup}_${operationName}`, nameUtils_js_1.NameType.Interface);
}
function getParameterTypeName(baseNameOrOperationGroup, operationName) {
    if (operationName) {
        baseNameOrOperationGroup = getParameterBaseName(baseNameOrOperationGroup, operationName);
    }
    return (0, nameUtils_js_1.normalizeName)(`${baseNameOrOperationGroup}_Parameters`, nameUtils_js_1.NameType.Interface);
}
/**
 * This is a helper function that gets the right import module depending on the type of
 * library being generated
 */
function getImportModuleName(name, codeModel) {
    var _a;
    if (((_a = codeModel.options) === null || _a === void 0 ? void 0 : _a.moduleKind) === "cjs") {
        return name.cjsName;
    }
    return name.esModulesName;
}
function getClientName(model) {
    var _a;
    const clientName = model.libraryName;
    const clientInterfaceName = ((_a = model.options) === null || _a === void 0 ? void 0 : _a.isModularLibrary)
        ? model.libraryName
        : clientName.endsWith("Client")
            ? `${clientName}`
            : `${clientName}Client`;
    return clientInterfaceName;
}
function getMultipartPartTypeName(schemaName, partName) {
    const name = (0, nameUtils_js_1.normalizeName)(partName, nameUtils_js_1.NameType.Interface);
    const bodyParamName = (0, nameUtils_js_1.normalizeName)(schemaName, nameUtils_js_1.NameType.Interface);
    return (0, nameUtils_js_1.normalizeName)(`${bodyParamName}_${name}_PartDescriptor`, nameUtils_js_1.NameType.Interface);
}
