// @azure-tools/rlc-common/dist/helpers/operationHelpers.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildMethodDefinitions = buildMethodDefinitions;
exports.getPathParamDefinitions = getPathParamDefinitions;
exports.hasPagingOperations = hasPagingOperations;
exports.hasPollingOperations = hasPollingOperations;
exports.hasMultiCollection = hasMultiCollection;
exports.hasPipeCollection = hasPipeCollection;
exports.hasSsvCollection = hasSsvCollection;
exports.hasTsvCollection = hasTsvCollection;
exports.hasCsvCollection = hasCsvCollection;
exports.hasUnexpectedHelper = hasUnexpectedHelper;
exports.hasInputModels = hasInputModels;
exports.hasOutputModels = hasOutputModels;
exports.getGeneratedWrapperTypes = getGeneratedWrapperTypes;
const interfaces_js_1 = require("../interfaces.js");
const nameUtils_js_1 = require("./nameUtils.js");
const schemaHelpers_js_1 = require("./schemaHelpers.js");
function buildMethodDefinitions(methods, pathParams = []) {
    const methodDefinitions = [];
    for (const key of Object.keys(methods)) {
        const verbMethods = methods[key];
        for (const method of verbMethods) {
            const description = method.description;
            const areAllOptional = method.hasOptionalOptions;
            methodDefinitions.push({
                name: key,
                ...(description && { docs: [{ description }] }),
                parameters: [
                    ...getPathParamDefinitions(pathParams),
                    {
                        name: "options",
                        hasQuestionToken: areAllOptional,
                        type: (0, nameUtils_js_1.pascalCase)(method.optionsName)
                    }
                ],
                returnType: `StreamableMethod<${method.returnType}>`
            });
        }
    }
    return methodDefinitions;
}
function getPathParamDefinitions(pathParams) {
    return pathParams.map(({ name, type, description }) => {
        return {
            name: (0, nameUtils_js_1.normalizeName)(name, nameUtils_js_1.NameType.Parameter),
            type,
            description
        };
    });
}
function hasPagingOperations(model) {
    var _a;
    return Boolean((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.hasPaging);
}
function hasPollingOperations(model) {
    var _a;
    return Boolean((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.hasLongRunning);
}
function hasMultiCollection(model) {
    var _a;
    return Boolean((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.hasMultiCollection);
}
function hasPipeCollection(model) {
    var _a;
    return Boolean((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.hasPipeCollection);
}
function hasSsvCollection(model) {
    var _a;
    return Boolean((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.hasSsvCollection);
}
function hasTsvCollection(model) {
    var _a;
    return Boolean((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.hasTsvCollection);
}
function hasCsvCollection(model) {
    var _a;
    return Boolean((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.hasCsvCollection);
}
function hasUnexpectedHelper(model) {
    const pathDictionary = model.paths;
    for (const details of Object.values(pathDictionary)) {
        for (const methodDetails of Object.values(details.methods)) {
            const successTypes = methodDetails[0].responseTypes.success;
            const errorTypes = methodDetails[0].responseTypes.error;
            if (successTypes.length > 0 && errorTypes.length > 0 && !!errorTypes[0]) {
                return true;
            }
        }
    }
    return false;
}
function hasInputModels(model) {
    return hasSchemaContextObject(model, [interfaces_js_1.SchemaContext.Input]);
}
function hasOutputModels(model) {
    return hasSchemaContextObject(model, [
        interfaces_js_1.SchemaContext.Output,
        interfaces_js_1.SchemaContext.Exception
    ]);
}
function hasSchemaContextObject(model, schemaUsage) {
    var _a;
    const objectSchemas = ((_a = model.schemas) !== null && _a !== void 0 ? _a : []).filter((o) => {
        var _a;
        return (0, schemaHelpers_js_1.isObjectSchema)(o) &&
            ((_a = o.usage) === null || _a === void 0 ? void 0 : _a.some((u) => schemaUsage.includes(u)));
    });
    return objectSchemas.length > 0;
}
function getGeneratedWrapperTypes(params) {
    const wrapperTypes = params
        .map((qp) => isParameterMetadata(qp) ? qp.param.wrapperType : qp.wrapperType)
        .filter((v) => v !== undefined);
    const wrapperFromObjects = wrapperTypes.filter((wrap) => wrap.type === "object");
    const wrapperFromUnions = wrapperTypes
        .filter((wrap) => wrap.type === "union")
        .flatMap((wrapperType) => { var _a; return (_a = wrapperType === null || wrapperType === void 0 ? void 0 : wrapperType.enum) !== null && _a !== void 0 ? _a : []; })
        .filter((v) => v.type === "object");
    return [...wrapperFromUnions, ...wrapperFromObjects];
}
function isParameterMetadata(param) {
    return param.param !== undefined;
}
