// @azure-tools/rlc-common/dist/helpers/shortcutMethods.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateMethodShortcuts = generateMethodShortcuts;
const operationHelpers_js_1 = require("./operationHelpers.js");
const nameUtils_js_1 = require("./nameUtils.js");
function generateMethodShortcuts(paths) {
    const keys = {};
    for (const path in paths) {
        const groupName = paths[path].operationGroupName;
        const definitions = buildOperationDefinitions(paths[path]);
        if (!keys[groupName]) {
            keys[groupName] = definitions;
        }
        else {
            keys[groupName] = [...keys[groupName], ...definitions];
        }
    }
    const interfaces = [];
    for (const interfaceName in keys) {
        const methods = keys[interfaceName];
        interfaces.push({
            name: `${interfaceName}Operations`,
            methods: methods,
            isExported: true,
            docs: [`Contains operations for ${interfaceName} operations`]
        });
    }
    return interfaces;
}
function buildOperationDefinitions(path) {
    let ops = [];
    for (const verb in path.methods) {
        const methods = path.methods[verb];
        for (const method of methods) {
            const name = (0, nameUtils_js_1.normalizeName)(method.operationName, nameUtils_js_1.NameType.Property);
            const pathParams = path.pathParameters;
            const methodDefinitions = (0, operationHelpers_js_1.buildMethodDefinitions)({ [name]: [method] }, pathParams);
            ops = [...ops, ...methodDefinitions];
        }
    }
    return ops;
}
