// @azure-tools/rlc-common/dist/helpers/typeUtil.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeScriptType = void 0;
exports.isRecord = isRecord;
exports.getRecordType = getRecordType;
exports.isArray = isArray;
exports.isArrayObject = isArrayObject;
exports.getArrayObjectType = getArrayObjectType;
exports.isNativeArray = isNativeArray;
exports.getNativeArrayType = getNativeArrayType;
exports.isUnion = isUnion;
exports.getUnionType = getUnionType;
exports.leaveBracket = leaveBracket;
exports.leaveStringQuotes = leaveStringQuotes;
exports.toTypeScriptTypeFromSchema = toTypeScriptTypeFromSchema;
exports.toTypeScriptTypeFromName = toTypeScriptTypeFromName;
exports.isConstant = isConstant;
exports.isNumericLiteral = isNumericLiteral;
exports.isBoolLiteral = isBoolLiteral;
exports.isStringLiteral = isStringLiteral;
function isRecord(type) {
    return /^Record<([a-zA-Z]+),(\s*)(?<type>.+)>$/.test(type);
}
function getRecordType(type) {
    var _a, _b;
    return (_b = (_a = /^Record<([a-zA-Z]+),(\s*)(?<type>.+)>$/.exec(type)) === null || _a === void 0 ? void 0 : _a.groups) === null || _b === void 0 ? void 0 : _b.type;
}
function isArray(type) {
    return isArrayObject(type) || isNativeArray(type);
}
function isArrayObject(type) {
    return /(^Array<(.+)>$)/g.test(type);
}
function getArrayObjectType(type) {
    var _a, _b;
    return (_b = (_a = /^Array<(?<type>.+)>$/g.exec(type)) === null || _a === void 0 ? void 0 : _a.groups) === null || _b === void 0 ? void 0 : _b.type;
}
function isNativeArray(type) {
    return /(^.+)\[\]$/g.test(type);
}
function getNativeArrayType(type) {
    var _a, _b;
    return (_b = (_a = /(?<type>.+)\[\]/g.exec(type)) === null || _a === void 0 ? void 0 : _a.groups) === null || _b === void 0 ? void 0 : _b.type;
}
function isUnion(type) {
    const members = type.split("|").map((m) => m.trim());
    return members.length > 1 && !isStringLiteral(type) && !isRecord(type);
}
function getUnionType(type) {
    return leaveBracket(type.split("|").map((m) => m.trim())[0]);
}
function leaveBracket(type) {
    if (!type || type.length < 2) {
        return type;
    }
    else if (type.startsWith("(") && type.endsWith(")")) {
        return type.slice(1, type.length - 1);
    }
    return type;
}
function leaveStringQuotes(str) {
    if (isStringLiteral(str)) {
        return str.slice(1, str.length - 1);
    }
    return str;
}
var TypeScriptType;
(function (TypeScriptType) {
    TypeScriptType[TypeScriptType["string"] = 0] = "string";
    TypeScriptType[TypeScriptType["date"] = 1] = "date";
    TypeScriptType[TypeScriptType["number"] = 2] = "number";
    TypeScriptType[TypeScriptType["boolean"] = 3] = "boolean";
    TypeScriptType[TypeScriptType["constant"] = 4] = "constant";
    TypeScriptType[TypeScriptType["record"] = 5] = "record";
    TypeScriptType[TypeScriptType["array"] = 6] = "array";
    TypeScriptType[TypeScriptType["object"] = 7] = "object";
    TypeScriptType[TypeScriptType["union"] = 8] = "union";
    TypeScriptType[TypeScriptType["unknown"] = 9] = "unknown";
    TypeScriptType[TypeScriptType["enum"] = 10] = "enum";
})(TypeScriptType || (exports.TypeScriptType = TypeScriptType = {}));
function toTypeScriptTypeFromSchema(schema) {
    var _a;
    schema.type = schema.type.trim();
    if (schema.type === "string" &&
        schema.typeName === "Date | string" &&
        schema.outputTypeName === "string") {
        return TypeScriptType.date;
    }
    else if (schema.enum && schema.enum.length > 0) {
        if (schema.type === "union" || schema.type === "object") {
            // Include both union and named union
            return TypeScriptType.union;
        }
        else {
            // Include both extensible and fixed enum
            return TypeScriptType.enum;
        }
    }
    else if (schema.isConstant === true ||
        isConstant((_a = schema.typeName) !== null && _a !== void 0 ? _a : schema.type)) {
        return TypeScriptType.constant;
    }
    else if (schema.type === "number") {
        return TypeScriptType.number;
    }
    else if (schema.type === "boolean") {
        return TypeScriptType.boolean;
    }
    else if (schema.type === "string") {
        return TypeScriptType.string;
    }
    else if (schema.type === "unknown") {
        return TypeScriptType.unknown;
    }
    else if (schema.type === "dictionary") {
        return TypeScriptType.record;
    }
    else if (schema.type === "array") {
        return TypeScriptType.array;
    }
    else if (schema.type === "object") {
        return TypeScriptType.object;
    }
}
function toTypeScriptTypeFromName(typeName) {
    typeName = typeName.trim();
    if (typeName === "Date") {
        return TypeScriptType.date;
    }
    else if (typeName === "string") {
        return TypeScriptType.string;
    }
    else if (typeName === "number") {
        return TypeScriptType.number;
    }
    else if (typeName === "boolean") {
        return TypeScriptType.boolean;
    }
    else if (typeName === "unknown") {
        return TypeScriptType.unknown;
    }
    else if (isConstant(typeName)) {
        return TypeScriptType.constant;
    }
    else if (isRecord(typeName)) {
        return TypeScriptType.record;
    }
    else if (isArray(typeName)) {
        return TypeScriptType.array;
    }
    else if (isUnion(typeName)) {
        return TypeScriptType.union;
    }
}
function isConstant(typeName) {
    return (["never", "null"].includes(typeName) ||
        isBoolLiteral(typeName) ||
        isNumericLiteral(typeName) ||
        isStringLiteral(typeName));
}
function isNumericLiteral(str) {
    if (typeof str !== "string")
        return false;
    return !isNaN(Number(str)) && !isNaN(parseFloat(str));
}
function isBoolLiteral(str) {
    return str === "true" || str === "false";
}
function isStringLiteral(type) {
    if (type.length < 2) {
        return false;
    }
    const first = type[0];
    const lastPos = type.length - 1;
    return (first === type[lastPos] &&
        (first === '"' || first === "'") &&
        type.indexOf(first, 1) === lastPos);
}
