// @azure-tools/rlc-common/dist/helpers/valueGenerationUtil.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateParameterTypeValue = generateParameterTypeValue;
const buildObjectTypes_js_1 = require("../buildObjectTypes.js");
const interfaces_js_1 = require("../interfaces.js");
const typeUtil_js_1 = require("./typeUtil.js");
/**
 * Generate parameter type value for the given type and parameter name
 * @param type the typescript type
 * @param parameterName the parameter name
 * @param schemaMap the schema info to help generate the value
 * @param path optional path to help detect self reference
 * @param _allowMockValue the flag to indicate whether to allow mock value, currently we always generate mock value
 * @returns
 */
function generateParameterTypeValue(type, parameterName, schemaMap, path = new Set(), _allowMockValue = true) {
    type = (0, typeUtil_js_1.leaveBracket)(type === null || type === void 0 ? void 0 : type.trim());
    let tsType;
    // Give priority to suggest the ts-type from schema
    if (schemaMap.has(type)) {
        tsType = (0, typeUtil_js_1.toTypeScriptTypeFromSchema)(schemaMap.get(type));
    }
    // Fallback to suggest ts-type from the type iteself
    if (!tsType) {
        tsType = (0, typeUtil_js_1.toTypeScriptTypeFromName)(type);
    }
    switch (tsType) {
        case typeUtil_js_1.TypeScriptType.string:
            return `"{Your ${(0, typeUtil_js_1.leaveStringQuotes)(parameterName)}}"`;
        case typeUtil_js_1.TypeScriptType.number:
            return "123";
        case typeUtil_js_1.TypeScriptType.boolean:
            return "true";
        case typeUtil_js_1.TypeScriptType.date:
            return "new Date()";
        case typeUtil_js_1.TypeScriptType.unknown:
            return `"Unknown Type"`;
        case typeUtil_js_1.TypeScriptType.object: {
            return generateObjectValues(type, parameterName, schemaMap, path);
        }
        case typeUtil_js_1.TypeScriptType.array: {
            return generateArrayValues(type, parameterName, schemaMap, path);
        }
        case typeUtil_js_1.TypeScriptType.record: {
            return generateRecordValues(type, parameterName, schemaMap, path);
        }
        case typeUtil_js_1.TypeScriptType.enum: {
            return mockEnumValues(type, parameterName, schemaMap, path);
        }
        case typeUtil_js_1.TypeScriptType.union: {
            return mockUnionValues(type, parameterName, schemaMap, path);
        }
        case typeUtil_js_1.TypeScriptType.constant:
            return type;
    }
    return `undefined /**FIXME */`;
}
function mockEnumValues(type, parameterName, schemaMap, path = new Set()) {
    const schema = schemaMap.get(type);
    if (schema && schema.enum && schema.enum.length > 0) {
        return schema.enum[0].type;
    }
    return generateParameterTypeValue((0, typeUtil_js_1.getUnionType)(type), parameterName, schemaMap, path);
}
function mockUnionValues(type, parameterName, schemaMap, path = new Set()) {
    var _a;
    const schema = schemaMap.get(type);
    if (schema && schema.enum && schema.enum.length > 0) {
        addToSchemaMap(schemaMap, schema.enum[0]);
        return generateParameterTypeValue((_a = getAccurateTypeName(schema.enum[0])) !== null && _a !== void 0 ? _a : schema.enum[0], parameterName, schemaMap, path);
    }
    return generateParameterTypeValue((0, typeUtil_js_1.getUnionType)(type), parameterName, schemaMap, path);
}
function generateRecordValues(type, parameterName, schemaMap, path = new Set()) {
    let recordType = (0, typeUtil_js_1.getRecordType)(type);
    const schema = schemaMap.get(type);
    if (schema && schema.additionalProperties) {
        recordType = getAccurateTypeName(schema.additionalProperties);
        addToSchemaMap(schemaMap, schema.additionalProperties);
    }
    return recordType
        ? `{"key": ${generateParameterTypeValue(recordType, parameterName, schemaMap, path)}}`
        : `{}`;
}
function generateArrayValues(type, parameterName, schemaMap, path = new Set()) {
    let arrayType;
    const schema = schemaMap.get(type);
    if (schema && schema.items) {
        arrayType = getAccurateTypeName(schema.items);
        addToSchemaMap(schemaMap, schema.items);
    }
    else if ((0, typeUtil_js_1.isArrayObject)(type)) {
        arrayType = (0, typeUtil_js_1.getArrayObjectType)(type);
    }
    else if ((0, typeUtil_js_1.isNativeArray)(type)) {
        arrayType = (0, typeUtil_js_1.getNativeArrayType)(type);
    }
    const itemValue = arrayType
        ? generateParameterTypeValue(arrayType, parameterName, schemaMap, path)
        : undefined;
    return itemValue ? `[${itemValue}]` : `[]`;
}
function generateObjectValues(type, _parameterName, schemaMap, path = new Set()) {
    if (path.has(type)) {
        // skip generating if self referenced
        return `{} as any /**FIXME */`;
    }
    path.add(type);
    // Extract all properties from the schema
    const allProperties = getAllProperties(schemaMap.get(type), schemaMap);
    const values = extractObjectProperties(allProperties, schemaMap, path);
    path.delete(type);
    return `{${values.join(", ")}}`;
}
function getAllProperties(schema, schemaMap = new Map()) {
    var _a;
    const propertiesMap = new Map();
    if (!schema) {
        return new Map();
    }
    (_a = (0, buildObjectTypes_js_1.getImmediateParentsNames)(schema, [interfaces_js_1.SchemaContext.Input])) === null || _a === void 0 ? void 0 : _a.forEach((p) => {
        const parentProperties = getAllProperties(schemaMap.get(p), schemaMap);
        for (const prop of parentProperties.keys()) {
            propertiesMap.set(prop, parentProperties.get(prop));
        }
    });
    for (const prop in schema.properties) {
        propertiesMap.set(prop, schema.properties[prop]);
    }
    return propertiesMap;
}
function extractObjectProperties(properties, schemaMap = new Map(), path = new Set()) {
    const values = [];
    for (const name of properties.keys()) {
        const property = properties.get(name);
        if (!property || property.readOnly || property.type === "never") {
            continue;
        }
        addToSchemaMap(schemaMap, property);
        values.push(`${name}: ` +
            generateParameterTypeValue(getAccurateTypeName(property), name, schemaMap, path));
    }
    return values;
}
function getAccurateTypeName(schema) {
    var _a;
    // For extensible enum, fallback to use the type
    if (schema.typeName === "string" && schema.enum && schema.enum.length > 0) {
        return schema.type;
    }
    return (_a = schema.typeName) !== null && _a !== void 0 ? _a : schema.type;
}
function addToSchemaMap(schemaMap, schema) {
    const type = getAccurateTypeName(schema);
    if (!type) {
        return;
    }
    if (!schemaMap.has(type) &&
        !["string", "number", "boolean"].includes(schema.type)) {
        schemaMap.set(type, schema);
    }
}
