// @azure-tools/rlc-common/dist/buildIndexFile.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildIndexFile = buildIndexFile;
const ts_morph_1 = require("ts-morph");
const nameUtils_js_1 = require("./helpers/nameUtils.js");
const operationHelpers_js_1 = require("./helpers/operationHelpers.js");
const path = require("path");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
function buildIndexFile(model) {
    var _a, _b;
    const multiClient = Boolean((_a = model.options) === null || _a === void 0 ? void 0 : _a.multiClient), batch = (_b = model.options) === null || _b === void 0 ? void 0 : _b.batch;
    const project = new ts_morph_1.Project();
    const { srcPath } = model;
    const filePath = path.join(srcPath, `index.ts`);
    const indexFile = project.createSourceFile(filePath, undefined, {
        overwrite: true
    });
    if (!multiClient || !batch || (batch === null || batch === void 0 ? void 0 : batch.length) === 1) {
        // if we are generate single client package for RLC
        generateRLCIndex(indexFile, model);
    }
    else {
        generateRLCIndexForMultiClient(indexFile, model);
    }
    return {
        path: filePath,
        content: indexFile.getFullText()
    };
}
// to generate a index.ts for each single module inside the multi client RLC package
function generateRLCIndexForMultiClient(file, model) {
    const clientName = model.libraryName;
    const createClientFuncName = `createClient`;
    const moduleName = (0, nameUtils_js_1.normalizeName)(clientName, nameUtils_js_1.NameType.File);
    file.addImportDeclaration({
        namespaceImport: "Parameters",
        moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({ cjsName: "./parameters", esModulesName: "./parameters.js" }, model)
    });
    file.addImportDeclaration({
        namespaceImport: "Responses",
        moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({ cjsName: "./responses", esModulesName: "./responses.js" }, model)
    });
    file.addImportDeclaration({
        namespaceImport: "Client",
        moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
            cjsName: "./clientDefinitions",
            esModulesName: "./clientDefinitions.js"
        }, model)
    });
    const exports = ["Parameters", "Responses", "Client"];
    if ((0, operationHelpers_js_1.hasInputModels)(model)) {
        file.addImportDeclaration({
            namespaceImport: "Models",
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: "./models",
                esModulesName: "./models.js"
            }, model)
        });
        exports.push("Models");
    }
    if ((0, operationHelpers_js_1.hasOutputModels)(model)) {
        file.addImportDeclaration({
            namespaceImport: "OutputModels",
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: "./outputModels",
                esModulesName: "./outputModels.js"
            }, model)
        });
        exports.push("OutputModels");
    }
    if ((0, operationHelpers_js_1.hasPagingOperations)(model)) {
        file.addImportDeclaration({
            namespaceImport: "PaginateHelper",
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: "./paginateHelper",
                esModulesName: "./paginateHelper.js"
            }, model)
        });
        exports.push("PaginateHelper");
    }
    if ((0, operationHelpers_js_1.hasUnexpectedHelper)(model)) {
        file.addImportDeclaration({
            namespaceImport: "UnexpectedHelper",
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: "./isUnexpected",
                esModulesName: "./isUnexpected.js"
            }, model)
        });
        exports.push("UnexpectedHelper");
    }
    if ((0, operationHelpers_js_1.hasPollingOperations)(model)) {
        file.addImportDeclaration({
            namespaceImport: "PollingHelper",
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: "./pollingHelper",
                esModulesName: "./pollingHelper.js"
            }, model)
        });
        exports.push("PollingHelper");
    }
    if ((0, operationHelpers_js_1.hasMultiCollection)(model) ||
        (0, operationHelpers_js_1.hasSsvCollection)(model) ||
        (0, operationHelpers_js_1.hasPipeCollection)(model) ||
        (0, operationHelpers_js_1.hasTsvCollection)(model) ||
        (0, operationHelpers_js_1.hasCsvCollection)(model)) {
        file.addImportDeclaration({
            namespaceImport: "SerializeHelper",
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: "./serializeHelper",
                esModulesName: "./serializeHelper.js"
            }, model)
        });
        exports.push("SerializeHelper");
    }
    file.addExportDeclarations([
        {
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: `./${moduleName}`,
                esModulesName: `./${moduleName}.js`
            }, model),
            namedExports: [`${createClientFuncName}`, `${clientName}ClientOptions`]
        },
        {
            namedExports: [...exports]
        }
    ]);
}
function generateRLCIndex(file, model) {
    const clientName = model.libraryName;
    const createClientFuncName = `${clientName}`;
    const moduleName = (0, nameUtils_js_1.normalizeName)(clientName, nameUtils_js_1.NameType.File);
    file.addImportDeclaration({
        moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
            cjsName: `./${moduleName}`,
            esModulesName: `./${moduleName}.js`
        }, model),
        defaultImport: createClientFuncName
    });
    file.addExportDeclarations([
        {
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: `./${moduleName}`,
                esModulesName: `./${moduleName}.js`
            }, model)
        },
        {
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: `./parameters`,
                esModulesName: `./parameters.js`
            }, model)
        },
        {
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: `./responses`,
                esModulesName: `./responses.js`
            }, model)
        },
        {
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: `./clientDefinitions`,
                esModulesName: `./clientDefinitions.js`
            }, model)
        }
    ]);
    if ((0, operationHelpers_js_1.hasUnexpectedHelper)(model)) {
        file.addExportDeclarations([
            {
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./isUnexpected`,
                    esModulesName: `./isUnexpected.js`
                }, model)
            }
        ]);
    }
    if ((0, operationHelpers_js_1.hasInputModels)(model)) {
        file.addExportDeclarations([
            {
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./models`,
                    esModulesName: `./models.js`
                }, model)
            }
        ]);
    }
    if ((0, operationHelpers_js_1.hasOutputModels)(model)) {
        file.addExportDeclarations([
            {
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./outputModels`,
                    esModulesName: `./outputModels.js`
                }, model)
            }
        ]);
    }
    if ((0, operationHelpers_js_1.hasPagingOperations)(model)) {
        file.addExportDeclarations([
            {
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./paginateHelper`,
                    esModulesName: `./paginateHelper.js`
                }, model)
            }
        ]);
    }
    if ((0, operationHelpers_js_1.hasPollingOperations)(model)) {
        file.addExportDeclarations([
            {
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./pollingHelper`,
                    esModulesName: `./pollingHelper.js`
                }, model)
            }
        ]);
    }
    if ((0, operationHelpers_js_1.hasMultiCollection)(model) ||
        (0, operationHelpers_js_1.hasSsvCollection)(model) ||
        (0, operationHelpers_js_1.hasPipeCollection)(model) ||
        (0, operationHelpers_js_1.hasTsvCollection)(model) ||
        (0, operationHelpers_js_1.hasCsvCollection)(model)) {
        file.addExportDeclarations([
            {
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./serializeHelper`,
                    esModulesName: `./serializeHelper.js`
                }, model)
            }
        ]);
    }
    file.addExportAssignment({
        expression: createClientFuncName,
        isExportEquals: false
    });
}
