// @azure-tools/rlc-common/dist/buildObjectTypes.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildObjectInterfaces = buildObjectInterfaces;
exports.buildObjectAliases = buildObjectAliases;
exports.buildPolymorphicAliases = buildPolymorphicAliases;
exports.getObjectInterfaceDeclaration = getObjectInterfaceDeclaration;
exports.getImmediateParentsNames = getImmediateParentsNames;
exports.getPropertySignature = getPropertySignature;
const ts_morph_1 = require("ts-morph");
const nameUtils_js_1 = require("./helpers/nameUtils.js");
const schemaHelpers_js_1 = require("./helpers/schemaHelpers.js");
const interfaces_js_1 = require("./interfaces.js");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
/**
 * Generates interfaces for ObjectSchemas
 */
function buildObjectInterfaces(model, importedModels, schemaUsage) {
    var _a, _b;
    const objectSchemas = ((_a = model.schemas) !== null && _a !== void 0 ? _a : []).filter((o) => {
        var _a;
        return (0, schemaHelpers_js_1.isObjectSchema)(o) &&
            ((_a = o.usage) === null || _a === void 0 ? void 0 : _a.some((u) => schemaUsage.includes(u)));
    });
    const objectInterfaces = [];
    for (const objectSchema of objectSchemas) {
        if (objectSchema.alias ||
            objectSchema.outputAlias ||
            objectSchema.fromCore) {
            continue;
        }
        // FIXME: disabling new multipart generation for modular while we figure out the story
        if (objectSchema.isMultipartBody && !((_b = model.options) === null || _b === void 0 ? void 0 : _b.isModularLibrary)) {
            objectInterfaces.push(...buildMultipartPartDefinitions(objectSchema, importedModels, schemaUsage));
            continue;
        }
        const baseName = getObjectBaseName(objectSchema, schemaUsage);
        const interfaceDeclaration = getObjectInterfaceDeclaration(model, baseName, objectSchema, schemaUsage, importedModels);
        objectInterfaces.push(interfaceDeclaration);
    }
    return objectInterfaces;
}
const MULTIPART_FILE_METADATA_PROPERTIES = [
    {
        name: "filename",
        hasQuestionToken: true,
        type: "string"
    },
    {
        name: "contentType",
        hasQuestionToken: true,
        type: "string"
    }
];
function buildMultipartPartDefinitions(schema, importedModels, schemaUsage) {
    var _a, _b, _c, _d;
    if (!schema.isMultipartBody) {
        return [];
    }
    // Transform property signatures into individual models
    const propertySignatures = getPropertySignatures((_a = schema.properties) !== null && _a !== void 0 ? _a : {}, schemaUsage, importedModels, { flattenBinaryArrays: true });
    const structures = [];
    for (const signature of propertySignatures) {
        const name = signature.name;
        const propertySchema = (_b = schema.properties) === null || _b === void 0 ? void 0 : _b[name];
        const typeName = (0, nameConstructors_js_1.getMultipartPartTypeName)(schema.name, name);
        const isFileUpload = (_d = (_c = signature.type) === null || _c === void 0 ? void 0 : _c.toString().includes("File")) !== null && _d !== void 0 ? _d : false;
        const additionalProperties = [];
        if (propertySchema.multipartOptions) {
            const multipartOptions = propertySchema.multipartOptions;
            if (multipartOptions.filenameSchema) {
                additionalProperties.push(getPropertySignature({ name: "filename", ...multipartOptions.filenameSchema }, [interfaces_js_1.SchemaContext.Input], importedModels));
            }
            if (multipartOptions.contentTypeSchema) {
                additionalProperties.push(getPropertySignature({ name: "contentType", ...multipartOptions.contentTypeSchema }, [interfaces_js_1.SchemaContext.Input], importedModels));
            }
        }
        else if (isFileUpload) {
            // default additional file metadata properties (legacy)
            additionalProperties.push(...MULTIPART_FILE_METADATA_PROPERTIES);
        }
        structures.push({
            kind: ts_morph_1.StructureKind.Interface,
            isExported: true,
            name: typeName,
            properties: [
                {
                    name: "name",
                    type: name
                },
                {
                    name: "body",
                    type: signature.type
                },
                ...additionalProperties
            ]
        });
    }
    return structures;
}
function buildObjectAliases(model, importedModels, schemaUsage) {
    var _a, _b, _c;
    const objectSchemas = ((_a = model.schemas) !== null && _a !== void 0 ? _a : []).filter((o) => {
        var _a;
        return (0, schemaHelpers_js_1.isObjectSchema)(o) &&
            ((_a = o.usage) === null || _a === void 0 ? void 0 : _a.some((u) => schemaUsage.includes(u)));
    });
    const objectAliases = [];
    for (const objectSchema of objectSchemas) {
        // FIXME: disabling new multipart generation for modular while we figure out the story
        if (objectSchema.isMultipartBody && !((_b = model.options) === null || _b === void 0 ? void 0 : _b.isModularLibrary)) {
            const propertySignatures = getPropertySignatures((_c = objectSchema.properties) !== null && _c !== void 0 ? _c : {}, schemaUsage, importedModels, { flattenBinaryArrays: true });
            const objectTypeNames = propertySignatures.map((sig) => (0, nameConstructors_js_1.getMultipartPartTypeName)(objectSchema.name, sig.name));
            objectAliases.push({
                kind: ts_morph_1.StructureKind.TypeAlias,
                ...(objectSchema.description && {
                    docs: [{ description: objectSchema.description }]
                }),
                name: objectSchema.typeName,
                type: `FormData | Array<${objectTypeNames.join("|") || "unknown"}>`,
                isExported: true
            });
        }
        if (objectSchema.alias || objectSchema.outputAlias) {
            const description = objectSchema.description;
            const modelName = schemaUsage.includes(interfaces_js_1.SchemaContext.Input)
                ? `${objectSchema.typeName}`
                : `${objectSchema.outputTypeName}`;
            objectAliases.push({
                kind: ts_morph_1.StructureKind.TypeAlias,
                ...(description && { docs: [{ description }] }),
                name: modelName,
                type: schemaUsage.includes(interfaces_js_1.SchemaContext.Input)
                    ? `${objectSchema.alias}`
                    : `${objectSchema.outputAlias}`,
                isExported: true,
                docs: [description !== null && description !== void 0 ? description : "Alias for " + modelName]
            });
        }
    }
    return objectAliases;
}
function buildPolymorphicAliases(model, schemaUsage) {
    var _a;
    // We'll add aliases for polymorphic objects
    const objectAliases = [];
    const objectSchemas = ((_a = model.schemas) !== null && _a !== void 0 ? _a : []).filter((o) => {
        var _a;
        return (0, schemaHelpers_js_1.isObjectSchema)(o) &&
            ((_a = o.usage) === null || _a === void 0 ? void 0 : _a.some((u) => schemaUsage.includes(u)));
    });
    for (const objectSchema of objectSchemas) {
        const baseName = getObjectBaseName(objectSchema, schemaUsage);
        const typeAlias = getPolymorphicTypeAlias(baseName, objectSchema, schemaUsage);
        if (typeAlias) {
            objectAliases.push(typeAlias);
        }
    }
    return objectAliases;
}
/**
 * Gets a base name for an object schema this is tipically used with suffixes when building interface or type names
 */
function getObjectBaseName(objectSchema, schemaUsage) {
    const nameSuffix = schemaUsage.includes(interfaces_js_1.SchemaContext.Output) ? "Output" : "";
    const name = (0, nameUtils_js_1.normalizeName)(objectSchema.name, nameUtils_js_1.NameType.Interface, true /** guard name */);
    return `${name}${nameSuffix}`;
}
/**
 * If the current object is a Polymorphic parent, we need to create
 * a type alias with the union of its children to enable polymorphism
 */
function getPolymorphicTypeAlias(baseName, objectSchema, schemaUsage) {
    var _a, _b;
    if (!isPolymorphicParent(objectSchema)) {
        return undefined;
    }
    const unionTypes = [];
    // If the object itself has a discriminatorValue add its base to the union
    if (objectSchema.discriminatorValue) {
        unionTypes.push(`${baseName}Parent`);
    }
    for (const child of (_b = (_a = objectSchema.children) === null || _a === void 0 ? void 0 : _a.all) !== null && _b !== void 0 ? _b : []) {
        const nameSuffix = schemaUsage.includes(interfaces_js_1.SchemaContext.Output)
            ? "Output"
            : "";
        const name = (0, nameUtils_js_1.normalizeName)(child.name, nameUtils_js_1.NameType.Interface, true /** shouldGuard */);
        unionTypes.push(`${name}${nameSuffix}`);
    }
    const description = objectSchema.description;
    return {
        kind: ts_morph_1.StructureKind.TypeAlias,
        ...(description && { docs: [{ description }] }),
        name: `${baseName}`,
        type: unionTypes.join(" | "),
        isExported: true
    };
}
/**
 * Builds the interface for the current object schema. If it is a polymorphic
 * root node it will suffix it with Base.
 */
function getObjectInterfaceDeclaration(model, baseName, objectSchema, schemaUsage, importedModels) {
    var _a;
    let interfaceName = `${baseName}`;
    if (isPolymorphicParent(objectSchema)) {
        interfaceName = `${baseName}Parent`;
    }
    const properties = (_a = objectSchema.properties) !== null && _a !== void 0 ? _a : {};
    let propertySignatures = getPropertySignatures(properties, schemaUsage, importedModels);
    // Add the polymorphic property if exists
    propertySignatures = addDiscriminatorProperty(model, objectSchema, propertySignatures, schemaUsage);
    // Calculate the parents of the current object
    const extendFrom = getImmediateParentsNames(objectSchema, schemaUsage);
    const description = objectSchema.description;
    return {
        kind: ts_morph_1.StructureKind.Interface,
        ...(description && { docs: [{ description }] }),
        name: interfaceName,
        isExported: true,
        properties: propertySignatures,
        ...(extendFrom && { extends: extendFrom })
    };
}
function isPolymorphicParent(objectSchema) {
    return objectSchema.isPolyParent ? true : false;
}
function addDiscriminatorProperty(model, objectSchema, properties, schemaUsage) {
    const polymorphicProperty = getDiscriminatorProperty(model, objectSchema, schemaUsage);
    if (polymorphicProperty) {
        // It is possible that the polymorphic property needs to override an existing property.
        // This is usually the case on the top level parent where the property already has a type of string
        // we need to replace it with the polymorphic values of its children
        const filteredProperties = properties.filter((p) => p.name !== polymorphicProperty.name);
        return [...filteredProperties, polymorphicProperty];
    }
    return properties;
}
/**
 * Finds the name of the property used as discriminator and the discriminator value.
 */
function getDiscriminatorProperty(model, objectSchema, schemaUsage) {
    var _a, _b, _c, _d, _e, _f;
    const discriminatorValue = objectSchema.discriminatorValue;
    if (!discriminatorValue && !objectSchema.discriminator) {
        return undefined;
    }
    const discriminators = getDiscriminatorValue(objectSchema);
    const discriminatorPropertyName = getDiscriminatorPropertyName(objectSchema);
    if (discriminators) {
        if (discriminatorPropertyName === undefined) {
            throw new Error(`getDiscriminatorProperty: Expected object ${objectSchema.name} to have a discriminator in its hierarchy but found none`);
        }
        const inputTypeName = (_b = (_a = objectSchema.discriminator) === null || _a === void 0 ? void 0 : _a.typeName) !== null && _b !== void 0 ? _b : (_c = objectSchema.discriminator) === null || _c === void 0 ? void 0 : _c.type;
        return {
            kind: ts_morph_1.StructureKind.PropertySignature,
            name: `"${discriminatorPropertyName}"`,
            type: ((_d = model.options) === null || _d === void 0 ? void 0 : _d.sourceFrom) === "Swagger"
                ? discriminators
                : schemaUsage.includes(interfaces_js_1.SchemaContext.Output)
                    ? ((_f = (_e = objectSchema.discriminator) === null || _e === void 0 ? void 0 : _e.outputTypeName) !== null && _f !== void 0 ? _f : inputTypeName)
                    : inputTypeName
        };
    }
    return undefined;
}
/**
 * Finds the closest discriminator property
 */
function getDiscriminatorPropertyName(objectSchema) {
    var _a, _b;
    if (objectSchema.discriminator !== undefined) {
        return objectSchema.discriminator.name;
    }
    const allParents = (_b = (_a = objectSchema.parents) === null || _a === void 0 ? void 0 : _a.all) !== null && _b !== void 0 ? _b : [];
    for (const parent of allParents) {
        if ((0, schemaHelpers_js_1.isObjectSchema)(parent) && parent.discriminator) {
            return parent.discriminator.name;
        }
    }
    return undefined;
}
/**
 * Calculates the discriminator values that a given object needs
 */
function getDiscriminatorValue(objectSchema) {
    var _a, _b, _c, _d;
    const discriminatorValue = objectSchema.discriminatorValue
        ? objectSchema.discriminatorValue
        : objectSchema.discriminator
            ? objectSchema.name
            : undefined;
    const children = (_b = (_a = objectSchema.children) === null || _a === void 0 ? void 0 : _a.immediate) !== null && _b !== void 0 ? _b : [];
    // If the current object has a discriminatorValue but doesn't have any children
    // it is a leaf node and the only discriminator value needed is itself
    if (discriminatorValue && !children.length) {
        return `"${discriminatorValue}"`;
    }
    // when the current object has both discriminator and discriminatorValue
    if (children) {
        const discriminatorProperty = objectSchema.discriminator;
        // Even when there are children, if no discriminatorProperty is present this is a leaf in the polymorphism tree
        if (!discriminatorProperty) {
            return `"${discriminatorValue}"`;
        }
        // the current object has discriminated children we need to find all the discriminatorValues for each of its children
        const allChildren = (_d = (_c = objectSchema.children) === null || _c === void 0 ? void 0 : _c.all) !== null && _d !== void 0 ? _d : [];
        // Top level parents may not have a discriminator of their own.
        const selfDiscriminator = discriminatorValue
            ? [`"${discriminatorValue}"`]
            : [];
        const childValues = getChildDiscriminatorValues(allChildren).map((v) => `"${v}"`);
        return [...selfDiscriminator, ...childValues].join(" | ");
    }
    return undefined;
}
/**
 * Looks into the children and grabs all possible discriminatorValues
 */
function getChildDiscriminatorValues(children) {
    const discriminatorValues = new Set();
    for (const child of children) {
        if ((0, schemaHelpers_js_1.isObjectSchema)(child) && child.discriminatorValue) {
            discriminatorValues.add(child.discriminatorValue);
        }
    }
    return [...discriminatorValues];
}
/**
 * Gets a list of types a given object may extend from
 */
function getImmediateParentsNames(objectSchema, schemaUsage) {
    var _a;
    if (!((_a = objectSchema.parents) === null || _a === void 0 ? void 0 : _a.immediate)) {
        return [];
    }
    const extendFrom = [];
    // If an immediate parent is an empty DictionarySchema, that means that the object has been marked
    // with additional properties. We need to add Record<string, unknown> to the extend list and
    if (objectSchema.parents.immediate.find((im) => (0, schemaHelpers_js_1.isDictionarySchema)(im, { filterEmpty: true }))) {
        extendFrom.push("Record<string, unknown>");
    }
    // Get the rest of the parents excluding any DictionarySchemas
    const parents = objectSchema.parents.immediate
        .filter((p) => !(0, schemaHelpers_js_1.isDictionarySchema)(p, { filterEmpty: true }))
        .map((parent) => {
        var _a, _b;
        const nameSuffix = schemaUsage.includes(interfaces_js_1.SchemaContext.Output)
            ? "Output"
            : "";
        const name = (0, schemaHelpers_js_1.isDictionarySchema)(parent)
            ? ((_a = Object.entries(objectSchema.properties)) === null || _a === void 0 ? void 0 : _a.some((prop) => {
                var _a, _b, _c;
                const typeName = (_a = prop[1].typeName) !== null && _a !== void 0 ? _a : prop[1].type;
                return (`Record<string, ${typeName}>` !== parent.typeName &&
                    !((_c = (_b = parent.additionalProperties) === null || _b === void 0 ? void 0 : _b.typeName) === null || _c === void 0 ? void 0 : _c.includes(typeName)));
            }))
                ? schemaUsage.includes(interfaces_js_1.SchemaContext.Output)
                    ? "Record<string, any>"
                    : "Record<string, unknown>"
                : `${(_b = (schemaUsage.includes(interfaces_js_1.SchemaContext.Output)
                    ? parent.outputTypeName
                    : parent.typeName)) !== null && _b !== void 0 ? _b : parent.name}`
            : `${(0, nameUtils_js_1.normalizeName)(parent.name, nameUtils_js_1.NameType.Interface, true /** shouldGuard */)}${nameSuffix}`;
        return (0, schemaHelpers_js_1.isObjectSchema)(parent) && isPolymorphicParent(parent)
            ? `${name}Parent`
            : name;
    });
    return [...parents, ...extendFrom];
}
function getPropertySignatures(properties, schemaUsage, importedModels, options = {}) {
    let validProperties = Object.keys(properties);
    const readOnlyFilter = (name) => !(schemaUsage.includes(interfaces_js_1.SchemaContext.Input) && properties[name].readOnly);
    const neverFilter = (name) => properties[name].type !== "never";
    validProperties = validProperties.filter(readOnlyFilter).filter(neverFilter);
    return validProperties.map((p) => getPropertySignature({ ...properties[p], name: p }, schemaUsage, importedModels, options));
}
function isBinaryArray(schema) {
    var _a, _b, _c, _d;
    return Boolean((0, schemaHelpers_js_1.isArraySchema)(schema) &&
        (((_b = (_a = schema.items) === null || _a === void 0 ? void 0 : _a.typeName) === null || _b === void 0 ? void 0 : _b.includes("NodeJS.ReadableStream")) ||
            ((_d = (_c = schema.items) === null || _c === void 0 ? void 0 : _c.outputTypeName) === null || _d === void 0 ? void 0 : _d.includes("NodeJS.ReadableStream"))));
}
/**
 * Builds a Typescript property or parameter signature
 * @param schema - Property or parameter to get the Typescript signature for
 * @param importedModels - Set to track the models that need to be imported
 * @returns a PropertySignatureStructure for the property.
 */
function getPropertySignature(property, schemaUsage, importedModels, options = {}) {
    var _a, _b, _c;
    let schema;
    if (options.flattenBinaryArrays && isBinaryArray(property)) {
        schema = {
            ...((_a = property.items) !== null && _a !== void 0 ? _a : property),
            name: property.name
        };
    }
    else {
        schema = property;
    }
    const propertyName = schema.name;
    const description = schema.description;
    let type;
    const hasCoreInArray = schema.type === "array" &&
        schema.items &&
        schema.items.fromCore;
    const hasCoreInRecord = schema.type === "dictionary" &&
        schema.additionalProperties &&
        schema.additionalProperties.fromCore;
    if (hasCoreInArray && schema.typeName) {
        type = schema.typeName;
        importedModels.add((_b = schema.items.typeName) !== null && _b !== void 0 ? _b : schema.items.name);
    }
    else if (hasCoreInRecord && schema.typeName) {
        type = schema.typeName;
        importedModels.add((_c = schema.additionalProperties.typeName) !== null && _c !== void 0 ? _c : schema.additionalProperties.name);
    }
    else {
        type =
            generateForOutput(schemaUsage, schema.usage) && schema.outputTypeName
                ? schema.outputTypeName
                : schema.typeName
                    ? schema.typeName
                    : schema.type;
        if (schema.typeName && schema.fromCore) {
            importedModels.add(schema.typeName);
            type = schema.typeName;
        }
    }
    return {
        name: propertyName,
        ...(description && { docs: [{ description }] }),
        hasQuestionToken: !schema.required,
        isReadonly: generateForOutput(schemaUsage, schema.usage) && schema.readOnly,
        type,
        kind: ts_morph_1.StructureKind.PropertySignature
    };
}
function generateForOutput(schemaUsage, propertyUsage) {
    return ((schemaUsage.includes(interfaces_js_1.SchemaContext.Output) &&
        (propertyUsage === null || propertyUsage === void 0 ? void 0 : propertyUsage.includes(interfaces_js_1.SchemaContext.Output))) ||
        (schemaUsage.includes(interfaces_js_1.SchemaContext.Exception) &&
            (propertyUsage === null || propertyUsage === void 0 ? void 0 : propertyUsage.includes(interfaces_js_1.SchemaContext.Exception))));
}
