// @azure-tools/rlc-common/dist/buildClientDefinitions.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildClientDefinitions = buildClientDefinitions;
const ts_morph_1 = require("ts-morph");
const path = require("path");
const operationHelpers_js_1 = require("./helpers/operationHelpers.js");
const shortcutMethods_js_1 = require("./helpers/shortcutMethods.js");
const buildMethodShortcuts_js_1 = require("./buildMethodShortcuts.js");
const nameUtils_js_1 = require("./helpers/nameUtils.js");
const nameUtils_js_2 = require("./helpers/nameUtils.js");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
const importsUtil_js_1 = require("./helpers/importsUtil.js");
function buildClientDefinitions(model) {
    var _a, _b, _c;
    const options = {
        importedParameters: new Set(),
        importedResponses: new Set(),
        clientImports: new Set()
    };
    const project = new ts_morph_1.Project();
    const srcPath = model.srcPath;
    const filePath = path.join(srcPath, `clientDefinitions.ts`);
    const clientDefinitionsFile = project.createSourceFile(filePath, undefined, {
        overwrite: true
    });
    // Get all paths
    const pathDictionary = model.paths;
    let shortcuts = [];
    // There may be operations without an operation group, those shortcut
    // methods need to be handled differently.
    let shortcutsInOperationGroup = [];
    if ((_a = model.options) === null || _a === void 0 ? void 0 : _a.includeShortcuts) {
        shortcuts = (0, shortcutMethods_js_1.generateMethodShortcuts)(model.paths);
        clientDefinitionsFile.addInterfaces(shortcuts);
        shortcutsInOperationGroup = shortcuts
            .filter((s) => s.name !== "ClientOperations")
            .map((s) => getShortcutName(s.name));
    }
    clientDefinitionsFile.addInterface({
        name: "Routes",
        isExported: true,
        callSignatures: getPathFirstRoutesInterfaceDefinition(pathDictionary, clientDefinitionsFile, options)
    });
    const clientInterfaceName = (0, nameConstructors_js_1.getClientName)(model);
    clientDefinitionsFile.addTypeAlias({
        isExported: true,
        name: clientInterfaceName,
        type: ts_morph_1.Writers.intersectionType("Client", ts_morph_1.Writers.objectType({
            properties: [
                { name: "path", type: "Routes" },
                ...shortcutsInOperationGroup
            ]
        }), 
        // If the length of shortcuts in operation group and all shortcutsInOperationGroup
        // is the same, then we don't have any operations at the client level
        // Otherwise we need to make the client interface name an union with the
        // definition of all client level shortcut methods
        ...(shortcutsInOperationGroup.length !== shortcuts.length
            ? [`ClientOperations`]
            : []))
    });
    if (options.importedParameters.size) {
        clientDefinitionsFile.addImportDeclaration({
            isTypeOnly: true,
            namedImports: [...options.importedParameters],
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({ cjsName: "./parameters", esModulesName: "./parameters.js" }, model)
        });
    }
    if (options.importedResponses.size) {
        clientDefinitionsFile.addImportDeclaration({
            isTypeOnly: true,
            namedImports: [...options.importedResponses],
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({ cjsName: "./responses", esModulesName: "./responses.js" }, model)
        });
    }
    if (((_c = (_b = model.importInfo.internalImports.rlcClientDefinition.importsSet) === null || _b === void 0 ? void 0 : _b.size) !== null && _c !== void 0 ? _c : 0) > 0) {
        clientDefinitionsFile.addImportDeclaration({
            isTypeOnly: true,
            namedImports: Array.from(model.importInfo.internalImports.rlcClientDefinition.importsSet),
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({ cjsName: "./models", esModulesName: "./models.js" }, model)
        });
    }
    options.clientImports.add("Client");
    options.clientImports.add("StreamableMethod");
    clientDefinitionsFile.addImportDeclarations([
        {
            isTypeOnly: true,
            namedImports: [...options.clientImports],
            moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restClient", model.importInfo.runtimeImports)
        }
    ]);
    return { path: filePath, content: clientDefinitionsFile.getFullText() };
}
function getPathFirstRoutesInterfaceDefinition(paths, sourcefile, options) {
    const operationGroupCount = getOperationGroupCount(paths);
    const signatures = [];
    for (const key of Object.keys(paths)) {
        for (const verb of Object.keys(paths[key].methods)) {
            for (const method of paths[key].methods[verb]) {
                options.importedParameters.add(method.optionsName);
                method.returnType
                    .split(" | ")
                    .forEach((item) => options.importedResponses.add(item));
            }
        }
        generatePathFirstRouteMethodsDefinition(paths[key], operationGroupCount, sourcefile);
        const pathParams = paths[key].pathParameters;
        (0, operationHelpers_js_1.getGeneratedWrapperTypes)(pathParams).forEach((p) => { var _a; return options.importedParameters.add((_a = p.name) !== null && _a !== void 0 ? _a : p.type); });
        signatures.push({
            docs: [
                `Resource for '${key
                    .replace(/}/g, "\\}")
                    .replace(/{/g, "\\{")}' has methods for the following verbs: ${Object.keys(paths[key].methods).join(", ")}`
            ],
            parameters: [
                { name: "path", type: `"${key}"` },
                ...(0, operationHelpers_js_1.getPathParamDefinitions)(pathParams)
            ],
            returnType: getOperationReturnTypeName(paths[key], getOperationGroupCount(paths)),
            kind: ts_morph_1.StructureKind.CallSignature
        });
    }
    return signatures;
}
function getOperationGroupCount(paths) {
    const operationGroups = Object.keys(paths)
        .map((p) => paths[p].operationGroupName)
        .filter((p) => p && p !== "Client");
    const uniqueNames = new Set(operationGroups);
    return uniqueNames.size;
}
function getOperationReturnTypeName({ operationGroupName, name }, operationGroupCount) {
    if (operationGroupCount > 1 &&
        operationGroupName &&
        operationGroupName !== "Client") {
        return (0, nameUtils_js_1.normalizeName)(`${(0, nameUtils_js_2.pascalCase)(operationGroupName)}${(0, nameUtils_js_2.pascalCase)(name)}`, nameUtils_js_1.NameType.Interface);
    }
    return (0, nameUtils_js_2.pascalCase)(name);
}
function generatePathFirstRouteMethodsDefinition(path, operationGroupCount, file) {
    const methodDefinitions = (0, operationHelpers_js_1.buildMethodDefinitions)(path.methods);
    const interfaceDef = {
        methods: methodDefinitions,
        name: getOperationReturnTypeName(path, operationGroupCount),
        isExported: true
    };
    file.addInterface(interfaceDef);
}
function getShortcutName(interfaceName) {
    const endIndex = shouldKeepSuffix(interfaceName)
        ? undefined
        : interfaceName.length - "Operations".length;
    const clientProperty = (0, nameUtils_js_1.normalizeName)(interfaceName.substring(0, endIndex), nameUtils_js_1.NameType.OperationGroup, true, buildMethodShortcuts_js_1.REST_CLIENT_RESERVED, nameUtils_js_1.CasingConvention.Camel);
    return {
        name: clientProperty,
        type: interfaceName
    };
}
function shouldKeepSuffix(name) {
    const reservedNames = [
        "pipelineOperations",
        "pathOperations",
        "pathUncheckedOperations"
    ];
    return reservedNames.some((r) => r.toLowerCase() === name.toLowerCase());
}
