// @azure-tools/rlc-common/dist/buildSchemaType.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildSchemaTypes = buildSchemaTypes;
exports.generateModelFiles = generateModelFiles;
const ts_morph_1 = require("ts-morph");
const path = require("path");
const buildObjectTypes_js_1 = require("./buildObjectTypes.js");
const interfaces_js_1 = require("./interfaces.js");
const importsUtil_js_1 = require("./helpers/importsUtil.js");
/**
 * Generates types to represent schema definitions in the swagger
 */
function buildSchemaTypes(model) {
    const { srcPath } = model;
    const project = new ts_morph_1.Project();
    let filePath = path.join(srcPath, `models.ts`);
    const inputModelFile = generateModelFiles(model, project, filePath, [
        interfaces_js_1.SchemaContext.Input
    ]);
    filePath = path.join(srcPath, `outputModels.ts`);
    const outputModelFile = generateModelFiles(model, project, filePath, [
        interfaces_js_1.SchemaContext.Output,
        interfaces_js_1.SchemaContext.Exception
    ]);
    return { inputModelFile, outputModelFile };
}
function generateModelFiles(model, project, filePath, schemaContext) {
    // Track models that need to be imported
    const importedModels = new Set();
    const objectsDefinitions = (0, buildObjectTypes_js_1.buildObjectInterfaces)(model, importedModels, schemaContext);
    const objectTypeAliases = (0, buildObjectTypes_js_1.buildPolymorphicAliases)(model, schemaContext);
    const objectAliases = (0, buildObjectTypes_js_1.buildObjectAliases)(model, importedModels, schemaContext);
    if (objectTypeAliases.length ||
        objectsDefinitions.length ||
        objectAliases.length) {
        const modelsFile = project.createSourceFile(filePath, undefined, {
            overwrite: true
        });
        modelsFile.addInterfaces(objectsDefinitions);
        modelsFile.addTypeAliases(objectTypeAliases);
        modelsFile.addTypeAliases(objectAliases);
        if (importedModels.size > 0) {
            modelsFile.addImportDeclarations([
                {
                    isTypeOnly: true,
                    namedImports: [...Array.from(importedModels || [])],
                    moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restClient", model.importInfo.runtimeImports)
                }
            ]);
        }
        return { path: filePath, content: modelsFile.getFullText() };
    }
    return undefined;
}
