// @azure-tools/rlc-common/dist/buildClient.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildClient = buildClient;
exports.getClientFactoryBody = getClientFactoryBody;
const ts_morph_1 = require("ts-morph");
const path = require("path");
const nameUtils_js_1 = require("./helpers/nameUtils.js");
const buildMethodShortcuts_js_1 = require("./buildMethodShortcuts.js");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
const importsUtil_js_1 = require("./helpers/importsUtil.js");
const packageUtil_js_1 = require("./helpers/packageUtil.js");
function getClientOptionsInterface(model, clientName, optionalUrlParameters) {
    var _a, _b, _c, _d, _e;
    if ((!optionalUrlParameters || optionalUrlParameters.length === 0) &&
        !model.apiVersionInfo) {
        return undefined;
    }
    const properties = (_a = optionalUrlParameters === null || optionalUrlParameters === void 0 ? void 0 : optionalUrlParameters.map((param) => {
        var _a;
        return {
            name: param.name,
            type: param.type,
            hasQuestionToken: true,
            docs: [
                (_a = param.description) !== null && _a !== void 0 ? _a : "client level optional parameter " + param.name
            ]
        };
    })) !== null && _a !== void 0 ? _a : [];
    if (((_b = model.apiVersionInfo) === null || _b === void 0 ? void 0 : _b.isCrossedVersion) === false &&
        !((_d = (_c = model.urlInfo) === null || _c === void 0 ? void 0 : _c.urlParameters) === null || _d === void 0 ? void 0 : _d.find((p) => p.name === "apiVersion")) &&
        (model.apiVersionInfo.defaultValue || !((_e = model.apiVersionInfo) === null || _e === void 0 ? void 0 : _e.required))) {
        properties.push({
            name: "apiVersion",
            type: "string",
            hasQuestionToken: true,
            docs: ["The api version option of the client"]
        });
    }
    return {
        name: `${clientName}Options`,
        extends: ["ClientOptions"],
        isExported: true,
        properties,
        docs: ["The optional parameters for the client"]
    };
}
function buildClient(model) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    const name = (0, nameUtils_js_1.normalizeName)(model.libraryName, nameUtils_js_1.NameType.File);
    const { srcPath } = model;
    const project = new ts_morph_1.Project();
    const filePath = path.join(srcPath, `${name}.ts`);
    const clientFile = project.createSourceFile(filePath, undefined, {
        overwrite: true
    });
    // Get all paths
    const clientInterfaceName = (0, nameConstructors_js_1.getClientName)(model);
    normalizeUrlInfo(model);
    const urlParameters = (_b = (_a = model === null || model === void 0 ? void 0 : model.urlInfo) === null || _a === void 0 ? void 0 : _a.urlParameters) === null || _b === void 0 ? void 0 : _b.filter(
    // Do not include parameters with constant values in the signature, these should go in the options bag
    (p) => p.value === undefined);
    const optionalUrlParameters = (_d = (_c = model === null || model === void 0 ? void 0 : model.urlInfo) === null || _c === void 0 ? void 0 : _c.urlParameters) === null || _d === void 0 ? void 0 : _d.filter(
    // Do not include parameters with constant values in the signature, these should go in the options bag
    (p) => Boolean(p.value));
    const clientOptionsInterface = getClientOptionsInterface(model, clientInterfaceName, optionalUrlParameters);
    if (clientOptionsInterface) {
        clientFile.addInterface(clientOptionsInterface);
    }
    if (!model.options) {
        return undefined;
    }
    const { multiClient, batch } = model.options;
    const { addCredentials, credentialScopes, credentialKeyHeaderName, customHttpAuthHeaderName, customHttpAuthSharedKeyPrefix } = model.options;
    const credentialTypes = credentialScopes ? ["TokenCredential"] : [];
    if (credentialKeyHeaderName || customHttpAuthHeaderName) {
        credentialTypes.push("KeyCredential");
    }
    const commonClientParams = [
        ...(urlParameters !== null && urlParameters !== void 0 ? urlParameters : []),
        ...(addCredentials === false ||
            !isSecurityInfoDefined(credentialScopes, credentialKeyHeaderName, customHttpAuthHeaderName)
            ? []
            : [
                {
                    name: "credentials",
                    type: credentialTypes.join(" | "),
                    description: `uniquely identify client credential`
                }
            ])
    ];
    let apiVersionStatement = "";
    // Set the default api-version when we have a default AND its position is query
    if (((_e = model.apiVersionInfo) === null || _e === void 0 ? void 0 : _e.isCrossedVersion) === false &&
        !!((_f = model.apiVersionInfo) === null || _f === void 0 ? void 0 : _f.defaultValue)) {
        apiVersionStatement = `
    apiVersion = "${(_g = model.apiVersionInfo) === null || _g === void 0 ? void 0 : _g.defaultValue}"`;
    }
    else if (((_h = model.apiVersionInfo) === null || _h === void 0 ? void 0 : _h.isCrossedVersion) === false &&
        !model.apiVersionInfo.required) {
        apiVersionStatement = `
    apiVersion`;
    }
    const allClientParams = [
        ...commonClientParams,
        {
            name: apiVersionStatement === ""
                ? "options"
                : `{${apiVersionStatement}, ...options}`,
            documentName: "options",
            type: `${(_j = clientOptionsInterface === null || clientOptionsInterface === void 0 ? void 0 : clientOptionsInterface.name) !== null && _j !== void 0 ? _j : "ClientOptions"} = {}`,
            description: "the parameter for all optional parameters"
        }
    ];
    const functionStatement = {
        isExported: true,
        name: `createClient`,
        parameters: allClientParams,
        docs: [
            {
                description: `Initialize a new instance of \`${clientInterfaceName}\`\n` +
                    allClientParams
                        .map((param) => {
                        var _a, _b;
                        return `@param ${(_a = param.documentName) !== null && _a !== void 0 ? _a : param.name} - ${(_b = param.description) !== null && _b !== void 0 ? _b : "The parameter " + param.name}`;
                    })
                        .join("\n")
            }
        ],
        returnType: clientInterfaceName,
        isDefaultExport: false,
        statements: getClientFactoryBody(model, clientInterfaceName, {
            isMultipleCredential: credentialTypes.length > 1
        })
    };
    if (!multiClient || !batch || batch.length === 1) {
        functionStatement.isDefaultExport = true;
    }
    clientFile.addFunction(functionStatement);
    const paths = srcPath.replace(/\//g, path.sep).split(path.sep);
    while (paths.length > 0 && paths[paths.length - 1] === "") {
        paths.pop();
    }
    const parentPath = paths.lastIndexOf("src") > -1
        ? paths.length - 1 - paths.lastIndexOf("src")
        : 0;
    const loggerPath = `${parentPath > 0 ? "../".repeat(parentPath) : "./"}logger`;
    clientFile.addImportDeclarations([
        {
            isTypeOnly: true,
            namedImports: ["ClientOptions"],
            moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restClient", model.importInfo.runtimeImports)
        }
    ]);
    clientFile.addImportDeclarations([
        {
            namedImports: ["getClient"],
            moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restClient", model.importInfo.runtimeImports)
        }
    ]);
    if ((0, packageUtil_js_1.isAzurePackage)(model)) {
        clientFile.addImportDeclarations([
            {
                namedImports: ["logger"],
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: loggerPath,
                    esModulesName: `${loggerPath}.js`
                }, model)
            }
        ]);
    }
    const includeKeyCredentialHelper = customHttpAuthHeaderName &&
        customHttpAuthSharedKeyPrefix &&
        credentialTypes.length > 1 &&
        credentialTypes.includes("KeyCredential");
    if (addCredentials &&
        isSecurityInfoDefined(credentialScopes, credentialKeyHeaderName, customHttpAuthHeaderName)) {
        clientFile.addImportDeclarations([
            {
                isTypeOnly: true,
                namedImports: credentialTypes,
                moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("coreAuth", model.importInfo.runtimeImports)
            }
        ]);
        if (includeKeyCredentialHelper) {
            clientFile.addImportDeclarations([
                {
                    namedImports: ["isKeyCredential"],
                    moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("coreAuth", model.importInfo.runtimeImports)
                }
            ]);
        }
    }
    clientFile.addImportDeclarations([
        {
            isTypeOnly: true,
            namedImports: [`${clientInterfaceName}`],
            moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                cjsName: "./clientDefinitions",
                esModulesName: "./clientDefinitions.js"
            }, model)
        }
    ]);
    if (((_o = (_m = (_l = (_k = model.importInfo.internalImports) === null || _k === void 0 ? void 0 : _k.rlcClientFactory) === null || _l === void 0 ? void 0 : _l.importsSet) === null || _m === void 0 ? void 0 : _m.size) !== null && _o !== void 0 ? _o : 0) > 0) {
        clientFile.addImportDeclarations([
            {
                isTypeOnly: true,
                namedImports: Array.from(model.importInfo.internalImports.rlcClientFactory.importsSet),
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./models`,
                    esModulesName: `./models.js`
                }, model)
            }
        ]);
    }
    return { path: filePath, content: clientFile.getFullText() };
}
function isSecurityInfoDefined(credentialScopes, credentialKeyHeaderName, customHttpAuthHeaderName) {
    return (credentialScopes || credentialKeyHeaderName || customHttpAuthHeaderName);
}
function getClientFactoryBody(model, clientTypeName, options = { isMultipleCredential: false }) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    if (!model.options || !model.options.packageDetails || !model.urlInfo) {
        return "";
    }
    const { includeShortcuts, packageDetails, addCredentials } = model.options;
    let clientPackageName = (_b = (_a = packageDetails.nameWithoutScope) !== null && _a !== void 0 ? _a : packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.name) !== null && _b !== void 0 ? _b : "";
    const packageVersion = packageDetails.version;
    const { endpoint, urlParameters } = model.urlInfo;
    const optionalUrlParameters = [];
    for (const param of urlParameters !== null && urlParameters !== void 0 ? urlParameters : []) {
        if (param.name === "apiVersion") {
            continue;
        }
        if (param.value) {
            const value = typeof param.value === "string" ? `"${param.value}"` : param.value;
            optionalUrlParameters.push(`const ${param.name} = options.${param.name} ?? ${value};`);
        }
    }
    let endpointUrl;
    if (urlParameters && endpoint) {
        let parsedEndpoint = endpoint;
        urlParameters.forEach((urlParameter) => {
            parsedEndpoint = parsedEndpoint.replace(`{${urlParameter.name}}`, `\${${urlParameter.name}}`);
        });
        endpointUrl = `options.endpoint ?? \`${parsedEndpoint}\``;
    }
    else {
        endpointUrl = `options.endpoint ?? "${endpoint}"`;
    }
    if (!model.options.isModularLibrary && !clientPackageName.endsWith("-rest")) {
        clientPackageName += "-rest";
    }
    const userAgentInfoStatement = "const userAgentInfo = `azsdk-js-" +
        clientPackageName +
        "/" +
        packageVersion +
        "`;";
    const userAgentPrefix = "options.userAgentOptions && options.userAgentOptions.userAgentPrefix ? `${options.userAgentOptions.userAgentPrefix} ${userAgentInfo}`: `${userAgentInfo}`;";
    const userAgentStatement = {
        kind: ts_morph_1.StructureKind.VariableStatement,
        declarationKind: ts_morph_1.VariableDeclarationKind.Const,
        declarations: [{ name: "userAgentPrefix", initializer: userAgentPrefix }]
    };
    const customHeaderOptions = ((_c = model.telemetryOptions) === null || _c === void 0 ? void 0 : _c.customRequestIdHeaderName)
        ? `,
    telemetryOptions: {
      clientRequestIdHeaderName:
        options.telemetryOptions?.clientRequestIdHeaderName ??
        "${(_d = model.telemetryOptions) === null || _d === void 0 ? void 0 : _d.customRequestIdHeaderName}"
    }`
        : "";
    const endpointUrlStatement = {
        kind: ts_morph_1.StructureKind.VariableStatement,
        declarationKind: ts_morph_1.VariableDeclarationKind.Const,
        declarations: [{ name: "endpointUrl", initializer: endpointUrl }]
    };
    const { credentialScopes, credentialKeyHeaderName } = model.options;
    const scopesString = credentialScopes
        ? credentialScopes.map((cs) => `"${cs}"`).join(", ") ||
            "`${endpointUrl}/.default`"
        : "";
    const scopes = scopesString
        ? `scopes: options.credentials?.scopes ?? [${scopesString}],`
        : "";
    const apiKeyHeaderName = credentialKeyHeaderName
        ? `apiKeyHeaderName: options.credentials?.apiKeyHeaderName ?? "${credentialKeyHeaderName}",`
        : "";
    const loggerOptions = (0, packageUtil_js_1.isAzurePackage)(model)
        ? `,
  loggingOptions: {
    logger: options.loggingOptions?.logger ?? logger.info
  }`
        : "";
    const credentialsOptions = (scopes || apiKeyHeaderName) && addCredentials
        ? `,
      credentials: {
        ${scopes}
        ${apiKeyHeaderName}
      }`
        : "";
    const overrideOptionsStatement = `options = {
        ...options,
        userAgentOptions: {
          userAgentPrefix
        }${loggerOptions}${customHeaderOptions}${credentialsOptions}
      };`;
    const getClient = `const client = getClient(
        endpointUrl, ${credentialsOptions ? "credentials," : ""} options
      ) as ${clientTypeName};
      `;
    const { customHttpAuthHeaderName, customHttpAuthSharedKeyPrefix } = model.options;
    let customHttpAuthStatement = "";
    if (customHttpAuthHeaderName && customHttpAuthSharedKeyPrefix) {
        if (options.isMultipleCredential) {
            customHttpAuthStatement = `if (isKeyCredential(credentials)) {
        client.pipeline.addPolicy({
          name: "customKeyCredentialPolicy",
          async sendRequest(request, next) {
            request.headers.set("Authorization", "Bearer " + credentials.key);
            return next(request);
          },
        });
      }`;
        }
        else {
            customHttpAuthStatement = `
      client.pipeline.addPolicy({
        name: "customKeyCredentialPolicy",
        async sendRequest(request, next) {
          request.headers.set("${customHttpAuthHeaderName}", "${customHttpAuthSharedKeyPrefix} " + credentials.key);
          return next(request);
        }
      });`;
        }
    }
    let apiVersionPolicyStatement = `client.pipeline.removePolicy({ name: "ApiVersionPolicy" });`;
    if ((0, packageUtil_js_1.isAzurePackage)(model) &&
        ((_e = model.apiVersionInfo) === null || _e === void 0 ? void 0 : _e.isCrossedVersion) !== false) {
        apiVersionPolicyStatement += `
      if (options.apiVersion) {
        logger.warning("This client does not support client api-version, please change it at the operation level");
      }`;
    }
    else if ((0, packageUtil_js_1.isAzurePackage)(model) &&
        !((_f = model.apiVersionInfo) === null || _f === void 0 ? void 0 : _f.defaultValue) &&
        ((_g = model.apiVersionInfo) === null || _g === void 0 ? void 0 : _g.required)) {
        apiVersionPolicyStatement += `
      if (options.apiVersion) {
        logger.warning("This client does not support to set api-version in options, please change it at positional argument");
      }`;
    }
    if (((_h = model.apiVersionInfo) === null || _h === void 0 ? void 0 : _h.isCrossedVersion) === false &&
        ((_j = model.apiVersionInfo) === null || _j === void 0 ? void 0 : _j.definedPosition) === "query") {
        apiVersionPolicyStatement += `
      client.pipeline.addPolicy({
        name: 'ClientApiVersionPolicy',
        sendRequest: (req, next) => {
          // Use the apiVersion defined in request url directly
          // Append one if there is no apiVersion and we have one at client options
          const url = new URL(req.url);
          if (!url.searchParams.get("api-version") && apiVersion) {
            req.url = \`\${req.url}\${
              Array.from(url.searchParams.keys()).length > 0 ? "&" : "?"
            }api-version=\${apiVersion}\`;
          }
    
          return next(req);
        },
      });`;
    }
    let returnStatement = `return client;`;
    if (includeShortcuts) {
        const shortcutImplementations = (0, buildMethodShortcuts_js_1.buildMethodShortcutImplementation)(model.paths);
        const shortcutBody = Object.keys(shortcutImplementations).map((key) => {
            // If the operation group has an empty name, it means its operations are client
            // level operations so we need to spread the definitions. Otherwise they are
            // within an operation group so we add them as key: value
            return `${key && key !== "client" ? `"${key}":` : "..."} {${shortcutImplementations[key].join()}}`;
        });
        returnStatement = `return { ...client, ${shortcutBody.join()} };`;
    }
    return [
        ...optionalUrlParameters,
        endpointUrlStatement,
        userAgentInfoStatement,
        userAgentStatement,
        overrideOptionsStatement,
        getClient,
        apiVersionPolicyStatement,
        customHttpAuthStatement,
        returnStatement
    ];
}
function normalizeUrlInfo(model) {
    if (!model ||
        !model.urlInfo ||
        !model.urlInfo.endpoint ||
        !model.urlInfo.urlParameters ||
        model.urlInfo.urlParameters.length === 0) {
        return;
    }
    let parsedEndpoint = model.urlInfo.endpoint;
    const urlParameters = model.urlInfo.urlParameters;
    urlParameters.forEach((urlParameter) => {
        const name = urlParameter.name;
        const normalizedName = (0, nameUtils_js_1.normalizeName)(name, nameUtils_js_1.NameType.Parameter);
        if (name !== normalizedName) {
            urlParameter.name = normalizedName;
            parsedEndpoint = parsedEndpoint.replace(`{${name}}`, `{${normalizedName}}`);
        }
    });
    model.urlInfo.endpoint = parsedEndpoint;
}
