// @azure-tools/rlc-common/dist/buildMethodShortcuts.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.REST_CLIENT_RESERVED = void 0;
exports.buildMethodShortcutImplementation = buildMethodShortcutImplementation;
const nameUtils_js_1 = require("./helpers/nameUtils.js");
exports.REST_CLIENT_RESERVED = [
    { name: "path", reservedFor: [nameUtils_js_1.NameType.Property, nameUtils_js_1.NameType.OperationGroup] },
    {
        name: "pathUnchecked",
        reservedFor: [nameUtils_js_1.NameType.Property, nameUtils_js_1.NameType.OperationGroup]
    },
    {
        name: "pipeline",
        reservedFor: [nameUtils_js_1.NameType.Property, nameUtils_js_1.NameType.OperationGroup]
    }
];
function buildMethodShortcutImplementation(paths) {
    const keys = {};
    for (const path of Object.keys(paths)) {
        const groupName = (0, nameUtils_js_1.normalizeName)(paths[path].operationGroupName, nameUtils_js_1.NameType.OperationGroup, true, exports.REST_CLIENT_RESERVED, nameUtils_js_1.CasingConvention.Camel);
        if (keys[groupName]) {
            keys[groupName].push(...buildOperationDeclarations(path, paths[path]));
        }
        else {
            keys[groupName] = buildOperationDeclarations(path, paths[path]);
        }
    }
    return keys;
}
function buildOperationDeclarations(path, pathMetadata) {
    let ops = [];
    for (const method of Object.keys(pathMetadata.methods)) {
        for (const op of pathMetadata.methods[method]) {
            const pathParams = pathMetadata === null || pathMetadata === void 0 ? void 0 : pathMetadata.pathParameters;
            const name = (0, nameUtils_js_1.normalizeName)(op.operationName, nameUtils_js_1.NameType.Property);
            const methodDefinitions = generateOperationDeclaration(path, name, method, pathParams);
            ops = [...ops, methodDefinitions];
        }
    }
    return ops;
}
function generateOperationDeclaration(path, operationName, method, pathParams = []) {
    const pathParamNames = `${pathParams.length > 0 ? `${pathParams.map((p) => p.name)},` : ""}`;
    return `"${operationName}": (${pathParamNames} options) => {
      return client.path("${path}", ${pathParamNames}).${method}(options);
    }`;
}
