// @azure-tools/rlc-common/dist/buildResponseTypes.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildResponseTypes = buildResponseTypes;
const ts_morph_1 = require("ts-morph");
const path = require("path");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
const importsUtil_js_1 = require("./helpers/importsUtil.js");
let hasErrorResponse = false;
function buildResponseTypes(model) {
    var _a, _b, _c, _d;
    const project = new ts_morph_1.Project();
    const srcPath = model.srcPath;
    const filePath = path.join(srcPath, `responses.ts`);
    hasErrorResponse = false;
    const responsesFile = project.createSourceFile(filePath, undefined, {
        overwrite: true
    });
    // Set used to track down which models need to be imported
    // Track if we need to import RawHttpHeaders
    let hasHeaders = false;
    if (!model.responses) {
        return;
    }
    for (const operationResponse of model.responses) {
        for (const response of operationResponse.responses) {
            // Building the response type base name
            const baseResponseName = (0, nameConstructors_js_1.getResponseBaseName)(operationResponse.operationGroup, operationResponse.operationName, response.statusCode);
            // Build the response header
            const headersInterface = getResponseHeaderInterfaceDefinition(response, baseResponseName);
            if (headersInterface) {
                hasHeaders = true;
                responsesFile.addInterface(headersInterface);
            }
            // Get the information to build the Response Interface
            const responseTypeName = (_a = response.predefinedName) !== null && _a !== void 0 ? _a : (0, nameConstructors_js_1.getResponseTypeName)(baseResponseName);
            const responseProperties = getResponseInterfaceProperties(response, headersInterface === null || headersInterface === void 0 ? void 0 : headersInterface.name);
            const responseInterfaceDefinition = {
                name: responseTypeName,
                properties: responseProperties,
                isExported: true,
                extends: ["HttpResponse"]
            };
            // Only add a description if one was provided in the Swagger
            // otherwise skip to avoid having empty TSDoc lines
            if (response.description) {
                responseInterfaceDefinition.docs = [
                    { description: response.description }
                ];
            }
            // Add the response interface to the responses file
            responsesFile.addInterface(responseInterfaceDefinition);
        }
    }
    if (hasHeaders) {
        responsesFile.addImportDeclarations([
            {
                isTypeOnly: true,
                namedImports: ["RawHttpHeaders"],
                moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restPipeline", model.importInfo.runtimeImports)
            }
        ]);
    }
    const namedImports = ["HttpResponse"];
    if (hasErrorResponse) {
        namedImports.push("ErrorResponse");
    }
    responsesFile.addImportDeclarations([
        {
            isTypeOnly: true,
            namedImports,
            moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restClient", model.importInfo.runtimeImports)
        }
    ]);
    if (((_d = (_c = (_b = model.importInfo.internalImports.response) === null || _b === void 0 ? void 0 : _b.importsSet) === null || _c === void 0 ? void 0 : _c.size) !== null && _d !== void 0 ? _d : 0) > 0) {
        const modelNamedImports = Array.from(model.importInfo.internalImports.response.importsSet).filter((modelName) => {
            return !(modelName === "ErrorResponseOutput" && hasErrorResponse);
        });
        responsesFile.addImportDeclarations([
            {
                isTypeOnly: true,
                namedImports: modelNamedImports,
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./outputModels`,
                    esModulesName: `./outputModels.js`
                }, model)
            }
        ]);
    }
    return { path: filePath, content: responsesFile.getFullText() };
}
function getResponseHeaderInterfaceDefinition(response, baseName) {
    // Check if there are any required headers
    if (!response.headers) {
        return;
    }
    const headersInterfaceName = `${baseName}Headers`;
    return {
        kind: ts_morph_1.StructureKind.Interface,
        isExported: true,
        name: headersInterfaceName,
        properties: response === null || response === void 0 ? void 0 : response.headers.map((h) => {
            const description = h.description;
            return {
                name: h.name,
                ...(description && { docs: [{ description }] }),
                type: h.type,
                hasQuestionToken: !h.required
            };
        })
    };
}
/**
 * Gets the properties that need to be part of the response interface
 */
function getResponseInterfaceProperties(response, headersInterfaceName) {
    const statusCode = response.statusCode;
    const responseProperties = [
        {
            name: "status",
            type: statusCode === "default" ? `string` : `"${statusCode}"`,
            kind: ts_morph_1.StructureKind.PropertySignature
        }
    ];
    if (response.body) {
        const description = response.body.description;
        let type = response.body.type;
        if (response.body.type === "ErrorResponseOutput" &&
            response.body.fromCore) {
            type = "ErrorResponse";
            hasErrorResponse = true;
        }
        responseProperties.push({
            name: "body",
            type,
            kind: ts_morph_1.StructureKind.PropertySignature,
            ...(description && { docs: [{ description }] })
        });
    }
    if (headersInterfaceName) {
        responseProperties.push({
            name: "headers",
            type: `RawHttpHeaders & ${headersInterfaceName}`,
            kind: ts_morph_1.StructureKind.PropertySignature
        });
    }
    return responseProperties;
}
