// @azure-tools/rlc-common/dist/buildIsUnexpectedHelper.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildIsUnexpectedHelper = buildIsUnexpectedHelper;
const path = require("path");
const ts_morph_1 = require("ts-morph");
const operationHelpers_js_1 = require("./helpers/operationHelpers.js");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
function buildIsUnexpectedHelper(model) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    if (!(0, operationHelpers_js_1.hasUnexpectedHelper)(model)) {
        return;
    }
    const project = new ts_morph_1.Project();
    const srcPath = model.srcPath;
    const filePath = path.join(srcPath, `isUnexpected.ts`);
    const isErrorHelper = project.createSourceFile(filePath, undefined, {
        overwrite: true
    });
    let map = {};
    const allResponseTypes = new Set();
    const allErrorTypes = new Set();
    const overloads = [];
    const pathDictionary = model.paths;
    for (const [path, details] of Object.entries(pathDictionary)) {
        for (const [methodName, methodDetails] of Object.entries(details.methods)) {
            const originalMethod = methodName.toUpperCase();
            const operation = `${originalMethod} ${path}`;
            const successCodeSet = new Set((_a = map.operation) !== null && _a !== void 0 ? _a : []);
            for (const detail of methodDetails) {
                detail.successStatus.forEach(successCodeSet.add, successCodeSet);
                // LROs may call the same path but with GET
                // to get the operation status.
                if (((_c = (_b = detail.operationHelperDetail) === null || _b === void 0 ? void 0 : _b.lroDetails) === null || _c === void 0 ? void 0 : _c.isLongRunning) &&
                    originalMethod !== "GET") {
                    const operation = `GET ${path}`;
                    const logicalSuccessCodes = ((_f = (_e = (_d = detail.operationHelperDetail) === null || _d === void 0 ? void 0 : _d.lroDetails) === null || _e === void 0 ? void 0 : _e.logicalResponseTypes) === null || _f === void 0 ? void 0 : _f.success)
                        ? ["200"]
                        : [];
                    const initialSuccessCodes = (_h = (pathDictionary[path].methods["get"] &&
                        ((_g = pathDictionary[path].methods["get"][0]) === null || _g === void 0 ? void 0 : _g.successStatus))) !== null && _h !== void 0 ? _h : detail.successStatus;
                    map = {
                        ...map,
                        ...{
                            [operation]: Array.from(new Set(logicalSuccessCodes.concat(initialSuccessCodes)))
                        }
                    };
                }
                const successTypes = [...detail.responseTypes.success];
                const errorTypes = [...detail.responseTypes.error];
                if (((_j = model.helperDetails) === null || _j === void 0 ? void 0 : _j.clientLroOverload) &&
                    ((_m = (_l = (_k = detail.operationHelperDetail) === null || _k === void 0 ? void 0 : _k.lroDetails) === null || _l === void 0 ? void 0 : _l.logicalResponseTypes) === null || _m === void 0 ? void 0 : _m.success)) {
                    successTypes.push(...((_o = detail.operationHelperDetail.lroDetails.logicalResponseTypes
                        .success) !== null && _o !== void 0 ? _o : []));
                }
                if (!successTypes.length || !errorTypes.length || !errorTypes[0]) {
                    continue;
                }
                successTypes.forEach((t) => allResponseTypes.add(t));
                errorTypes.forEach((t) => {
                    allResponseTypes.add(t);
                    allErrorTypes.add(t);
                });
                overloads.push({
                    isExported: true,
                    parameters: [
                        {
                            name: "response",
                            type: [...successTypes, ...errorTypes].join(" | ")
                        }
                    ],
                    returnType: `response is ${errorTypes[0]}`
                });
            }
            map = { ...map, ...{ [operation]: Array.from(successCodeSet) } };
        }
    }
    isErrorHelper.addImportDeclaration({
        isTypeOnly: true,
        namedImports: [...allResponseTypes],
        moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
            cjsName: `./responses`,
            esModulesName: `./responses.js`
        }, model)
    });
    isErrorHelper.addVariableStatement({
        declarations: [
            {
                name: "responseMap",
                initializer: JSON.stringify(map),
                type: "Record<string, string[]>"
            }
        ],
        declarationKind: ts_morph_1.VariableDeclarationKind.Const
    });
    if (allErrorTypes.size) {
        isErrorHelper.addFunction({
            overloads,
            isExported: true,
            name: "isUnexpected",
            parameters: [
                {
                    name: "response",
                    type: [...allResponseTypes].join(" | ")
                }
            ],
            returnType: `response is ${[...allErrorTypes].join(" | ")}`,
            statements: [
                `
          const lroOriginal = response.headers["x-ms-original-url"];
          const url = new URL(lroOriginal ?? response.request.url);
          const method = response.request.method;
          let pathDetails = responseMap[\`\${method} \${url.pathname}\`];
          if (!pathDetails) {
            pathDetails = getParametrizedPathSuccess(method, url.pathname);
          }
          return !pathDetails.includes(response.status);
        `
            ]
        });
        isErrorHelper.addFunction({
            isExported: false,
            name: "getParametrizedPathSuccess",
            parameters: [
                {
                    name: "method",
                    type: "string"
                },
                {
                    name: "path",
                    type: "string"
                }
            ],
            returnType: `string[]`,
            statements: [
                `
          const pathParts = path.split("/");
            
          // Traverse list to match the longest candidate
          // matchedLen: the length of candidate path
          // matchedValue: the matched status code array
          let matchedLen = -1,
            matchedValue: string[] = [];
          
          // Iterate the responseMap to find a match
          for (const [key, value] of Object.entries(responseMap)) {
            // Extracting the path from the map key which is in format
            // GET /path/foo
            if (!key.startsWith(method)) {
              continue;
            }
            const candidatePath = getPathFromMapKey(key);
            // Get each part of the url path
            const candidateParts = candidatePath.split("/");
        
            // track if we have found a match to return the values found.
            let found = true;
            for (
              let i = candidateParts.length - 1, j = pathParts.length - 1;
              i >= 1 && j >= 1;
              i--, j--
            ) {
              if (
                candidateParts[i]?.startsWith("{") &&
                candidateParts[i]?.indexOf("}") !== -1
              ) {
                const start = candidateParts[i]!.indexOf("}") + 1,
                  end = candidateParts[i]?.length;
                // If the current part of the candidate is a "template" part
                // Try to use the suffix of pattern to match the path
                // {guid} ==> $
                // {guid}:export ==> :export$
                const isMatched = new RegExp(
                  \`\${candidateParts[i]?.slice(start, end)}\`
                ).test(pathParts[j] || '');
        
                if (!isMatched) {
                  found = false;
                  break;
                }
                continue;
              }
        
              // If the candidate part is not a template and
              // the parts don't match mark the candidate as not found
              // to move on with the next candidate path.
              if (candidateParts[i] !== pathParts[j]) {
                found = false;
                break;
              }
            }
        
            // We finished evaluating the current candidate parts
            // Update the matched value if and only if we found the longer pattern
            if (found && candidatePath.length > matchedLen) {
              matchedLen = candidatePath.length;
              matchedValue = value;
            }
          }
        
          return matchedValue;
        `
            ]
        });
        isErrorHelper.addFunction({
            isExported: false,
            name: "getPathFromMapKey",
            parameters: [
                {
                    name: "mapKey",
                    type: "string"
                }
            ],
            returnType: `string`,
            statements: [
                `const pathStart = mapKey.indexOf("/");
         return mapKey.slice(pathStart);`
            ]
        });
    }
    return {
        path: filePath,
        content: isErrorHelper.getFullText()
    };
}
