// @azure-tools/rlc-common/dist/buildPollingHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildPollingHelper = buildPollingHelper;
const interfaces_js_1 = require("./interfaces.js");
const path = require("path");
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore: to fix the handlebars issue
const handlebars_1 = require("handlebars");
const operationHelpers_js_1 = require("./helpers/operationHelpers.js");
const pollingContent_js_1 = require("./static/pollingContent.js");
function buildPollingHelper(model) {
    if (!(0, operationHelpers_js_1.hasPollingOperations)(model)) {
        return;
    }
    const lroDetail = buildLroHelperDetail(model);
    const readmeFileContents = handlebars_1.default.compile(pollingContent_js_1.pollingContent, { noEscape: true });
    const { srcPath } = model;
    return {
        path: path.join(srcPath, "pollingHelper.ts"),
        content: readmeFileContents(lroDetail)
    };
}
function buildLroHelperDetail(model) {
    var _a, _b, _c, _d, _e, _f;
    if (!((_a = model.helperDetails) === null || _a === void 0 ? void 0 : _a.clientLroOverload)) {
        return {
            clientOverload: false,
            isEsm: ((_b = model.options) === null || _b === void 0 ? void 0 : _b.moduleKind) === "esm"
        };
    }
    const mapDetail = [];
    const pathDictionary = model.paths;
    const responses = new Set();
    for (const details of Object.values(pathDictionary)) {
        for (const methodDetails of Object.values(details.methods)) {
            const lroDetail = (_c = methodDetails[0].operationHelperDetail) === null || _c === void 0 ? void 0 : _c.lroDetails;
            if (lroDetail === null || lroDetail === void 0 ? void 0 : lroDetail.isLongRunning) {
                const initialResponses = methodDetails[0].responseTypes.success.concat(methodDetails[0].responseTypes.error);
                const finalResponse = (_d = lroDetail.logicalResponseTypes) === null || _d === void 0 ? void 0 : _d.success.concat(methodDetails[0].responseTypes.error);
                if (initialResponses && finalResponse) {
                    initialResponses.forEach((n) => responses.add(n));
                    finalResponse.forEach((n) => responses.add(n));
                    mapDetail.push({
                        initialResponses: initialResponses.join("|"),
                        finalResponses: finalResponse.join("|"),
                        precedence: (_e = lroDetail.precedence) !== null && _e !== void 0 ? _e : interfaces_js_1.OPERATION_LRO_HIGH_PRIORITY
                    });
                }
            }
        }
    }
    // Sorted by the precedence
    mapDetail.sort((d1, d2) => d1.precedence - d2.precedence);
    return {
        clientOverload: responses.size > 0 && mapDetail.length > 0,
        importedResponses: Array.from(responses),
        overloadMap: mapDetail,
        isEsm: ((_f = model.options) === null || _f === void 0 ? void 0 : _f.moduleKind) === "esm"
    };
}
