// @azure-tools/rlc-common/dist/buildParameterTypes.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildParameterTypes = buildParameterTypes;
exports.buildBodyTypeAlias = buildBodyTypeAlias;
const ts_morph_1 = require("ts-morph");
const path = require("path");
const interfaces_js_1 = require("./interfaces.js");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
const importsUtil_js_1 = require("./helpers/importsUtil.js");
const buildObjectTypes_js_1 = require("./buildObjectTypes.js");
const operationHelpers_js_1 = require("./helpers/operationHelpers.js");
function buildParameterTypes(model) {
    var _a, _b, _c, _d, _e, _f;
    const project = new ts_morph_1.Project();
    const srcPath = model.srcPath;
    const filePath = path.join(srcPath, `parameters.ts`);
    const partialBodyTypeNames = new Set();
    const parametersFile = project.createSourceFile(filePath, undefined, {
        overwrite: true
    });
    let hasHeaders = false;
    if (!model.parameters) {
        return;
    }
    for (const requestParameter of model.parameters) {
        const baseParameterName = (0, nameConstructors_js_1.getParameterBaseName)(requestParameter.operationGroup, requestParameter.operationName);
        const requestCount = (_b = (_a = requestParameter === null || requestParameter === void 0 ? void 0 : requestParameter.parameters) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
        const topParamName = (0, nameConstructors_js_1.getParameterTypeName)(baseParameterName);
        const subParamNames = [];
        // We need to loop the requests. An operation with multiple requests means that
        // the operation can get different values for content-type and each value may
        // have a different type associated to it.
        for (let i = 0; i < requestCount; i++) {
            const parameter = requestParameter.parameters[i];
            const internalReferences = new Set();
            // In case we have more than one request to model we need to add a suffix to differentiate
            const nameSuffix = i > 0 ? `${i}` : "";
            const parameterInterfaceName = requestCount > 1
                ? `${baseParameterName}RequestParameters${nameSuffix}`
                : topParamName;
            const queryParameterDefinitions = buildQueryParameterDefinition(model, parameter, baseParameterName, internalReferences, i);
            const pathParameterDefinitions = buildPathParameterDefinitions(model, parameter, baseParameterName, parametersFile, internalReferences, i);
            const headerParameterDefinitions = buildHeaderParameterDefinitions(parameter, baseParameterName, parametersFile, internalReferences, i);
            const contentTypeParameterDefinition = buildContentTypeParametersDefinition(parameter, baseParameterName, internalReferences, i);
            const bodyParameterDefinition = buildBodyParametersDefinition(parameter, baseParameterName, internalReferences, i);
            const bodyTypeAlias = buildBodyTypeAlias(parameter, partialBodyTypeNames);
            if (bodyTypeAlias) {
                parametersFile.addTypeAlias(bodyTypeAlias);
            }
            // Add interfaces for body and query parameters
            parametersFile.addInterfaces([
                ...(bodyParameterDefinition !== null && bodyParameterDefinition !== void 0 ? bodyParameterDefinition : []),
                ...(queryParameterDefinitions !== null && queryParameterDefinitions !== void 0 ? queryParameterDefinitions : []),
                ...(pathParameterDefinitions !== null && pathParameterDefinitions !== void 0 ? pathParameterDefinitions : []),
                ...(headerParameterDefinitions ? [headerParameterDefinitions] : []),
                ...(contentTypeParameterDefinition
                    ? [contentTypeParameterDefinition]
                    : [])
            ]);
            // Add Operation parameters type alias which is composed of the types we generated above
            // plus the common type RequestParameters
            parametersFile.addTypeAlias({
                name: parameterInterfaceName,
                isExported: true,
                type: [...internalReferences, "RequestParameters"].join(" & ")
            });
            subParamNames.push(parameterInterfaceName);
            if (headerParameterDefinitions !== undefined) {
                hasHeaders = true;
            }
        }
        // Add Operation parameters type alias which is composed of the types we generated above
        // plus the common type RequestParameters
        if (requestCount > 1) {
            parametersFile.addTypeAlias({
                name: topParamName,
                isExported: true,
                type: [...subParamNames].join(" | ")
            });
        }
    }
    if (hasHeaders) {
        parametersFile.addImportDeclarations([
            {
                isTypeOnly: true,
                namedImports: ["RawHttpHeadersInput"],
                moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restPipeline", model.importInfo.runtimeImports)
            }
        ]);
    }
    parametersFile.addImportDeclarations([
        {
            isTypeOnly: true,
            namedImports: ["RequestParameters"],
            moduleSpecifier: (0, importsUtil_js_1.getImportSpecifier)("restClient", model.importInfo.runtimeImports)
        }
    ]);
    if (((_f = (_e = (_d = (_c = model.importInfo.internalImports) === null || _c === void 0 ? void 0 : _c.parameter) === null || _d === void 0 ? void 0 : _d.importsSet) === null || _e === void 0 ? void 0 : _e.size) !== null && _f !== void 0 ? _f : 0) > 0) {
        parametersFile.addImportDeclarations([
            {
                isTypeOnly: true,
                namedImports: Array.from(model.importInfo.internalImports.parameter.importsSet),
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `./models`,
                    esModulesName: `./models.js`
                }, model)
            }
        ]);
    }
    return { path: filePath, content: parametersFile.getFullText() };
}
function buildQueryParameterDefinition(model, parameters, baseName, internalReferences, requestIndex) {
    const queryParameters = ((parameters === null || parameters === void 0 ? void 0 : parameters.parameters) || []).filter((p) => p.type === "query");
    if (!queryParameters.length) {
        return undefined;
    }
    const nameSuffix = requestIndex > 0 ? `${requestIndex}` : "";
    const queryParameterInterfaceName = `${baseName}QueryParam${nameSuffix}`;
    const queryParameterPropertiesName = `${baseName}QueryParamProperties`;
    // Get the property signature for each query parameter
    const propertiesDefinition = queryParameters.map((qp) => getPropertyFromSchema(qp.param));
    // Get wrapper types for query parameters
    const wrapperTypesDefinition = (0, operationHelpers_js_1.getGeneratedWrapperTypes)(queryParameters).map((wrapObj) => {
        return (0, buildObjectTypes_js_1.getObjectInterfaceDeclaration)(model, wrapObj.name, wrapObj, [interfaces_js_1.SchemaContext.Input], new Set());
    });
    const hasRequiredParameters = propertiesDefinition.some((p) => !p.hasQuestionToken);
    const propertiesInterface = {
        kind: ts_morph_1.StructureKind.Interface,
        isExported: true,
        name: queryParameterPropertiesName,
        properties: propertiesDefinition
    };
    const parameterInterface = {
        kind: ts_morph_1.StructureKind.Interface,
        isExported: true,
        name: queryParameterInterfaceName,
        properties: [
            {
                name: "queryParameters",
                type: queryParameterPropertiesName,
                // Mark as optional if there are no required parameters
                hasQuestionToken: !hasRequiredParameters
            }
        ]
    };
    // Mark the queryParameter interface for importing
    internalReferences.add(queryParameterInterfaceName);
    return [...wrapperTypesDefinition, propertiesInterface, parameterInterface];
}
function getPropertyFromSchema(schema) {
    const description = schema.description;
    return {
        name: schema.name,
        ...(description && { docs: [{ description }] }),
        type: schema.type,
        hasQuestionToken: !schema.required,
        kind: ts_morph_1.StructureKind.PropertySignature
    };
}
function buildPathParameterDefinitions(model, parameters, baseName, parametersFile, internalReferences, requestIndex) {
    const pathParameters = (parameters.parameters || []).filter((p) => p.type === "path");
    if (!pathParameters.length) {
        return undefined;
    }
    const allDefinitions = [];
    buildClientPathParameters();
    buildMethodWrapParameters();
    return allDefinitions;
    function buildClientPathParameters() {
        var _a;
        // we only have client-level path parameters if the source is from swagger
        if (((_a = model.options) === null || _a === void 0 ? void 0 : _a.sourceFrom) === "TypeSpec") {
            return;
        }
        const clientPathParams = pathParameters.length > 0 ? pathParameters : [];
        const nameSuffix = requestIndex > 0 ? `${requestIndex}` : "";
        const pathParameterInterfaceName = `${baseName}PathParam${nameSuffix}`;
        const pathInterface = getPathInterfaceDefinition(clientPathParams, baseName);
        if (pathInterface) {
            parametersFile.addInterface(pathInterface);
        }
        internalReferences.add(pathParameterInterfaceName);
        allDefinitions.push({
            isExported: true,
            kind: ts_morph_1.StructureKind.Interface,
            name: pathParameterInterfaceName,
            properties: [
                {
                    name: "pathParameters",
                    type: `${baseName}PathParameters`,
                    kind: ts_morph_1.StructureKind.PropertySignature
                }
            ]
        });
    }
    function buildMethodWrapParameters() {
        var _a;
        if (((_a = model.options) === null || _a === void 0 ? void 0 : _a.sourceFrom) === "Swagger") {
            return;
        }
        // we only have method-level path parameters if the source is from typespec
        const methodPathParams = pathParameters.length > 0 ? pathParameters : [];
        // we only need to build the wrapper types if the path parameters are objects
        const wrapperTypesDefinition = (0, operationHelpers_js_1.getGeneratedWrapperTypes)(methodPathParams).map((wrap) => {
            return (0, buildObjectTypes_js_1.getObjectInterfaceDeclaration)(model, wrap.name, wrap, [interfaces_js_1.SchemaContext.Input], new Set());
        });
        allDefinitions.push(...wrapperTypesDefinition);
    }
}
function getPathInterfaceDefinition(pathParameters, baseName) {
    const pathInterfaceName = `${baseName}PathParameters`;
    return {
        kind: ts_morph_1.StructureKind.Interface,
        isExported: true,
        name: pathInterfaceName,
        properties: pathParameters.map((p) => getPropertyFromSchema(p.param))
    };
}
function buildHeaderParameterDefinitions(parameters, baseName, parametersFile, internalReferences, requestIndex) {
    const headerParameters = (parameters.parameters || []).filter((p) => p.type === "header" && p.name !== "contentType");
    if (!headerParameters.length) {
        return undefined;
    }
    const nameSuffix = requestIndex > 0 ? `${requestIndex}` : "";
    const headerParameterInterfaceName = `${baseName}HeaderParam${nameSuffix}`;
    const headersInterface = getRequestHeaderInterfaceDefinition(headerParameters, baseName);
    let isOptional = true;
    if (headersInterface) {
        parametersFile.addInterface(headersInterface);
        isOptional = !(headersInterface.properties || []).some((prop) => prop.hasQuestionToken === false);
    }
    internalReferences.add(headerParameterInterfaceName);
    return {
        isExported: true,
        kind: ts_morph_1.StructureKind.Interface,
        name: headerParameterInterfaceName,
        properties: [
            {
                name: "headers",
                type: `RawHttpHeadersInput & ${baseName}Headers`,
                kind: ts_morph_1.StructureKind.PropertySignature,
                hasQuestionToken: isOptional
            }
        ]
    };
}
function getRequestHeaderInterfaceDefinition(headerParameters, baseName) {
    const headersInterfaceName = `${baseName}Headers`;
    return {
        kind: ts_morph_1.StructureKind.Interface,
        isExported: true,
        name: headersInterfaceName,
        properties: headerParameters.map((h) => getPropertyFromSchema(h.param))
    };
}
function buildContentTypeParametersDefinition(parameters, baseName, internalReferences, requestIndex) {
    const mediaTypeParameters = (parameters.parameters || []).filter((p) => p.type === "header" && p.name === "contentType");
    if (!mediaTypeParameters.length) {
        return undefined;
    }
    const nameSuffix = requestIndex > 0 ? `${requestIndex}` : "";
    const mediaTypesParameterInterfaceName = `${baseName}MediaTypesParam${nameSuffix}`;
    // Mark the queryParameter interface for importing
    internalReferences.add(mediaTypesParameterInterfaceName);
    const mediaParam = mediaTypeParameters[0].param;
    return {
        isExported: true,
        kind: ts_morph_1.StructureKind.Interface,
        name: mediaTypesParameterInterfaceName,
        properties: [getPropertyFromSchema(mediaParam)]
    };
}
function buildBodyParametersDefinition(parameters, baseName, internalReferences, requestIndex) {
    const bodyParameters = parameters.body;
    if (!bodyParameters ||
        !(bodyParameters === null || bodyParameters === void 0 ? void 0 : bodyParameters.body) ||
        !(bodyParameters === null || bodyParameters === void 0 ? void 0 : bodyParameters.body.length)) {
        return [];
    }
    const nameSuffix = requestIndex > 0 ? `${requestIndex}` : "";
    const bodyParameterInterfaceName = `${baseName}BodyParam${nameSuffix}`;
    internalReferences.add(bodyParameterInterfaceName);
    // In case of formData we'd get multiple properties in body marked as partialBody
    if (bodyParameters.isPartialBody) {
        let allOptionalParts = true;
        const propertiesDefinitions = [];
        for (const param of bodyParameters.body) {
            if (param.required) {
                allOptionalParts = false;
            }
            propertiesDefinitions.push(getPropertyFromSchema(param));
        }
        const formBodyName = `${baseName}FormBody`;
        const formBodyInterface = {
            isExported: true,
            kind: ts_morph_1.StructureKind.Interface,
            name: formBodyName,
            properties: propertiesDefinitions
        };
        return [
            {
                isExported: true,
                kind: ts_morph_1.StructureKind.Interface,
                name: bodyParameterInterfaceName,
                properties: [
                    {
                        name: "body",
                        type: formBodyName,
                        hasQuestionToken: allOptionalParts
                    }
                ]
            },
            formBodyInterface
        ];
    }
    else {
        const bodySignature = getPropertyFromSchema(bodyParameters.body[0]);
        return [
            {
                isExported: true,
                kind: ts_morph_1.StructureKind.Interface,
                name: bodyParameterInterfaceName,
                properties: [
                    {
                        docs: bodySignature.docs,
                        name: "body",
                        type: bodySignature.type,
                        hasQuestionToken: bodySignature.hasQuestionToken
                    }
                ]
            }
        ];
    }
}
function buildBodyTypeAlias(parameters, partialBodyTypeNames) {
    const bodyParameters = parameters.body;
    if (!bodyParameters ||
        !(bodyParameters === null || bodyParameters === void 0 ? void 0 : bodyParameters.body) ||
        !(bodyParameters === null || bodyParameters === void 0 ? void 0 : bodyParameters.body.length)) {
        return undefined;
    }
    const schema = bodyParameters.body[0];
    const headerParameters = (parameters.parameters || []).filter((p) => p.type === "header" && p.name === "contentType");
    if (!headerParameters.length || headerParameters.length > 1) {
        return undefined;
    }
    const contentType = headerParameters[0].param.type;
    const description = `${schema.description}`;
    const typeName = `${schema.typeName}ResourceMergeAndPatch`;
    if (partialBodyTypeNames.has(typeName)) {
        return null;
    }
    else {
        partialBodyTypeNames.add(typeName);
    }
    if (contentType.includes("application/merge-patch+json")) {
        const type = `Partial<${schema.typeName}>`;
        return {
            // kind: StructureKind.TypeAlias,
            ...(description && { docs: [{ description }] }),
            name: `${typeName}`,
            type,
            isExported: true
        };
    }
}
