// @azure-tools/rlc-common/dist/buildTopLevelIndexFile.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTopLevelIndex = buildTopLevelIndex;
const nameUtils_js_1 = require("./helpers/nameUtils.js");
const ts_morph_1 = require("ts-morph");
const path = require("path");
const pathUtils_js_1 = require("./helpers/pathUtils.js");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
const batchOutputFolder = [];
function buildTopLevelIndex(model) {
    if (!model.options) {
        return undefined;
    }
    const project = new ts_morph_1.Project();
    const { srcPath } = model;
    const { multiClient } = model.options;
    const batch = model.options.batch;
    if (srcPath) {
        const clientName = model.libraryName;
        const moduleName = (0, nameUtils_js_1.normalizeName)(clientName, nameUtils_js_1.NameType.File);
        const relativePath = "./" +
            (0, pathUtils_js_1.getRelativePartFromSrcPath)(srcPath, model.options.isModularLibrary);
        batchOutputFolder.push([relativePath, clientName, moduleName]);
    }
    if (multiClient &&
        batch &&
        batch.length > 1 &&
        batchOutputFolder.length === batch.length) {
        const indexFile = project.createSourceFile("index.ts", undefined, {
            overwrite: true
        });
        const allModules = [];
        batchOutputFolder.forEach((item) => {
            indexFile.addImportDeclaration({
                isTypeOnly: true,
                namespaceImport: item[1],
                moduleSpecifier: (0, nameConstructors_js_1.getImportModuleName)({
                    cjsName: `${item[0]}`,
                    esModulesName: `${item[0]}/index.js`
                }, model)
            });
            allModules.push(item[1]);
        });
        indexFile.addExportDeclaration({
            namedExports: [...allModules]
        });
        const content = indexFile.getFullText();
        const filePath = path.join(srcPath.substring(0, srcPath.lastIndexOf("src") + 4), model.options.isModularLibrary ? "rest" : "", `index.ts`);
        return { path: filePath, content };
    }
}
