// @azure-tools/rlc-common/dist/transformSampleGroups.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformSampleGroups = transformSampleGroups;
const valueGenerationUtil_js_1 = require("./helpers/valueGenerationUtil.js");
const nameConstructors_js_1 = require("./helpers/nameConstructors.js");
const nameUtils_js_1 = require("./helpers/nameUtils.js");
const schemaHelpers_js_1 = require("./helpers/schemaHelpers.js");
const packageUtil_js_1 = require("./helpers/packageUtil.js");
/**
 * Transform the sample data based RLC detail e.g path, operations & schemas
 * @param model RLC detail
 * @param allowMockValue allow to mock value if not exist, currently we always generate mock value
 * @returns Generated sample data or undefined if not support to generate
 */
function transformSampleGroups(model, allowMockValue = true) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    if (((_a = model.options) === null || _a === void 0 ? void 0 : _a.multiClient) || ((_b = model.options) === null || _b === void 0 ? void 0 : _b.isModularLibrary)) {
        // Not support to generate if multiple clients
        // Not support to generate if modular libraries
        return;
    }
    if ((model.sampleGroups && model.sampleGroups.length > 0) ||
        !allowMockValue) {
        // Skip to transform if already has sample data
        // Skip to transform if not allow to mock value
        return;
    }
    const rlcSampleGroups = [];
    // Get all paths
    const paths = model.paths;
    const clientName = (0, nameConstructors_js_1.getClientName)(model);
    const clientInterfaceName = clientName.endsWith("Client")
        ? `${clientName}`
        : `${clientName}Client`;
    const defaultFactoryName = (0, nameUtils_js_1.normalizeName)((0, nameUtils_js_1.camelCase)(`create ${clientInterfaceName}`), nameUtils_js_1.NameType.Method);
    const packageName = (_e = (_d = (_c = model.options) === null || _c === void 0 ? void 0 : _c.packageDetails) === null || _d === void 0 ? void 0 : _d.name) !== null && _e !== void 0 ? _e : "";
    const methodParameterMap = buildMethodParamMap(model);
    const schemaObjectMap = (0, schemaHelpers_js_1.buildSchemaObjectMap)(model);
    for (const path in paths) {
        const pathDetails = paths[path];
        const methods = pathDetails.methods;
        for (const method in methods) {
            const importedDict = {};
            const detail = methods[method][0];
            const operatonConcante = getOperationConcate(detail.operationName, pathDetails.operationGroupName, (_f = model.options) === null || _f === void 0 ? void 0 : _f.sourceFrom);
            const operationPrefix = (0, nameUtils_js_1.normalizeName)((0, nameUtils_js_1.camelCase)(transformSpecialLetterToSpace(operatonConcante)), nameUtils_js_1.NameType.Operation);
            const sampleGroup = {
                filename: `${operationPrefix}Sample`,
                defaultFactoryName,
                clientPackageName: packageName,
                samples: []
            };
            // initialize the sample
            const sample = {
                description: `call operation ${detail.operationName}`,
                name: `${operationPrefix}Sample`,
                path,
                defaultFactoryName,
                clientParamAssignments: [],
                pathParamAssignments: [],
                methodParamAssignments: [],
                clientParamNames: "",
                pathParamNames: "",
                methodParamNames: "",
                method,
                isLRO: (_j = (_h = (_g = detail.operationHelperDetail) === null || _g === void 0 ? void 0 : _g.lroDetails) === null || _h === void 0 ? void 0 : _h.isLongRunning) !== null && _j !== void 0 ? _j : false,
                isPaging: (_l = (_k = detail.operationHelperDetail) === null || _k === void 0 ? void 0 : _k.isPaging) !== null && _l !== void 0 ? _l : false,
                useLegacyLro: false
            };
            // client-level, path-level and method-level parameter preparation
            const parameters = {
                client: convertClientLevelParameters(model, importedDict, schemaObjectMap),
                path: convertPathLevelParameters(pathDetails, path, schemaObjectMap),
                method: convertMethodLevelParameters(methods[method], schemaObjectMap, methodParameterMap.get(operatonConcante))
            };
            // enrich parameter details
            enrichParameterInSample(sample, parameters);
            // enrich LRO and pagination info
            enrichLROAndPagingInSample(detail, importedDict, packageName);
            sampleGroup.samples.push(sample);
            rlcSampleGroups.push(sampleGroup);
            enrichImportedString(sampleGroup, importedDict, defaultFactoryName, packageName);
        }
    }
    return rlcSampleGroups;
}
function enrichLROAndPagingInSample(operation, importedDict, packageName) {
    var _a, _b, _c, _d, _e;
    const isLRO = (_c = (_b = (_a = operation.operationHelperDetail) === null || _a === void 0 ? void 0 : _a.lroDetails) === null || _b === void 0 ? void 0 : _b.isLongRunning) !== null && _c !== void 0 ? _c : false, isPaging = (_e = (_d = operation.operationHelperDetail) === null || _d === void 0 ? void 0 : _d.isPaging) !== null && _e !== void 0 ? _e : false;
    if (isPaging) {
        if (isLRO) {
            // TODO: report warning this is not supported
        }
        addValueInImportedDict(packageName, "paginate", importedDict);
    }
    else if (isLRO) {
        addValueInImportedDict(packageName, "getLongRunningPoller", importedDict);
    }
}
function transformSpecialLetterToSpace(str) {
    if (!str) {
        return str;
    }
    return str
        .replace(/_/g, " ")
        .replace(/\//g, " Or ")
        .replace(/,|\.|\(|\)/g, " ")
        .replace("'s ", " ");
}
function enrichImportedString(sampleGroup, importedDict, defaultFactoryName, packageName) {
    const importedTypes = [];
    if (!importedDict[packageName] || importedDict[packageName].size === 0) {
        importedTypes.push(`import ${defaultFactoryName} from "${packageName}";`);
    }
    for (const key in importedDict) {
        const values = Array.from(importedDict[key]).join(", ");
        const hasDefaultFactory = key === packageName ? `${defaultFactoryName},` : "";
        importedTypes.push(`import ${hasDefaultFactory} { ${values} } from "${key}";`);
    }
    sampleGroup.importedTypes = importedTypes;
}
function enrichParameterInSample(sample, parameters) {
    var _a;
    sample.clientParamAssignments = getAssignmentStrArray(parameters.client);
    sample.clientParamNames = getContactParameterNames(parameters.client);
    sample.pathParamAssignments = getAssignmentStrArray(parameters.path);
    sample.pathParamNames = getContactParameterNames(parameters.path);
    // Directly apply the inline option value as method parameter
    sample.methodParamNames =
        parameters.method.length > 0 ? ((_a = parameters.method[0].value) !== null && _a !== void 0 ? _a : "") : "";
}
function getAssignmentStrArray(parameters) {
    return parameters.filter((p) => !!p.assignment).map((p) => p.assignment);
}
function getContactParameterNames(parameters) {
    return parameters
        .filter((p) => p.name != null)
        .map((p) => p.name)
        .join(",");
}
function convertClientLevelParameters(model, importedDict, schemaMap) {
    var _a, _b;
    if (!model.options) {
        return [];
    }
    const clientParams = [];
    const urlParameters = (_b = (_a = model === null || model === void 0 ? void 0 : model.urlInfo) === null || _a === void 0 ? void 0 : _a.urlParameters) === null || _b === void 0 ? void 0 : _b.filter(
    // Do not include parameters with constant values in the signature, these should go in the options bag
    (p) => p.value === undefined);
    const { addCredentials, credentialScopes, credentialKeyHeaderName, customHttpAuthHeaderName, flavor } = model.options;
    const hasUrlParameter = !!urlParameters, hasCredentials = addCredentials &&
        (credentialScopes || credentialKeyHeaderName || customHttpAuthHeaderName);
    if (hasUrlParameter) {
        // convert the host parameters in url
        const clientParamAssignments = urlParameters.map((urlParameter) => {
            const urlValue = (0, valueGenerationUtil_js_1.generateParameterTypeValue)(urlParameter.type, urlParameter.name, schemaMap);
            const normalizedName = (0, nameUtils_js_1.normalizeName)(urlParameter.name, nameUtils_js_1.NameType.Parameter);
            return {
                name: normalizedName,
                assignment: `const ${normalizedName} = ` + urlValue + `;`
            };
        });
        clientParams.push(...clientParamAssignments);
    }
    if (hasCredentials) {
        // Currently only support token credential
        const apiKeyCredentialPackage = (0, packageUtil_js_1.isAzurePackage)(model)
            ? "@azure/core-auth"
            : "@typespec/ts-http-runtime";
        const tokenCredentialPackage = (0, packageUtil_js_1.isAzurePackage)(model)
            ? "@azure/identity"
            : "@typespec/ts-http-runtime";
        if (credentialKeyHeaderName && (0, packageUtil_js_1.isAzurePackage)(model)) {
            clientParams.push({
                name: "credential",
                assignment: `const credential = new AzureKeyCredential("{Your API key}");`
            });
            addValueInImportedDict(apiKeyCredentialPackage, "AzureKeyCredential", importedDict);
        }
        else if ((credentialKeyHeaderName && flavor !== "azure") ||
            customHttpAuthHeaderName) {
            clientParams.push({
                name: "credential",
                assignment: `const credential = { key: "{Your API key}"};`
            });
        }
        else if ((0, packageUtil_js_1.isAzurePackage)(model)) {
            clientParams.push({
                name: "credential",
                assignment: "const credential = new DefaultAzureCredential();"
            });
            addValueInImportedDict(tokenCredentialPackage, "DefaultAzureCredential", importedDict);
        }
        else {
            clientParams.push({
                name: "credential",
                assignment: `const credential = {getToken: () => Promise.resolve({ token: "{Your token}", expiresOnTimestamp: 0 })};`
            });
        }
    }
    return clientParams;
}
function convertPathLevelParameters(pathDetail, path, schemaMap) {
    const pathItself = {
        name: `"${path}"`
    };
    const pathParams = (pathDetail || []).pathParameters.map((p) => {
        const pathParam = {
            name: (0, nameUtils_js_1.normalizeName)(p.name, nameUtils_js_1.NameType.Parameter)
        };
        const value = (0, valueGenerationUtil_js_1.generateParameterTypeValue)(p.type, p.name, schemaMap);
        pathParam.assignment = `const ${pathParam.name} =` + value + `;`;
        return pathParam;
    });
    return [pathItself].concat(pathParams);
}
function convertMethodLevelParameters(methods, schemaMap, operationParameter) {
    var _a, _b, _c, _d, _e;
    if (!methods ||
        methods.length === 0 ||
        !operationParameter ||
        operationParameter.parameters.length === 0) {
        return [];
    }
    const rawMethodParams = operationParameter.parameters;
    const method = methods[0];
    const requestParameter = rawMethodParams[0];
    const hasInputParams = !!rawMethodParams && rawMethodParams.length > 0, requireParam = !method.hasOptionalOptions;
    if (!hasInputParams && !requireParam) {
        return [];
    }
    const allSideAssignments = [], querySideAssignments = [], headerSideAssignments = [];
    if (!!requestParameter.body &&
        requestParameter.body &&
        requestParameter.body.body &&
        ((_a = requestParameter.body.body) === null || _a === void 0 ? void 0 : _a.length) > 0) {
        const body = requestParameter.body.body[0];
        const bodyTypeName = (_b = body.typeName) !== null && _b !== void 0 ? _b : body.type;
        if (bodyTypeName !== "string" && body.oriSchema) {
            schemaMap.set(bodyTypeName, body.oriSchema);
        }
        allSideAssignments.push(` body: ` + (0, valueGenerationUtil_js_1.generateParameterTypeValue)(bodyTypeName, "body", schemaMap));
    }
    (_c = requestParameter.parameters) === null || _c === void 0 ? void 0 : _c.filter((p) => p.type === "query").forEach((p) => {
        const name = `${p.name}`;
        querySideAssignments.push(`${name}: ` + (0, valueGenerationUtil_js_1.generateParameterTypeValue)(p.param.type, name, schemaMap));
    });
    if (querySideAssignments.length > 0) {
        allSideAssignments.push(` queryParameters: { ` + querySideAssignments.join(", ") + `}`);
    }
    (_d = requestParameter.parameters) === null || _d === void 0 ? void 0 : _d.filter((p) => p.type === "header").filter((p) => p.name.toLowerCase() !== "contenttype").forEach((p) => {
        const name = `${p.name}`;
        headerSideAssignments.push(`${name}: ` + (0, valueGenerationUtil_js_1.generateParameterTypeValue)(p.param.type, name, schemaMap));
    });
    if (headerSideAssignments.length > 0) {
        allSideAssignments.push(` headers: { ` + headerSideAssignments.join(", ") + `}`);
    }
    const contentType = (_e = requestParameter.parameters) === null || _e === void 0 ? void 0 : _e.filter((p) => p.type === "header").filter((p) => p.name.toLowerCase() === "contenttype");
    if (contentType && contentType.length > 0) {
        allSideAssignments.push(` ${contentType[0].name}: ` +
            (0, valueGenerationUtil_js_1.generateParameterTypeValue)(contentType[0].param.type, contentType[0].name, schemaMap));
    }
    let value = `{}`;
    if (allSideAssignments.length > 0) {
        value = `{ ` + allSideAssignments.join(", ") + `}`;
    }
    else {
        return [];
    }
    const optionParam = {
        name: "options",
        assignment: `const options =` + value + `;`,
        value
    };
    return [optionParam];
}
function addValueInImportedDict(key, val, importedDict) {
    if (!importedDict[key]) {
        importedDict[key] = new Set();
    }
    importedDict[key].add(val);
}
function buildMethodParamMap(model) {
    var _a;
    const map = new Map();
    ((_a = model.parameters) !== null && _a !== void 0 ? _a : []).forEach((p) => {
        var _a;
        const operatonConcante = getOperationConcate(p.operationName, p.operationGroup, (_a = model.options) === null || _a === void 0 ? void 0 : _a.sourceFrom);
        map.set(operatonConcante, p);
    });
    return map;
}
function getOperationConcate(opName, opGroup, sourceFrom) {
    return sourceFrom === "Swagger"
        ? opGroup === "" || opGroup === "Client"
            ? opName
            : `${opGroup}${opName}`
        : `${opGroup}_${opName}`;
}
