// @azure-tools/typespec-ts/dist/src/contextManager.js
/**
 * Manages shared contexts across the application to minimize prop drilling and enhance modularity.
 * Implements the Singleton pattern to ensure there is a single instance of ContextManager.
 */
class ContextManager {
    constructor() {
        this.contexts = new Map();
    }
    /**
     * Retrieves the singleton instance of the ContextManager.
     * @returns {ContextManager} The singleton instance.
     */
    static getInstance() {
        if (!ContextManager.instance) {
            ContextManager.instance = new ContextManager();
        }
        return ContextManager.instance;
    }
    /**
     * Sets a value for a specific context key.
     * @param {ContextKey} key - The key of the context to set.
     * @param {Contexts[K]} value - The value to set for the specified context.
     */
    setContext(key, value) {
        this.contexts.set(key, value);
    }
    /**
     * Retrieves the context value for a specific key.
     * @param {ContextKey} key - The key of the context to retrieve.
     * @returns {Contexts[K] | undefined} The value of the context if found, otherwise undefined.
     */
    getContext(key) {
        return this.contexts.get(key);
    }
}
// Expose the singleton instance of the context manager.
export const contextManager = ContextManager.getInstance();
/**
 * A utility function to use a context by key.
 * @param {ContextKey} key - The key of the context to retrieve.
 * @throws Will throw an error if the context is not found.
 * @returns {Contexts[K]} The context value.
 */
export function useContext(key) {
    const context = contextManager.getContext(key);
    if (!context) {
        throw new Error(`Context ${key} not found`);
    }
    return context;
}
/**
 * Provides a context with a value to be accessible across the program.
 * @param {ContextKey} key - The key of the context to provide.
 * @param {Contexts[K]} value - The value to set for the specified context.
 */
export function provideContext(key, value) {
    contextManager.setContext(key, value);
}
