// @azure-tools/typespec-ts/dist/src/index.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import * as fsextra from "fs-extra";
import { AzureCoreDependencies, AzureIdentityDependencies, AzurePollingDependencies, DefaultCoreDependencies, AzureTestDependencies } from "./modular/external-dependencies.js";
import { clearDirectory } from "./utils/fileSystemUtils.js";
import { CloudSettingHelpers, CreateRecorderHelpers, MultipartHelpers, PagingHelpers, PollingHelpers, SerializationHelpers, SimplePollerHelpers, StorageCompatHelpers, UrlTemplateHelpers, XmlHelpers } from "./modular/static-helpers-metadata.js";
import { buildApiExtractorConfig, buildClient, buildClientDefinitions, buildEsLintConfig, buildIndexFile, buildIsUnexpectedHelper, buildLicenseFile, buildLogger, buildPackageFile, buildParameterTypes, buildPollingHelper, buildPaginateHelper as buildRLCPaginateHelper, buildReadmeFile, hasClientNameChanged, updateReadmeFile, buildRecordedClientFile, buildResponseTypes, buildRollupConfig, buildSampleTest, buildSamples, buildSerializeHelper, buildTopLevelIndex, buildTsConfig, buildTsSnippetsConfig, buildTestBrowserTsConfig, buildTestNodeTsConfig, buildTestMainTsConfig, buildVitestConfig, buildWarpConfig, getClientName, hasUnexpectedHelper, isAzurePackage, updatePackageFile, buildSampleEnvFile, buildSnippets, buildTsSrcConfig, buildTsSampleConfig } from "@azure-tools/rlc-common";
import { buildRootIndex, buildSubClientIndexFile } from "./modular/buildRootIndex.js";
import { emitContentByBuilder, emitModels } from "./utils/emitUtil.js";
import { provideContext, useContext } from "./contextManager.js";
import { Project } from "ts-morph";
import { buildClassicOperationFiles } from "./modular/buildClassicalOperationGroups.js";
import { buildClassicalClient } from "./modular/buildClassicalClient.js";
import { getClientContextPath, buildClientContext } from "./modular/buildClientContext.js";
import { buildApiOptions } from "./modular/emitModelsOptions.js";
import { buildOperationFiles } from "./modular/buildOperations.js";
import { buildRestorePoller } from "./modular/buildRestorePoller.js";
import { buildSubpathIndexFile } from "./modular/buildSubpathIndex.js";
import { createSdkContext, listAllServiceNamespaces } from "@azure-tools/typespec-client-generator-core";
import { transformModularEmitterOptions } from "./modular/buildModularOptions.js";
import { emitLoggerFile } from "./modular/emitLoggerFile.js";
import { emitTypes, emitNonModelResponseTypes } from "./modular/emitModels.js";
import { existsSync } from "fs";
import { getModuleExports } from "./modular/buildProjectFiles.js";
import { getClientHierarchyMap, getRLCClients, getModularClientOptions } from "./utils/clientUtils.js";
import { join } from "path";
import { loadStaticHelpers } from "./framework/load-static-helpers.js";
import { packageUsesXmlSerialization } from "./modular/serialization/buildXmlSerializerFunction.js";
import { provideBinder } from "./framework/hooks/binder.js";
import { provideSdkTypes } from "./framework/hooks/sdkTypes.js";
import { transformRLCModel } from "./transform/transform.js";
import { transformRLCOptions } from "./transform/transfromRLCOptions.js";
import { emitSamples } from "./modular/emitSamples.js";
import { emitTests } from "./modular/emitTests.js";
import { generateCrossLanguageDefinitionFile } from "./utils/crossLanguageDef.js";
import { getClassicalClientName } from "./modular/helpers/namingHelpers.js";
export * from "./lib.js";
export async function $onEmit(context) {
    var _a, _b, _c, _d;
    if (context.program.compilerOptions.noEmit || context.program.hasError()) {
        return;
    }
    /** Shared status */
    const outputProject = new Project();
    const program = context.program;
    const emitterOptions = context.options;
    const dpgContext = await createContextWithDefaultOptions(context);
    // Report any diagnostics from TCGC
    if (((_a = dpgContext.diagnostics) === null || _a === void 0 ? void 0 : _a.length) > 0) {
        program.reportDiagnostics(dpgContext.diagnostics);
    }
    // Enrich the dpg context with path detail and common options
    await enrichDpgContext();
    const rlcOptions = (_b = dpgContext.rlcOptions) !== null && _b !== void 0 ? _b : {};
    const needUnexpectedHelper = new Map();
    const serviceNameToRlcModelsMap = new Map();
    provideContext("rlcMetaTree", new Map());
    provideContext("symbolMap", new Map());
    provideContext("outputProject", outputProject);
    provideContext("emitContext", {
        compilerContext: context,
        tcgcContext: dpgContext
    });
    const staticHelpers = await loadStaticHelpers(outputProject, {
        ...SerializationHelpers,
        ...PagingHelpers,
        ...PollingHelpers,
        ...SimplePollerHelpers,
        ...UrlTemplateHelpers,
        ...MultipartHelpers,
        ...CloudSettingHelpers,
        ...XmlHelpers,
        ...(rlcOptions.generateTest ? CreateRecorderHelpers : {}),
        ...(rlcOptions.enableStorageCompat ? StorageCompatHelpers : {})
    }, {
        sourcesDir: (_c = dpgContext.generationPathDetail) === null || _c === void 0 ? void 0 : _c.modularSourcesDir,
        rootDir: (_d = dpgContext.generationPathDetail) === null || _d === void 0 ? void 0 : _d.rootDir,
        options: rlcOptions,
        program
    });
    const extraDependencies = isAzurePackage({ options: rlcOptions })
        ? {
            ...AzurePollingDependencies,
            ...AzureCoreDependencies,
            ...AzureIdentityDependencies,
            ...AzureTestDependencies
        }
        : { ...DefaultCoreDependencies };
    const binder = provideBinder(outputProject, {
        staticHelpers,
        dependencies: {
            ...extraDependencies
        }
    });
    provideSdkTypes(dpgContext);
    const rlcCodeModels = [];
    let modularEmitterOptions;
    // 1. Clear sources folder
    await clearSrcFolder();
    // 2. Generate RLC code model
    // TODO: skip this step in modular once modular generator is sufficiently decoupled
    await buildRLCCodeModels();
    // 3. Clear samples-dev folder if generateSample is true
    await clearSamplesDevFolder();
    // 4. Generate sources
    if (emitterOptions["is-modular-library"]) {
        await generateModularSources();
    }
    else {
        await generateRLCSources();
    }
    // 5. Generate metadata and test files
    function getTypespecTsVersion(context) {
        const emitterMetadata = context.program.emitters.find((emitter) => emitter.metadata.name === "@azure-tools/typespec-ts");
        return emitterMetadata === null || emitterMetadata === void 0 ? void 0 : emitterMetadata.metadata.version;
    }
    await generateMetadataAndTest(dpgContext);
    async function enrichDpgContext() {
        var _a, _b;
        const generationPathDetail = await calculateGenerationDir();
        dpgContext.generationPathDetail = generationPathDetail;
        dpgContext.allServiceNamespaces = listAllServiceNamespaces(dpgContext);
        const options = transformRLCOptions(emitterOptions, dpgContext);
        emitterOptions["is-modular-library"] = options.isModularLibrary;
        emitterOptions["generate-sample"] = options.generateSample;
        // clear output folder if needed
        if (options.clearOutputFolder) {
            // Clear output directory while preserving TempTypeSpecFiles
            await clearDirectory(context.emitterOutputDir, ["TempTypeSpecFiles"], program);
        }
        const hasTestFolder = await fsextra.pathExists(join((_b = (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.metadataDir) !== null && _b !== void 0 ? _b : "", "test"));
        options.generateTest =
            options.generateTest === true ||
                (options.generateTest === undefined &&
                    (!hasTestFolder || (options.azureSdkForJs && options.azureArm)) &&
                    isAzurePackage({ options: options }));
        dpgContext.rlcOptions = options;
    }
    async function calculateGenerationDir() {
        var _a;
        const projectRoot = (_a = context.emitterOutputDir) !== null && _a !== void 0 ? _a : "";
        const customizationFolder = join(projectRoot, "generated");
        const srcGeneratedFolder = join(projectRoot, "src", "generated");
        // if customization folder exists, use it as sources root
        const finalCustomizationFolder = (await fsextra.pathExists(srcGeneratedFolder))
            ? srcGeneratedFolder
            : customizationFolder;
        const sourcesRoot = (await fsextra.pathExists(finalCustomizationFolder))
            ? finalCustomizationFolder
            : join(projectRoot, "src");
        return {
            rootDir: projectRoot,
            metadataDir: projectRoot,
            rlcSourcesDir: sourcesRoot,
            modularSourcesDir: sourcesRoot
        };
    }
    async function clearSrcFolder() {
        var _a, _b, _c, _d;
        await fsextra.emptyDir((_d = (_b = (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.modularSourcesDir) !== null && _b !== void 0 ? _b : (_c = dpgContext.generationPathDetail) === null || _c === void 0 ? void 0 : _c.rlcSourcesDir) !== null && _d !== void 0 ? _d : "");
    }
    async function clearSamplesDevFolder() {
        var _a, _b;
        if (emitterOptions["generate-sample"] === true) {
            const samplesDevPath = join((_b = (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.rootDir) !== null && _b !== void 0 ? _b : "", "samples-dev");
            if (await fsextra.pathExists(samplesDevPath)) {
                await fsextra.emptyDir(samplesDevPath);
            }
        }
    }
    async function buildRLCCodeModels() {
        var _a, _b;
        const clients = getRLCClients(dpgContext);
        for (const client of clients) {
            const rlcModels = await transformRLCModel(client, dpgContext);
            rlcCodeModels.push(rlcModels);
            const serviceName = (_b = (_a = client.services[0]) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : "Unknown";
            serviceNameToRlcModelsMap.set(serviceName, rlcModels);
            needUnexpectedHelper.set(getClientName(rlcModels), hasUnexpectedHelper(rlcModels));
        }
    }
    async function generateRLCSources() {
        var _a;
        for (const rlcModels of rlcCodeModels) {
            await emitModels(rlcModels, program);
            await emitContentByBuilder(program, buildClientDefinitions, rlcModels);
            await emitContentByBuilder(program, buildResponseTypes, rlcModels);
            await emitContentByBuilder(program, buildClient, rlcModels);
            await emitContentByBuilder(program, buildParameterTypes, rlcModels);
            await emitContentByBuilder(program, buildIsUnexpectedHelper, rlcModels);
            await emitContentByBuilder(program, buildIndexFile, rlcModels);
            await emitContentByBuilder(program, buildLogger, rlcModels);
            await emitContentByBuilder(program, buildTopLevelIndex, rlcModels);
            await emitContentByBuilder(program, buildRLCPaginateHelper, rlcModels);
            await emitContentByBuilder(program, buildPollingHelper, rlcModels);
            await emitContentByBuilder(program, buildSerializeHelper, rlcModels);
            await emitContentByBuilder(program, buildSamples, rlcModels, (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.metadataDir);
        }
    }
    async function generateModularSources() {
        var _a, _b, _c, _d;
        const modularSourcesRoot = (_b = (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.modularSourcesDir) !== null && _b !== void 0 ? _b : "src";
        const project = useContext("outputProject");
        modularEmitterOptions = transformModularEmitterOptions(dpgContext, modularSourcesRoot, {
            casing: "camel"
        });
        emitLoggerFile(modularEmitterOptions, modularSourcesRoot);
        const rootIndexFile = project.createSourceFile(`${modularSourcesRoot}/index.ts`, "", {
            overwrite: true
        });
        emitTypes(dpgContext, { sourceRoot: modularSourcesRoot });
        emitNonModelResponseTypes(dpgContext, { sourceRoot: modularSourcesRoot });
        buildSubpathIndexFile(modularEmitterOptions, "models", undefined, {
            recursive: true
        });
        const clientMap = getClientHierarchyMap(dpgContext);
        if (clientMap.length === 0) {
            // If no clients, we still need to build the root index file
            buildRootIndex(dpgContext, modularEmitterOptions, rootIndexFile);
        }
        for (const subClient of clientMap) {
            await renameClientName(subClient[1], modularEmitterOptions);
            buildApiOptions(dpgContext, subClient, modularEmitterOptions);
            buildOperationFiles(dpgContext, subClient, modularEmitterOptions);
            buildClientContext(dpgContext, subClient, modularEmitterOptions);
            buildRestorePoller(dpgContext, subClient, modularEmitterOptions);
            if ((_c = dpgContext.rlcOptions) === null || _c === void 0 ? void 0 : _c.hierarchyClient) {
                buildSubpathIndexFile(modularEmitterOptions, "api", subClient, {
                    exportIndex: false,
                    recursive: true
                });
            }
            else {
                buildSubpathIndexFile(modularEmitterOptions, "api", subClient, {
                    recursive: true,
                    exportIndex: true
                });
            }
            buildClassicalClient(dpgContext, subClient, modularEmitterOptions);
            buildClassicOperationFiles(dpgContext, subClient, modularEmitterOptions);
            buildSubpathIndexFile(modularEmitterOptions, "classic", subClient, {
                exportIndex: true,
                interfaceOnly: true
            });
            const { subfolder } = getModularClientOptions(subClient);
            // Generate index file for clients with subfolders (multi-client scenarios and nested clients)
            if (subfolder) {
                buildSubClientIndexFile(dpgContext, subClient, modularEmitterOptions);
            }
            buildRootIndex(dpgContext, modularEmitterOptions, rootIndexFile, subClient);
        }
        // Enable modular sample generation when explicitly set to true or MPG
        if (emitterOptions["generate-sample"] === true) {
            const samples = emitSamples(dpgContext);
            // Refine the rlc sample generation logic
            // TODO: remember to remove this out when RLC is splitted from Modular
            if (samples.length > 0) {
                dpgContext.rlcOptions.generateSample = true;
            }
        }
        binder.resolveAllReferences(modularSourcesRoot, (_d = dpgContext.generationPathDetail) === null || _d === void 0 ? void 0 : _d.rootDir);
        if (program.compilerOptions.noEmit || program.hasError()) {
            return;
        }
        for (const file of project.getSourceFiles()) {
            await emitContentByBuilder(program, () => ({ content: file.getFullText(), path: file.getFilePath() }), modularEmitterOptions);
        }
    }
    function buildMetadataJson() {
        var _a;
        const apiVersions = dpgContext.sdkPackage.metadata.apiVersions;
        const emitterVersion = getTypespecTsVersion(context);
        if (apiVersions === undefined && emitterVersion === undefined) {
            return;
        }
        const content = {};
        if (apiVersions !== undefined && apiVersions.size > 0) {
            content.apiVersions = Object.fromEntries(apiVersions);
        }
        if (emitterVersion !== undefined) {
            content.emitterVersion = emitterVersion;
        }
        if ((_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.isModularLibrary) {
            content.crossLanguageDefinitions =
                generateCrossLanguageDefinitionFile(dpgContext);
        }
        return {
            path: "metadata.json",
            content: JSON.stringify(content, null, 2)
        };
    }
    async function generateMetadataAndTest(context) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        const project = useContext("outputProject");
        if (rlcCodeModels.length === 0 || !rlcCodeModels[0]) {
            return;
        }
        const rlcClient = rlcCodeModels[0];
        const option = dpgContext.rlcOptions;
        const isAzureFlavor = isAzurePackage({ options: option });
        // Generate metadata
        const existingPackageFilePath = join((_b = (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.metadataDir) !== null && _b !== void 0 ? _b : "", "package.json");
        const hasPackageFile = await existsSync(existingPackageFilePath);
        const existingReadmeFilePath = join((_d = (_c = dpgContext.generationPathDetail) === null || _c === void 0 ? void 0 : _c.metadataDir) !== null && _d !== void 0 ? _d : "", "README.md");
        const hasReadmeFile = await existsSync(existingReadmeFilePath);
        const shouldGenerateMetadata = option.generateMetadata === true || !hasPackageFile;
        const existingTestFolderPath = join((_f = (_e = dpgContext.generationPathDetail) === null || _e === void 0 ? void 0 : _e.metadataDir) !== null && _f !== void 0 ? _f : "", "test");
        const hasTestFolder = await existsSync(existingTestFolderPath);
        if (option.generateTest === undefined) {
            if (hasTestFolder) {
                option.generateTest = false;
            }
            else {
                option.generateTest = true;
            }
        }
        //TODO Need consider multi-client cases
        if (option.isModularLibrary) {
            for (const subClient of dpgContext.sdkPackage.clients) {
                rlcClient.libraryName = subClient.name;
            }
        }
        if (shouldGenerateMetadata) {
            const commonBuilders = [
                buildRollupConfig,
                buildApiExtractorConfig,
                buildReadmeFile,
                buildLicenseFile,
                buildSampleEnvFile
            ];
            if (option.generateTest) {
                commonBuilders.push((model) => buildVitestConfig(model, "node"));
                commonBuilders.push((model) => buildVitestConfig(model, "esm"));
                commonBuilders.push((model) => buildVitestConfig(model, "browser"));
                commonBuilders.push((model) => buildTestBrowserTsConfig(model));
                commonBuilders.push((model) => buildTestNodeTsConfig(model));
                commonBuilders.push((model) => buildTestMainTsConfig(model));
            }
            if (isAzureFlavor) {
                commonBuilders.push(buildEsLintConfig);
            }
            if (emitterOptions["generate-test"] === true &&
                option.azureSdkForJs === true &&
                emitterOptions["generate-metadata"] === true) {
                await emitTests(dpgContext);
            }
            let modularPackageInfo = {};
            if (option.isModularLibrary) {
                modularPackageInfo = {
                    exports: getModuleExports(context, modularEmitterOptions)
                };
                // Build dependencies
                const dependencies = {};
                if (isAzureFlavor) {
                    dependencies["@azure/core-util"] = "^1.9.2";
                }
                // Add fast-xml-parser if XML serialization is used
                if (packageUsesXmlSerialization(dpgContext.sdkPackage)) {
                    dependencies["fast-xml-parser"] = "^4.5.0";
                }
                if (isAzureFlavor) {
                    modularPackageInfo = {
                        ...modularPackageInfo,
                        dependencies,
                        clientContextPaths: getRelativeContextPaths(context, modularEmitterOptions)
                    };
                }
                else if (Object.keys(dependencies).length > 0) {
                    modularPackageInfo = {
                        ...modularPackageInfo,
                        dependencies
                    };
                }
            }
            commonBuilders.push((model) => buildPackageFile(model, modularPackageInfo));
            // Generate warp.config.yml for Azure monorepo ESM packages
            if (option.azureSdkForJs) {
                commonBuilders.push((model) => buildWarpConfig(model, modularPackageInfo));
            }
            commonBuilders.push(buildTsConfig);
            if (option.azureSdkForJs) {
                commonBuilders.push(buildTsSrcConfig);
                if (option.generateSample) {
                    commonBuilders.push(buildTsSampleConfig);
                }
            }
            // TODO: need support snippets generation for multi-client cases. https://github.com/Azure/autorest.typescript/issues/3048
            if (option.generateTest) {
                for (const subClient of dpgContext.sdkPackage.clients) {
                    commonBuilders.push((model) => buildSnippets(model, subClient.name, option.azureSdkForJs));
                }
                commonBuilders.push(buildTsSnippetsConfig);
            }
            // build metadata relevant files
            await emitContentByBuilder(program, commonBuilders, rlcClient, (_g = dpgContext.generationPathDetail) === null || _g === void 0 ? void 0 : _g.metadataDir);
            if (option.isModularLibrary) {
                for (const file of project.getSourceFiles()) {
                    await emitContentByBuilder(program, () => ({ content: file.getFullText(), path: file.getFilePath() }), modularEmitterOptions);
                }
            }
        }
        else if (hasPackageFile) {
            const updateBuilders = [];
            let modularPackageInfo = {};
            // update existing package.json file with correct dependencies
            if (option.isModularLibrary) {
                modularPackageInfo = {
                    exports: getModuleExports(context, modularEmitterOptions),
                    clientContextPaths: getRelativeContextPaths(context, modularEmitterOptions)
                };
                updateBuilders.push((model) => updatePackageFile(model, existingPackageFilePath, modularPackageInfo));
            }
            // Update warp.config.yml for Azure monorepo packages
            if (option.azureSdkForJs) {
                updateBuilders.push((model) => buildWarpConfig(model, modularPackageInfo));
            }
            // If the client name changed, regenerate the README and snippets completely;
            // otherwise update only the API reference link in-place.
            if (hasReadmeFile) {
                const clientNameChanged = hasClientNameChanged(rlcClient, existingReadmeFilePath);
                updateBuilders.push(clientNameChanged
                    ? buildReadmeFile
                    : (model) => updateReadmeFile(model, existingReadmeFilePath));
                // Regenerate snippets.spec.ts only when the client name changed
                if (clientNameChanged && option.azureSdkForJs) {
                    for (const subClient of dpgContext.sdkPackage.clients) {
                        updateBuilders.push((model) => buildSnippets(model, getClassicalClientName(subClient), option.azureSdkForJs));
                    }
                }
            }
            // update metadata relevant files
            await emitContentByBuilder(program, updateBuilders, rlcClient, (_h = dpgContext.generationPathDetail) === null || _h === void 0 ? void 0 : _h.metadataDir);
        }
        if (isAzureFlavor) {
            await emitContentByBuilder(program, buildMetadataJson, rlcClient, (_j = dpgContext.generationPathDetail) === null || _j === void 0 ? void 0 : _j.metadataDir);
        }
        // Generate test relevant files
        if (option.generateTest && !hasTestFolder) {
            await emitContentByBuilder(program, [buildRecordedClientFile, buildSampleTest], rlcClient, (_k = dpgContext.generationPathDetail) === null || _k === void 0 ? void 0 : _k.metadataDir);
        }
    }
    function getRelativeContextPaths(context, options) {
        const clientMap = getClientHierarchyMap(context);
        return Array.from(clientMap)
            .map((subClient) => getClientContextPath(subClient, options))
            .map((path) => path.substring(path.indexOf("src")));
    }
}
export async function createContextWithDefaultOptions(context) {
    var _a, _b;
    const flattenUnionAsEnum = context.options["experimental-extensible-enums"] === undefined
        ? isArm(context)
        : context.options["experimental-extensible-enums"];
    const tcgcSettings = {
        "generate-protocol-methods": true,
        "generate-convenience-methods": true,
        emitters: [
            {
                main: "@azure-tools/typespec-ts",
                metadata: { name: "@azure-tools/typespec-ts" }
            }
        ]
    };
    context.options = {
        ...context.options,
        ...tcgcSettings
    };
    return (await createSdkContext(context, (_b = (_a = context.program.emitters[0]) === null || _a === void 0 ? void 0 : _a.metadata.name) !== null && _b !== void 0 ? _b : "@azure-tools/typespec-ts", {
        flattenUnionAsEnum
    }));
}
// TODO: should be removed once tcgc issue is resolved https://github.com/Azure/typespec-azure/issues/1794
function isArm(context) {
    var _a, _b;
    const packageName = (_b = ((_a = context === null || context === void 0 ? void 0 : context.options["package-details"]) !== null && _a !== void 0 ? _a : {})["name"]) !== null && _b !== void 0 ? _b : "";
    return packageName === null || packageName === void 0 ? void 0 : packageName.startsWith("@azure/arm-");
}
export async function renameClientName(client, emitterOptions) {
    if (emitterOptions.options.typespecTitleMap &&
        emitterOptions.options.typespecTitleMap[client.name]) {
        client.name = emitterOptions.options.typespecTitleMap[client.name];
    }
}
