// @azure-tools/typespec-ts/dist/src/modular/helpers/classicalOperationHelpers.js
import { NameType, normalizeName } from "@azure-tools/rlc-common";
import { StructureKind } from "ts-morph";
import { getClassicalLayerPrefix } from "./namingHelpers.js";
import { getOperationFunction } from "./operationHelpers.js";
import { getModularClientOptions } from "../../utils/clientUtils.js";
import { refkey } from "../../framework/refkey.js";
import { resolveReference } from "../../framework/reference.js";
import { addDeclaration } from "../../framework/declaration.js";
import { SimplePollerHelpers, PagingHelpers } from "../static-helpers-metadata.js";
import { AzurePollingDependencies } from "../external-dependencies.js";
export function getPagingLROMethodName(operationName) {
    let initialOperationName = normalizeName(operationName, NameType.Operation);
    if (initialOperationName.indexOf("list") === 0) {
        initialOperationName = initialOperationName.replace("list", "");
    }
    else if (initialOperationName.indexOf("get") === 0) {
        initialOperationName = initialOperationName.replace("get", "");
    }
    else {
        initialOperationName = normalizeName(initialOperationName, NameType.Class);
    }
    return `beginList_${initialOperationName}_andWait`;
}
export function getClassicalOperation(dpgContext, clientMap, classicFile, operationGroup, layer = operationGroup[0].length - 1) {
    var _a, _b, _c, _d;
    const prefixes = operationGroup[0];
    const operations = operationGroup[1];
    const { rlcClientName } = getModularClientOptions(clientMap);
    const hasClientContextImport = classicFile
        .getImportDeclarations()
        .filter((i) => {
        return (i.getModuleSpecifierValue() ===
            `${"../".repeat(layer + 2)}api/${normalizeName(rlcClientName, NameType.File)}.js`);
    });
    if (!hasClientContextImport || hasClientContextImport.length === 0) {
        classicFile.addImportDeclaration({
            namedImports: [rlcClientName],
            moduleSpecifier: `${"../".repeat(layer + 2)}api/${normalizeName(rlcClientName, NameType.File)}.js`
        });
    }
    const operationDeclarationMap = new Map();
    const operationDeclarations = operations.map((operation) => {
        const declaration = getOperationFunction(dpgContext, [prefixes, operation], rlcClientName);
        operationDeclarationMap.set(declaration, {
            declaration,
            oriName: operation.oriName,
            declarationRefKey: resolveReference(refkey(operation, "api")),
            isLro: declaration.isLro,
            lroFinalReturnType: declaration.lroFinalReturnType,
            isLroPaging: declaration.isLroPaging,
            lropagingFinalReturnType: declaration.lropagingFinalReturnType
        });
        return declaration;
    });
    const interfaceNamePrefix = getClassicalLayerPrefix(prefixes, NameType.Interface, "", layer);
    const interfaceName = `${interfaceNamePrefix}Operations`;
    const nextLayerInterfaceName = `${getClassicalLayerPrefix(prefixes, NameType.Interface, "", layer + 1)}Operations`;
    const existInterface = classicFile
        .getInterfaces()
        .filter((i) => i.getName() === interfaceName)[0];
    const properties = [];
    if (layer !== prefixes.length - 1) {
        const name = normalizeName((_a = (layer === prefixes.length - 1 ? prefixes[layer] : prefixes[layer + 1])) !== null && _a !== void 0 ? _a : "", NameType.Property);
        if (!properties.some((x) => x.name === name) &&
            !(existInterface && existInterface.getProperty(name))) {
            properties.push({
                kind: StructureKind.PropertySignature,
                name,
                type: resolveReference(refkey(nextLayerInterfaceName, layer + 1, "classicOperations"))
            });
        }
    }
    else {
        operationDeclarations.forEach((d) => {
            var _a, _b, _c, _d;
            const operationInfo = operationDeclarationMap.get(d);
            const paramStr = (_a = d.parameters) === null || _a === void 0 ? void 0 : _a.filter((p) => p.name !== "context").map((p) => {
                var _a;
                return p.name +
                    (((_a = p.type) === null || _a === void 0 ? void 0 : _a.toString().endsWith("operationOptions__")) ||
                        p.hasQuestionToken
                        ? "?"
                        : "") +
                    ": " +
                    p.type;
            }).join(",");
            // add the operation method signature
            properties.push({
                kind: StructureKind.PropertySignature,
                name: getClassicalMethodName(d),
                type: `(${paramStr}) => ${d.returnType}`,
                docs: d.docs
            });
            // add LRO helper methods if applicable
            if (((_b = dpgContext.rlcOptions) === null || _b === void 0 ? void 0 : _b.compatibilityLro) &&
                ((operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.isLro) || (operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.isLroPaging))) {
                const operationStateReference = resolveReference(AzurePollingDependencies.OperationState);
                const simplePollerLikeReference = resolveReference(SimplePollerHelpers.SimplePollerLike);
                const beginName = `begin_${getClassicalMethodName(d)}`;
                const beginAndWaitName = `${beginName}_andWait`;
                if (operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.isLroPaging) {
                    // LRO+Paging operation
                    const returnType = (_c = operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.lropagingFinalReturnType) !== null && _c !== void 0 ? _c : "void";
                    const pagedAsyncIterableIteratorReference = resolveReference(PagingHelpers.PagedAsyncIterableIterator);
                    const beginListAndWaitName = getPagingLROMethodName(getClassicalMethodName(d));
                    properties.push({
                        kind: StructureKind.PropertySignature,
                        name: `${normalizeName(beginListAndWaitName, NameType.Method)}`,
                        type: `(${paramStr}) => ${pagedAsyncIterableIteratorReference}<${returnType}>`,
                        docs: [`@deprecated use ${getClassicalMethodName(d)} instead`]
                    });
                }
                else {
                    // Regular LRO operation
                    const returnType = (_d = operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.lroFinalReturnType) !== null && _d !== void 0 ? _d : "void";
                    properties.push({
                        kind: StructureKind.PropertySignature,
                        name: `${normalizeName(beginName, NameType.Method)}`,
                        type: `(${paramStr}) => Promise<${simplePollerLikeReference}<${operationStateReference}<${returnType}>, ${returnType}>>`,
                        docs: [`@deprecated use ${getClassicalMethodName(d)} instead`]
                    });
                    properties.push({
                        kind: StructureKind.PropertySignature,
                        name: `${normalizeName(beginAndWaitName, NameType.Method)}`,
                        type: `(${paramStr}) => Promise<${returnType}>`,
                        docs: [`@deprecated use ${getClassicalMethodName(d)} instead`]
                    });
                }
            }
        });
    }
    if (existInterface) {
        existInterface.addProperties([...properties]);
    }
    else {
        const interfaceDeclaration = {
            kind: StructureKind.Interface,
            name: interfaceName,
            isExported: true,
            properties,
            docs: [`Interface representing a ${interfaceNamePrefix} operations.`]
        };
        addDeclaration(classicFile, interfaceDeclaration, refkey(interfaceName, layer, "classicOperations"));
    }
    const functionName = `_get${getClassicalLayerPrefix(prefixes, NameType.Interface, "", layer)}`;
    if (layer === prefixes.length - 1) {
        const functionDeclaration = {
            kind: StructureKind.Function,
            name: functionName,
            isExported: false,
            parameters: [
                {
                    name: "context",
                    type: rlcClientName
                }
            ],
            statements: `return {
        ${operationDeclarations
                .map((d) => {
                var _a, _b, _c;
                const operationInfo = operationDeclarationMap.get(d);
                const classicalParamStr = (_a = d.parameters) === null || _a === void 0 ? void 0 : _a.filter((p) => p.name !== "context").map((p) => {
                    var _a;
                    return p.name +
                        (((_a = p.type) === null || _a === void 0 ? void 0 : _a.toString().endsWith("operationOptions__")) ||
                            p.hasQuestionToken
                            ? "?"
                            : "") +
                        ": " +
                        p.type;
                }).join(",");
                const apiParamStr = [
                    "context",
                    ...[
                        (_b = d.parameters) === null || _b === void 0 ? void 0 : _b.map((p) => p.name).filter((p) => p !== "context")
                    ]
                ].join(",");
                const ret = [
                    `${getClassicalMethodName(d)}: (${classicalParamStr}) => ${operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.declarationRefKey}(${apiParamStr})`
                ];
                // add LRO helper methods if applicable
                if (((_c = dpgContext.rlcOptions) === null || _c === void 0 ? void 0 : _c.compatibilityLro) &&
                    ((operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.isLro) || (operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.isLroPaging))) {
                    const getSimplePollerReference = resolveReference(SimplePollerHelpers.getSimplePoller);
                    const beginName = `begin_${getClassicalMethodName(d)}`;
                    const beginAndWaitName = `${beginName}_andWait`;
                    const beginListAndWaitName = getPagingLROMethodName(getClassicalMethodName(d));
                    if (operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.isLroPaging) {
                        ret.push(`${normalizeName(beginListAndWaitName, NameType.Method)}: (${classicalParamStr}) => {
                    return ${operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.declarationRefKey}(${apiParamStr});
                  }`);
                    }
                    else {
                        // Regular LRO operation
                        ret.push(`${normalizeName(beginName, NameType.Method)}: async (${classicalParamStr}) => {
                    const poller = ${operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.declarationRefKey}(${apiParamStr});
                    await poller.submitted();
                    return ${getSimplePollerReference}(poller);
                  }`);
                        ret.push(`${normalizeName(beginAndWaitName, NameType.Method)}: async (${classicalParamStr}) => {
                    return await ${operationInfo === null || operationInfo === void 0 ? void 0 : operationInfo.declarationRefKey}(${apiParamStr});
                  }`);
                    }
                }
                return ret.join(",");
            })
                .join(",")}
      }`
        };
        addDeclaration(classicFile, functionDeclaration, refkey(functionName, layer, "getClassicOperation"));
    }
    const operationFunctionName = `_get${getClassicalLayerPrefix(prefixes, NameType.Interface, "", layer)}Operations`;
    const nextLayerOperationFunctionName = `_get${getClassicalLayerPrefix(prefixes, NameType.Interface, "", layer + 1)}Operations`;
    const existFunction = classicFile
        .getFunctions()
        .filter((f) => f.getName() === operationFunctionName)[0];
    if (existFunction) {
        const returnStatement = existFunction.getBodyText();
        if (returnStatement) {
            let statement = undefined;
            if (layer !== prefixes.length - 1) {
                const name = normalizeName((_b = prefixes[layer + 1]) !== null && _b !== void 0 ? _b : "FIXME", NameType.Property);
                // HACK: check if the statement includes a group of this name already to prevent an operation group appearing multiple times
                // TODO: would be good to refactor so that we have an intermediate data structure before generating the return statement
                if (!returnStatement.includes(`${name}:`)) {
                    statement = `,
          ${normalizeName((_c = prefixes[layer + 1]) !== null && _c !== void 0 ? _c : "FIXME", NameType.Property)}: ${resolveReference(refkey(nextLayerOperationFunctionName, layer + 1, "getClassicOperations"))}(context)}`;
                }
            }
            else {
                statement = `,
      ...${resolveReference(refkey(nextLayerOperationFunctionName, layer + 1, "getClassicOperations"))}(context)}`;
            }
            if (statement) {
                const newReturnStatement = returnStatement.replace(/}$/, statement);
                existFunction.setBodyText(newReturnStatement);
            }
        }
    }
    else {
        const functions = {
            kind: StructureKind.Function,
            name: operationFunctionName,
            isExported: true,
            parameters: [
                {
                    name: "context",
                    type: rlcClientName
                }
            ],
            returnType: resolveReference(refkey(interfaceName, layer, "classicOperations")),
            statements: layer !== prefixes.length - 1
                ? `return {
            ${normalizeName((_d = prefixes[layer + 1]) !== null && _d !== void 0 ? _d : "FIXME", NameType.Property)}: ${resolveReference(refkey(nextLayerOperationFunctionName, layer + 1, "getClassicOperations"))}(context)   
      }`
                : `return {
        ...${resolveReference(refkey(functionName, layer, "getClassicOperation"))}(context)
      }`
        };
        addDeclaration(classicFile, functions, refkey(operationFunctionName, layer, "getClassicOperations"));
    }
    function getClassicalMethodName(declaration) {
        var _a, _b, _c, _d;
        return normalizeName((_d = (_c = (_b = (_a = operationDeclarationMap.get(declaration)) === null || _a === void 0 ? void 0 : _a.oriName) !== null && _b !== void 0 ? _b : declaration.propertyName) !== null && _c !== void 0 ? _c : declaration.name) !== null && _d !== void 0 ? _d : "FIXME", NameType.Method);
    }
}
