// @azure-tools/typespec-ts/dist/src/modular/helpers/namingHelpers.js
import { NameType, normalizeName, ReservedModelNames } from "@azure-tools/rlc-common";
export function getClientName(client) {
    return client.name.replace(/Client$/, "");
}
export function getClassicalClientName(client) {
    return client.name;
}
export function getOperationName(operation) {
    const norm = normalizeName(operation.name, NameType.Method, true);
    if (isReservedName(operation.name, NameType.Method)) {
        return {
            name: norm,
            fixme: [
                `${operation.name} is a reserved word that cannot be used as an operation name. 
        Please add @clientName("clientName") or @clientName("<JS-Specific-Name>", "javascript") 
        to the operation to override the generated name.`
            ]
        };
    }
    return {
        name: norm
    };
}
export function isReservedName(name, nameType) {
    return ReservedModelNames.some((reservedName) => reservedName.name === name.toLowerCase() &&
        reservedName.reservedFor.includes(nameType));
}
export function getClassicalLayerPrefix(prefixes, nameType, separator = "", layer = prefixes.length - 1) {
    var _a, _b;
    const prefix = [];
    if (layer < 0) {
        return prefix.join(separator);
    }
    if (layer === 0) {
        return normalizeName((_a = prefixes[0]) !== null && _a !== void 0 ? _a : "", nameType);
    }
    for (let i = 0; i <= layer; i++) {
        prefix.push(normalizeName((_b = prefixes[i]) !== null && _b !== void 0 ? _b : "", nameType));
    }
    return prefix.join(separator);
}
export function isDefined(thing) {
    return typeof thing !== "undefined" && thing !== null;
}
/**
 * Generates a locally unique name within a set of existing names.
 * @param name - The base name.
 * @param existingNames - A set of names already in use.
 * @returns A unique name not present in the existing names set.
 */
export function generateLocallyUniqueName(name, existingNames) {
    let uniqueName = name;
    let counter = 1;
    while (existingNames.has(uniqueName)) {
        uniqueName = `${name}_${counter}`;
        counter++;
    }
    return uniqueName;
}
