// @azure-tools/typespec-ts/dist/src/modular/helpers/typeHelpers.js
import { NameType, normalizeName } from "@azure-tools/rlc-common";
import { getPropertyWithOverrides } from "../serialization/serializeUtils.js";
import { getAllAncestors, getAllProperties } from "./operationHelpers.js";
export function getDirectSubtypes(type) {
    if (!type.discriminatedSubtypes) {
        return [];
    }
    // Filter out subtypes that extend from other discriminated subtypes (hierarchical inheritance)
    return Object.values(type.discriminatedSubtypes).filter((subtype) => subtype.baseModel === type);
}
export function getAdditionalPropertiesType(type) {
    if (!type ||
        !("additionalProperties" in type ||
            (type.kind === "model" && getAdditionalPropertiesType(type.baseModel)))) {
        return undefined;
    }
    if (type.additionalProperties) {
        return type.additionalProperties;
    }
    if (type.baseModel) {
        return getAdditionalPropertiesType(type.baseModel);
    }
    return undefined;
}
export function isTypeNullable(type) {
    return Boolean(type.kind === "nullable");
}
export function getNullableValidType(type) {
    if (type.kind === "nullable") {
        return type.type;
    }
    return type;
}
const NumericTypeKinds = [
    "numeric",
    "integer",
    "safeint",
    "int8",
    "int16",
    "int32",
    "int64",
    "uint8",
    "uint16",
    "uint32",
    "uint64",
    "float",
    "float32",
    "float64",
    "decimal",
    "decimal128"
];
const DateTimeTypeKinds = ["plainTime"];
// This may be a good candidate to move to TCGC
export function isNumericTypeKind(kind) {
    return NumericTypeKinds.includes(kind);
}
export function isDateTimeTypeKind(kind) {
    return DateTimeTypeKinds.includes(kind);
}
export function isCredentialType(type) {
    return type.kind === "credential";
}
/**
 * Builds a property name mapper between the serializedName and the name of the property.
 * Return empty map if the type is not a model.
 */
export function buildPropertyNameMapper(context, model, overrides) {
    var _a;
    const mapper = new Map();
    if (model.kind !== "model") {
        return mapper;
    }
    const allProperties = getAllProperties(context, model, getAllAncestors(model));
    for (const p of allProperties) {
        if (p.kind !== "property") {
            continue;
        }
        const prop = getPropertyWithOverrides(p, overrides);
        mapper.set(((_a = prop.serializationOptions.json) === null || _a === void 0 ? void 0 : _a.name) || prop.name, normalizeName(prop.name, NameType.Property));
    }
    return mapper;
}
/**
 * Checks if the body parameter is a spread parameter.
 * @param body
 * @returns
 */
export function isSpreadBodyParameter(body) {
    var _a, _b;
    const methodParams = body.methodParameterSegments;
    return (methodParams.length > 1 ||
        (methodParams.length === 1 &&
            ((_a = methodParams[0]) === null || _a === void 0 ? void 0 : _a.length) === 1 &&
            ((_b = methodParams[0][0]) === null || _b === void 0 ? void 0 : _b.type) !== body.type));
}
