// @azure-tools/rlc-common/dist-esm/helpers/nameConstructors.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { NameType, normalizeName } from "./nameUtils.js";
export function getResponseTypeName(baseNameOrOperationGroup, operationName, statusCode) {
    if (operationName) {
        baseNameOrOperationGroup = getResponseBaseName(baseNameOrOperationGroup, operationName, statusCode || "");
    }
    return normalizeName(`${baseNameOrOperationGroup}Response`, NameType.Interface);
}
export function getLroLogicalResponseName(operationGroup, operationName) {
    return normalizeName(`${operationGroup}_${operationName}_Logical_Response`, NameType.Interface);
}
/**
 * The prefix of all response types
 * @param operationGroup operation group name e.g string_PutEmpty
 * @param operationName operation name D e.g string_PutEmpty
 * @param statusCode 2XX, 4XX, 5XX, default etc.
 * @returns normolized base name e.g StringPutEmpty200
 */
export function getResponseBaseName(operationGroup, operationName, statusCode) {
    return normalizeName(`${operationGroup}_${normalizeName(operationName, NameType.Interface, true)}_${statusCode}`, NameType.Interface);
}
/**
 * The prefix of all parameter relevant types
 * @param operationName is composed with operationGroup and operationID e.g string_PutEmpty
 * @returns
 */
export function getParameterBaseName(operationGroup, operationName) {
    return normalizeName(`${operationGroup}_${operationName}`, NameType.Interface);
}
export function getParameterTypeName(baseNameOrOperationGroup, operationName) {
    if (operationName) {
        baseNameOrOperationGroup = getParameterBaseName(baseNameOrOperationGroup, operationName);
    }
    return normalizeName(`${baseNameOrOperationGroup}_Parameters`, NameType.Interface);
}
/**
 * This is a helper function that gets the right import module depending on the type of
 * library being generated
 */
export function getImportModuleName(name, codeModel) {
    if (codeModel.options?.moduleKind === "cjs") {
        return name.cjsName;
    }
    return name.esModulesName;
}
export function getClientName(model) {
    const clientName = model.libraryName;
    const clientInterfaceName = model.options?.isModularLibrary
        ? model.libraryName
        : clientName.endsWith("Client")
            ? `${clientName}`
            : `${clientName}Client`;
    return clientInterfaceName;
}
export function getMultipartPartTypeName(schemaName, partName) {
    const name = normalizeName(partName, NameType.Interface);
    const bodyParamName = normalizeName(schemaName, NameType.Interface);
    return normalizeName(`${bodyParamName}_${name}_PartDescriptor`, NameType.Interface);
}
