// @azure-tools/rlc-common/dist-esm/helpers/operationHelpers.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { SchemaContext } from "../interfaces.js";
import { NameType, normalizeName, pascalCase } from "./nameUtils.js";
import { isObjectSchema } from "./schemaHelpers.js";
export function buildMethodDefinitions(methods, pathParams = []) {
    const methodDefinitions = [];
    for (const key of Object.keys(methods)) {
        const verbMethods = methods[key];
        for (const method of verbMethods) {
            const description = method.description;
            const areAllOptional = method.hasOptionalOptions;
            methodDefinitions.push({
                name: key,
                ...(description && { docs: [{ description }] }),
                parameters: [
                    ...getPathParamDefinitions(pathParams),
                    {
                        name: "options",
                        hasQuestionToken: areAllOptional,
                        type: pascalCase(method.optionsName)
                    }
                ],
                returnType: `StreamableMethod<${method.returnType}>`
            });
        }
    }
    return methodDefinitions;
}
export function getPathParamDefinitions(pathParams) {
    return pathParams.map(({ name, type, description }) => {
        return {
            name: normalizeName(name, NameType.Parameter),
            type,
            description
        };
    });
}
export function hasPagingOperations(model) {
    return Boolean(model.helperDetails?.hasPaging);
}
export function hasPollingOperations(model) {
    return Boolean(model.helperDetails?.hasLongRunning);
}
export function hasMultiCollection(model) {
    return Boolean(model.helperDetails?.hasMultiCollection);
}
export function hasPipeCollection(model) {
    return Boolean(model.helperDetails?.hasPipeCollection);
}
export function hasSsvCollection(model) {
    return Boolean(model.helperDetails?.hasSsvCollection);
}
export function hasTsvCollection(model) {
    return Boolean(model.helperDetails?.hasTsvCollection);
}
export function hasCsvCollection(model) {
    return Boolean(model.helperDetails?.hasCsvCollection);
}
export function hasUnexpectedHelper(model) {
    const pathDictionary = model.paths;
    for (const details of Object.values(pathDictionary)) {
        for (const methodDetails of Object.values(details.methods)) {
            const successTypes = methodDetails[0].responseTypes.success;
            const errorTypes = methodDetails[0].responseTypes.error;
            if (successTypes.length > 0 && errorTypes.length > 0 && !!errorTypes[0]) {
                return true;
            }
        }
    }
    return false;
}
export function hasInputModels(model) {
    return hasSchemaContextObject(model, [SchemaContext.Input]);
}
export function hasOutputModels(model) {
    return hasSchemaContextObject(model, [
        SchemaContext.Output,
        SchemaContext.Exception
    ]);
}
function hasSchemaContextObject(model, schemaUsage) {
    const objectSchemas = (model.schemas ?? []).filter((o) => isObjectSchema(o) &&
        o.usage?.some((u) => schemaUsage.includes(u)));
    return objectSchemas.length > 0;
}
export function getGeneratedWrapperTypes(params) {
    const wrapperTypes = params
        .map((qp) => isParameterMetadata(qp) ? qp.param.wrapperType : qp.wrapperType)
        .filter((v) => v !== undefined);
    const wrapperFromObjects = wrapperTypes.filter((wrap) => wrap.type === "object");
    const wrapperFromUnions = wrapperTypes
        .filter((wrap) => wrap.type === "union")
        .flatMap((wrapperType) => wrapperType?.enum ?? [])
        .filter((v) => v.type === "object");
    return [...wrapperFromUnions, ...wrapperFromObjects];
}
function isParameterMetadata(param) {
    return param.param !== undefined;
}
