// @azure-tools/rlc-common/dist-esm/helpers/shortcutMethods.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { buildMethodDefinitions } from "./operationHelpers.js";
import { NameType, normalizeName } from "./nameUtils.js";
export function generateMethodShortcuts(paths) {
    const keys = {};
    for (const path in paths) {
        const groupName = paths[path].operationGroupName;
        const definitions = buildOperationDefinitions(paths[path]);
        if (!keys[groupName]) {
            keys[groupName] = definitions;
        }
        else {
            keys[groupName] = [...keys[groupName], ...definitions];
        }
    }
    const interfaces = [];
    for (const interfaceName in keys) {
        const methods = keys[interfaceName];
        interfaces.push({
            name: `${interfaceName}Operations`,
            methods: methods,
            isExported: true,
            docs: [`Contains operations for ${interfaceName} operations`]
        });
    }
    return interfaces;
}
function buildOperationDefinitions(path) {
    let ops = [];
    for (const verb in path.methods) {
        const methods = path.methods[verb];
        for (const method of methods) {
            const name = normalizeName(method.operationName, NameType.Property);
            const pathParams = path.pathParameters;
            const methodDefinitions = buildMethodDefinitions({ [name]: [method] }, pathParams);
            ops = [...ops, ...methodDefinitions];
        }
    }
    return ops;
}
