// @azure-tools/rlc-common/dist/static/serializeHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildCsvCollectionContent = exports.buildTsvCollectionContent = exports.buildSsvCollectionContent = exports.buildPipeCollectionContent = exports.buildMultiCollectionContent = void 0;
exports.buildMultiCollectionContent = `
export function buildMultiCollection(
  items: string[],
  parameterName: string
): string {
  return items
    .map((item, index) => {
      if (index === 0) {
        return item;
      }
      return \`\${parameterName}=\${item}\`;
    })
    .join("&");
}`;
exports.buildPipeCollectionContent = `
export function buildPipeCollection(items: string[] | number[]): string {
  return items.join("|");
}`;
exports.buildSsvCollectionContent = `
export function buildSsvCollection(items: string[] | number[]): string {
  return items.join(" ");
}`;
exports.buildTsvCollectionContent = `
export function buildTsvCollection(items: string[] | number[]): string {
  return items.join("\\t");
}`;
exports.buildCsvCollectionContent = `
export function buildCsvCollection(items: string[] | number[]): string {
  return items.join(",");
}`;
