// @azure-tools/rlc-common/dist/metadata/packageJson/buildFlavorlessPackage.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildFlavorlessPackage = buildFlavorlessPackage;
const packageCommon_js_1 = require("./packageCommon.js");
/**
 * Builds the package.json for a flavorless package.
 */
function buildFlavorlessPackage(config) {
    const packageInfo = {
        ...getFlavorlessPackageInfo(config),
        scripts: getFlavorlessScripts(config),
        devDependencies: {
            ...(0, packageCommon_js_1.getCommonPackageDevDependencies)(config),
            "@microsoft/api-extractor": "^7.40.3",
            rimraf: "^5.0.5",
            mkdirp: "^3.0.1"
        },
        dependencies: {
            ...packageCommon_js_1.commonPackageDependencies,
            "@typespec/ts-http-runtime": "0.1.0"
        }
    };
    return packageInfo;
}
function getFlavorlessPackageInfo(config) {
    const commonPackageInfo = (0, packageCommon_js_1.getPackageCommonInfo)(config);
    return {
        ...commonPackageInfo
    };
}
function getFlavorlessScripts(config) {
    return {
        ...(0, packageCommon_js_1.getCommonPackageScripts)(),
        ...getCjsScripts(config),
        ...getEsmScripts(config)
    };
}
function getCjsScripts({ moduleKind }) {
    if (moduleKind !== "cjs") {
        return {};
    }
    return {
        build: "npm run clean && tsc && npm run extract-api"
    };
}
function getEsmScripts({ moduleKind }) {
    if (moduleKind !== "esm") {
        return {};
    }
    return {
        build: "npm run clean && tshy && npm run extract-api"
    };
}
