// @azure-tools/typespec-ts/dist/src/transform/transfromRLCOptions.js
import { pascalCase, NameType, normalizeName } from "@azure-tools/rlc-common";
import { getHttpOperationWithCache } from "@azure-tools/typespec-client-generator-core";
import { getDoc, NoTarget } from "@typespec/compiler";
import { getAuthentication } from "@typespec/http";
import { reportDiagnostic } from "../lib.js";
import { getRLCClients, listOperationsUnderRLCClient } from "../utils/clientUtils.js";
import { getDefaultService } from "../utils/modelUtils.js";
import { detectModelConflicts } from "../utils/namespaceUtils.js";
import { getOperationName } from "../utils/operationUtil.js";
import { getSupportedHttpAuth } from "../utils/credentialUtils.js";
import { getClientParameters } from "../modular/helpers/clientHelpers.js";
export function transformRLCOptions(emitterOptions, dpgContext) {
    var _a, _b;
    // Extract the options from emitter option
    const options = extractRLCOptions(dpgContext, emitterOptions, (_b = (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.rootDir) !== null && _b !== void 0 ? _b : "");
    const batch = getRLCClients(dpgContext, options.isModularLibrary);
    options.batch = batch;
    return options;
}
function extractRLCOptions(dpgContext, emitterOptions, generationRootDir) {
    var _a, _b;
    const program = dpgContext.program;
    // Compute isModularLibrary early - defaults to true unless explicitly set to false
    const isModularLibrary = emitterOptions["is-modular-library"] !== false;
    const includeShortcuts = getIncludeShortcuts(emitterOptions);
    const packageDetails = getPackageDetails(program, emitterOptions, isModularLibrary);
    const flavor = getFlavor(emitterOptions, packageDetails);
    const moduleKind = getModuleKind(emitterOptions);
    const serviceInfo = getServiceInfo(program, isModularLibrary);
    const includeHeadersInResponse = emitterOptions["include-headers-in-response"] === true;
    const azureSdkForJs = getAzureSdkForJs(emitterOptions, flavor);
    const generateMetadata = getGenerateMetadata(emitterOptions);
    const generateTest = getGenerateTest(emitterOptions, flavor);
    const generateSample = getGenerateSample(dpgContext, emitterOptions);
    const credentialInfo = getCredentialInfo(program, emitterOptions, isModularLibrary);
    const azureOutputDirectory = getAzureOutputDirectory(generationRootDir);
    const enableOperationGroup = getEnableOperationGroup(dpgContext, emitterOptions, isModularLibrary);
    const enableModelNamespace = getEnableModelNamespace(dpgContext, emitterOptions);
    const hierarchyClient = getHierarchyClient(emitterOptions);
    const clearOutputFolder = getClearOutputFolder(emitterOptions);
    const multiClient = emitterOptions["multi-client"];
    const isTypeSpecTest = emitterOptions["is-typespec-test"];
    const title = emitterOptions.title;
    const dependencyInfo = emitterOptions["dependency-info"];
    const productDocLink = emitterOptions["product-doc-link"];
    const compatibilityMode = emitterOptions["compatibility-mode"];
    const compatibilityLro = emitterOptions["compatibility-lro"];
    const experimentalExtensibleEnums = emitterOptions["experimental-extensible-enums"];
    const ignorePropertyNameNormalize = emitterOptions["ignore-property-name-normalize"];
    const ignoreEnumMemberNameNormalize = emitterOptions["ignore-enum-member-name-normalize"];
    const compatibilityQueryMultiFormat = emitterOptions["compatibility-query-multi-format"];
    const enableStorageCompat = emitterOptions["enable-storage-compat"] === true;
    const treatUnknownAsRecord = emitterOptions["treat-unknown-as-record"] === true;
    const typespecTitleMap = emitterOptions["typespec-title-map"];
    const hasSubscriptionId = getSubscriptionId(dpgContext);
    const ignoreNullableOnOptional = getIgnoreNullableOnOptional(emitterOptions, flavor);
    const wrapNonModelReturn = getWrapNonModelReturn(emitterOptions, flavor);
    const isMultiService = ((_b = (_a = dpgContext.allServiceNamespaces) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) > 1;
    return {
        ...credentialInfo,
        includeHeadersInResponse,
        flavor,
        moduleKind,
        includeShortcuts,
        packageDetails,
        generateMetadata,
        generateTest,
        generateSample,
        azureSdkForJs,
        serviceInfo,
        azureOutputDirectory,
        sourceFrom: "TypeSpec",
        enableOperationGroup,
        enableModelNamespace,
        hierarchyClient,
        azureArm: dpgContext.arm,
        clearOutputFolder,
        multiClient,
        isTypeSpecTest,
        title,
        dependencyInfo,
        productDocLink,
        isModularLibrary,
        compatibilityMode,
        compatibilityLro,
        experimentalExtensibleEnums,
        ignorePropertyNameNormalize,
        compatibilityQueryMultiFormat,
        typespecTitleMap,
        ignoreEnumMemberNameNormalize,
        hasSubscriptionId,
        ignoreNullableOnOptional,
        wrapNonModelReturn,
        isMultiService,
        enableStorageCompat,
        treatUnknownAsRecord
    };
}
function processAuth(program, isModularLibrary) {
    var _a;
    const serviceNs = (_a = getDefaultService(program, isModularLibrary)) === null || _a === void 0 ? void 0 : _a.type;
    if (!serviceNs) {
        return undefined;
    }
    const authorization = getAuthentication(program, serviceNs);
    if (!authorization || !authorization.options) {
        return undefined;
    }
    const securityInfo = {};
    for (const auth of getSupportedHttpAuth(program, authorization)) {
        switch (auth.type) {
            case "http":
                securityInfo.addCredentials = true;
                securityInfo.customHttpAuthHeaderName = "Authorization";
                // If it is basic or bearer auth we should generate it as Basic or Bearer
                securityInfo.customHttpAuthSharedKeyPrefix = [
                    "basic",
                    "bearer"
                ].includes(auth.scheme.toLowerCase())
                    ? pascalCase(auth.scheme)
                    : auth.scheme;
                break;
            case "apiKey":
                if (auth.in === "cookie") {
                    return undefined;
                }
                securityInfo.addCredentials = true;
                securityInfo.credentialKeyHeaderName = auth.name;
                break;
            case "oauth2": {
                const flow = auth.flows[0];
                if (flow === undefined || !flow.scopes) {
                    return undefined;
                }
                securityInfo.addCredentials = true;
                if (!securityInfo.credentialScopes) {
                    securityInfo.credentialScopes = [];
                }
                if (flow.scopes.length === 0) {
                    reportDiagnostic(program, {
                        code: "no-credential-scopes",
                        target: NoTarget
                    });
                }
                // ignore the user_impersonation scope
                if (flow.scopes.length === 1 &&
                    flow.scopes[0] &&
                    flow.scopes[0].value.toLowerCase() === "user_impersonation") {
                    return securityInfo;
                }
                securityInfo.credentialScopes.push(...flow.scopes.map((item) => {
                    return item.value;
                }));
                break;
            }
            default:
                break;
        }
    }
    return securityInfo;
}
function getEnableOperationGroup(dpgContext, emitterOptions, isModularLibrary) {
    if (emitterOptions["enable-operation-group"] === true ||
        emitterOptions["enable-operation-group"] === false) {
        return emitterOptions["enable-operation-group"];
    }
    // Only detect if existing name conflicts if customers don't set hierarchyClient to true
    return detectIfNameConflicts(dpgContext, isModularLibrary);
}
function getEnableModelNamespace(dpgContext, emitterOptions) {
    if (emitterOptions["enable-model-namespace"] === true ||
        emitterOptions["enable-model-namespace"] === false) {
        return emitterOptions["enable-model-namespace"];
    }
    // Detect if existing name conflicts if customers didn't set the option explicitly
    return detectModelConflicts(dpgContext);
}
function getHierarchyClient(emitterOptions) {
    if (emitterOptions["hierarchy-client"] === true ||
        emitterOptions["hierarchy-client"] === false) {
        return emitterOptions["hierarchy-client"];
    }
    // enable hierarchy client by default if customers didn't set the option explicitly
    return true;
}
function getClearOutputFolder(emitterOptions) {
    return emitterOptions["clear-output-folder"] ? true : false;
}
function detectIfNameConflicts(dpgContext, isModularLibrary) {
    const clients = getRLCClients(dpgContext, isModularLibrary);
    for (const client of clients) {
        // only consider it's conflict when there are conflicts in the same client
        const nameSet = new Set();
        for (const op of listOperationsUnderRLCClient(client)) {
            const route = getHttpOperationWithCache(dpgContext, op);
            const name = getOperationName(dpgContext, route.operation);
            if (nameSet.has(name)) {
                return true;
            }
            else {
                nameSet.add(name);
            }
        }
    }
    // No conflicts if we didn't detect any
    return false;
}
function getIgnoreNullableOnOptional(emitterOptions, flavor) {
    // If explicitly set in options, use that value
    if (emitterOptions["ignore-nullable-on-optional"] !== undefined) {
        return Boolean(emitterOptions["ignore-nullable-on-optional"]);
    }
    // Default to true for Azure services (same as HLC behavior)
    return flavor === "azure";
}
function getWrapNonModelReturn(emitterOptions, flavor) {
    // If explicitly set in options, use that value
    if (emitterOptions["wrap-non-model-return"] !== undefined) {
        return Boolean(emitterOptions["wrap-non-model-return"]);
    }
    // Default to true for Azure services to maintain HLC backward compatibility
    return flavor === "azure";
}
function getIncludeShortcuts(emitterOptions) {
    return Boolean(emitterOptions["include-shortcuts"]);
}
function getModuleKind(emitterOptions) {
    var _a;
    return (_a = emitterOptions["module-kind"]) !== null && _a !== void 0 ? _a : "esm";
}
function getFlavor(emitterOptions, packageDetails) {
    const flavor = emitterOptions.flavor;
    if (flavor !== undefined) {
        if (flavor.toLowerCase() === "azure") {
            return "azure";
        }
        else {
            return undefined;
        }
    }
    const branded = emitterOptions.branded;
    if (branded !== undefined) {
        return branded ? "azure" : undefined;
    }
    const scopeName = packageDetails === null || packageDetails === void 0 ? void 0 : packageDetails.scopeName;
    if (scopeName !== undefined &&
        (scopeName.startsWith("azure") || scopeName.startsWith("msinternal"))) {
        return "azure";
    }
    else {
        return undefined;
    }
}
function buildPackageDetails(program, emitterOptions, isModularLibrary) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    const defaultDetail = {
        name: "@msinternal/unamedpackage",
        nameWithoutScope: "unamedpackage",
        version: "1.0.0-beta.1"
    };
    const isVersionUserProvided = Boolean((_a = emitterOptions["package-details"]) === null || _a === void 0 ? void 0 : _a.version);
    const packageDetails = {
        ...emitterOptions["package-details"],
        name: (_c = (_b = emitterOptions["package-details"]) === null || _b === void 0 ? void 0 : _b.name) !== null && _c !== void 0 ? _c : normalizeName((_f = (_d = emitterOptions === null || emitterOptions === void 0 ? void 0 : emitterOptions.title) !== null && _d !== void 0 ? _d : (_e = getDefaultService(program, isModularLibrary)) === null || _e === void 0 ? void 0 : _e.title) !== null && _f !== void 0 ? _f : "", NameType.Class),
        version: (_h = (_g = emitterOptions["package-details"]) === null || _g === void 0 ? void 0 : _g.version) !== null && _h !== void 0 ? _h : "1.0.0-beta.1",
        isVersionUserProvided
    };
    if ((_j = emitterOptions["package-details"]) === null || _j === void 0 ? void 0 : _j.name) {
        const nameParts = (_k = emitterOptions["package-details"]) === null || _k === void 0 ? void 0 : _k.name.split("/");
        if (nameParts.length === 2) {
            packageDetails.nameWithoutScope = nameParts[1];
            packageDetails.scopeName = (_l = nameParts[0]) === null || _l === void 0 ? void 0 : _l.replace("@", "");
        }
    }
    return packageDetails !== null && packageDetails !== void 0 ? packageDetails : defaultDetail;
}
function getPackageDetails(program, emitterOptions, isModularLibrary) {
    return buildPackageDetails(program, emitterOptions, isModularLibrary);
}
function getServiceInfo(program, isModularLibrary) {
    const defaultService = getDefaultService(program, isModularLibrary);
    return {
        title: defaultService === null || defaultService === void 0 ? void 0 : defaultService.title,
        description: defaultService && getDoc(program, defaultService.type)
    };
}
function getAzureSdkForJs(emitterOptions, flavor) {
    return flavor !== "azure"
        ? false
        : emitterOptions["azure-sdk-for-js"] === undefined ||
            emitterOptions["azure-sdk-for-js"] === null
            ? true
            : Boolean(emitterOptions["azure-sdk-for-js"]);
}
function getGenerateMetadata(emitterOptions) {
    if (emitterOptions["generate-metadata"] === undefined ||
        emitterOptions["generate-metadata"] === null) {
        return undefined;
    }
    return Boolean(emitterOptions["generate-metadata"]);
}
/**
 * In azure scope, by default we generate test.
 * @param emitterOptions
 * @returns
 */
function getGenerateTest(emitterOptions, flavor) {
    // Disable generateTest if azureSdkForJS is false
    if (!getAzureSdkForJs(emitterOptions, flavor)) {
        return false;
    }
    return emitterOptions["generate-test"];
}
/**
 * In azure scope, by default we generate test.
 * @param emitterOptions
 * @returns
 */
function getGenerateSample(dpgContext, emitterOptions) {
    if (dpgContext.arm && emitterOptions["generate-sample"] === undefined) {
        return true;
    }
    if (emitterOptions["generate-sample"] === undefined ||
        emitterOptions["generate-sample"] === null) {
        return undefined;
    }
    return Boolean(emitterOptions["generate-sample"]);
}
export function getCredentialInfo(program, emitterOptions, isModularLibrary = true) {
    const securityInfo = processAuth(program, isModularLibrary);
    const addCredentials = emitterOptions["add-credentials"] === false
        ? false
        : securityInfo
            ? securityInfo.addCredentials
            : emitterOptions["add-credentials"];
    const credentialScopes = securityInfo && securityInfo.credentialScopes
        ? securityInfo.credentialScopes
        : emitterOptions["credential-scopes"];
    const credentialKeyHeaderName = securityInfo && securityInfo.credentialKeyHeaderName
        ? securityInfo.credentialKeyHeaderName
        : emitterOptions["credential-key-header-name"];
    const customHttpAuthHeaderName = securityInfo && securityInfo.customHttpAuthHeaderName
        ? securityInfo.customHttpAuthHeaderName
        : emitterOptions["custom-http-auth-header-name"];
    const customHttpAuthSharedKeyPrefix = securityInfo && securityInfo.customHttpAuthSharedKeyPrefix
        ? securityInfo.customHttpAuthSharedKeyPrefix
        : emitterOptions["custom-http-auth-shared-key-prefix"];
    return {
        addCredentials,
        credentialScopes,
        credentialKeyHeaderName,
        customHttpAuthHeaderName,
        customHttpAuthSharedKeyPrefix
    };
}
function getAzureOutputDirectory(emitterOutputDir) {
    const sdkFolder = emitterOutputDir;
    const sdkReletivePath = sdkFolder === null || sdkFolder === void 0 ? void 0 : sdkFolder.replace(/\/$/, "").split("/").slice(-3).join("/");
    return (sdkReletivePath === null || sdkReletivePath === void 0 ? void 0 : sdkReletivePath.substring(0, 3)) === "sdk"
        ? sdkReletivePath
        : undefined;
}
export function getSubscriptionId(dpgContext) {
    var _a;
    //TODO Need consider multi-client cases, skip multi-client cases check for now
    if ((_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.multiClient) {
        return;
    }
    for (const client of dpgContext.sdkPackage.clients) {
        if (getClientParameters(client, dpgContext)
            .map((item) => item.name)
            .includes("subscriptionId")) {
            return true;
        }
    }
    return false;
}
