// @azure-tools/typespec-ts/dist/src/transform/transform.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { buildRuntimeImports, initInternalImports, NameType, normalizeName, SchemaContext, transformSampleGroups } from "@azure-tools/rlc-common";
import { getDoc } from "@typespec/compiler";
import { getServers } from "@typespec/http";
import * as path from "path";
import { getDefaultService, getFormattedPropertyDoc, getImportedModelName, getSchemaForType, getTypeName, predictDefaultValue } from "../utils/modelUtils.js";
import { getClientLroOverload } from "../utils/operationUtil.js";
import { transformApiVersionInfo } from "./transformApiVersionInfo.js";
import { transformHelperFunctionDetails } from "./transformHelperFunctionDetails.js";
import { transformToParameterTypes } from "./transformParameters.js";
import { transformPaths } from "./transformPaths.js";
import { transformToResponseTypes } from "./transformResponses.js";
import { transformSchemas } from "./transformSchemas.js";
import { transformTelemetryInfo } from "./transformTelemetryInfo.js";
export async function transformRLCModel(client, dpgContext) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const program = dpgContext.program;
    const options = dpgContext.rlcOptions;
    const rlcSourceDir = (_a = dpgContext.generationPathDetail) === null || _a === void 0 ? void 0 : _a.rlcSourcesDir;
    const srcPath = path.join((_c = (_b = dpgContext.generationPathDetail) === null || _b === void 0 ? void 0 : _b.rlcSourcesDir) !== null && _c !== void 0 ? _c : "", options.batch && options.batch.length > 1
        ? normalizeName(client.name.replace("Client", ""), NameType.File)
        : "");
    const libraryName = normalizeName(options.batch && (options.isModularLibrary || options.batch.length > 1)
        ? client.name
        : ((_g = (_e = (_d = options === null || options === void 0 ? void 0 : options.title) !== null && _d !== void 0 ? _d : client.name) !== null && _e !== void 0 ? _e : (_f = getDefaultService(program, options.isModularLibrary)) === null || _f === void 0 ? void 0 : _f.title) !== null && _g !== void 0 ? _g : ""), NameType.Class);
    const importSet = initInternalImports();
    const urlInfo = transformUrlInfo(client, dpgContext, importSet);
    const paths = transformPaths(client, dpgContext, importSet);
    const schemas = transformSchemas(client, dpgContext);
    const responses = transformToResponseTypes(client, dpgContext, importSet);
    const parameters = transformToParameterTypes(client, dpgContext, importSet, urlInfo === null || urlInfo === void 0 ? void 0 : urlInfo.apiVersionInfo);
    const helperDetails = transformHelperFunctionDetails(client, dpgContext, options.flavor);
    // Enrich client-level annotation detail
    helperDetails.clientLroOverload = getClientLroOverload(paths);
    const telemetryOptions = transformTelemetryInfo(client, dpgContext);
    const model = {
        srcPath,
        libraryName,
        paths,
        options,
        schemas,
        responses,
        apiVersionInfo: urlInfo === null || urlInfo === void 0 ? void 0 : urlInfo.apiVersionInfo,
        parameters,
        helperDetails,
        urlInfo,
        telemetryOptions,
        importInfo: {
            internalImports: importSet,
            runtimeImports: buildRuntimeImports(options.flavor)
        },
        rlcSourceDir
    };
    model.sampleGroups = transformSampleGroups(model, (options === null || options === void 0 ? void 0 : options.generateSample) ===
        true /* Enable mock sample content if generateSample === true */);
    options.generateSample =
        (options.generateSample === true || options.generateSample === undefined) &&
            ((_h = model.sampleGroups) !== null && _h !== void 0 ? _h : []).length > 0;
    return model;
}
export function transformUrlInfo(client, dpgContext, importDetails) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const importedModels = new Set();
    const usage = [SchemaContext.Exception, SchemaContext.Input];
    const program = dpgContext.program;
    const serviceNs = (_b = getDefaultService(program, (_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.isModularLibrary)) === null || _b === void 0 ? void 0 : _b.type;
    let endpoint = undefined;
    const urlParameters = [];
    if (serviceNs) {
        const host = getServers(program, serviceNs);
        if ((_c = host === null || host === void 0 ? void 0 : host[0]) === null || _c === void 0 ? void 0 : _c.url) {
            endpoint = host[0].url;
        }
        if (host && (host === null || host === void 0 ? void 0 : host[0]) && ((_d = host === null || host === void 0 ? void 0 : host[0]) === null || _d === void 0 ? void 0 : _d.parameters)) {
            // Currently we only support one parameter in the servers definition
            for (const key of host[0].parameters.keys()) {
                const property = (_e = host === null || host === void 0 ? void 0 : host[0]) === null || _e === void 0 ? void 0 : _e.parameters.get(key);
                const type = property === null || property === void 0 ? void 0 : property.type;
                if (!property || !type) {
                    continue;
                }
                const schema = getSchemaForType(dpgContext, type, {
                    usage: usage,
                    needRef: false,
                    relevantProperty: property
                });
                (_f = getImportedModelName(schema, usage)) === null || _f === void 0 ? void 0 : _f.forEach(importedModels.add, importedModels);
                const normName = normalizeName(key, NameType.Parameter, true);
                if (normName !== key && endpoint) {
                    endpoint = endpoint.replace(`{${key}}`, `{${normName}}`);
                }
                urlParameters.push({
                    oriName: key,
                    name: normalizeName(key, NameType.Parameter, true),
                    type: getTypeName(schema, usage),
                    description: (_g = (getDoc(program, property) &&
                        getFormattedPropertyDoc(program, property, schema, " "))) !== null && _g !== void 0 ? _g : getFormattedPropertyDoc(program, type, schema, " " /* separator*/),
                    value: predictDefaultValue(dpgContext, (_h = host === null || host === void 0 ? void 0 : host[0]) === null || _h === void 0 ? void 0 : _h.parameters.get(key))
                });
            }
        }
    }
    if (importedModels.size > 0) {
        importDetails.rlcClientFactory.importsSet = importedModels;
    }
    if (endpoint && urlParameters.length > 0) {
        for (const param of urlParameters) {
            if (param.oriName === "apiVersion") {
                dpgContext.hasApiVersionInClient = true;
            }
            if (param.oriName) {
                const regexp = new RegExp(`{${param.oriName}}`, "g");
                endpoint = endpoint.replace(regexp, `{${param.name}}`);
            }
        }
    }
    // Set the default value if missing endpoint parameter
    if (endpoint == undefined && urlParameters.length === 0) {
        endpoint = "{endpointParam}";
        urlParameters.push({
            name: "endpointParam",
            type: "string"
        });
    }
    const apiVersionInfo = transformApiVersionInfo(client, dpgContext, {
        endpoint,
        urlParameters
    });
    if (apiVersionInfo &&
        urlParameters &&
        (apiVersionInfo === null || apiVersionInfo === void 0 ? void 0 : apiVersionInfo.definedPosition) === "query" &&
        !apiVersionInfo.isCrossedVersion &&
        !apiVersionInfo.defaultValue) {
        urlParameters.push({
            oriName: "api-version",
            name: "apiVersion",
            type: "string"
        });
    }
    return { endpoint, urlParameters, apiVersionInfo };
}
