// @azure-tools/typespec-ts/dist/src/transform/transformApiVersionInfo.js
import { extractDefinedPosition, extractPathApiVersion, SchemaContext } from "@azure-tools/rlc-common";
import { getHttpOperationWithCache, isApiVersion } from "@azure-tools/typespec-client-generator-core";
import { getDefaultApiVersionString, getSchemaForType, trimUsage } from "../utils/modelUtils.js";
import { listOperationsUnderRLCClient } from "../utils/clientUtils.js";
export function transformApiVersionInfo(client, dpgContext, urlInfo) {
    var _a, _b, _c;
    const queryVersionDetail = getOperationApiVersion(client, dpgContext);
    const pathVersionDetail = extractPathApiVersion(urlInfo);
    const isCrossedVersion = queryVersionDetail || pathVersionDetail
        ? Boolean((pathVersionDetail === null || pathVersionDetail === void 0 ? void 0 : pathVersionDetail.isCrossedVersion) ||
            (queryVersionDetail === null || queryVersionDetail === void 0 ? void 0 : queryVersionDetail.isCrossedVersion))
        : undefined;
    const defaultValue = (pathVersionDetail || queryVersionDetail) && !isCrossedVersion
        ? ((_b = (_a = getDefaultApiVersionString(dpgContext)) !== null && _a !== void 0 ? _a : pathVersionDetail === null || pathVersionDetail === void 0 ? void 0 : pathVersionDetail.defaultValue) !== null && _b !== void 0 ? _b : queryVersionDetail === null || queryVersionDetail === void 0 ? void 0 : queryVersionDetail.defaultValue)
        : undefined;
    if (pathVersionDetail && queryVersionDetail) {
        return pathVersionDetail;
    }
    if (pathVersionDetail && !queryVersionDetail) {
        dpgContext.hasApiVersionInClient = true;
    }
    return {
        definedPosition: extractDefinedPosition(queryVersionDetail, pathVersionDetail),
        isCrossedVersion,
        defaultValue,
        required: (_c = pathVersionDetail === null || pathVersionDetail === void 0 ? void 0 : pathVersionDetail.required) !== null && _c !== void 0 ? _c : queryVersionDetail === null || queryVersionDetail === void 0 ? void 0 : queryVersionDetail.required
    };
}
export function getOperationApiVersion(client, dpgContext) {
    var _a;
    const apiVersionTypes = new Set();
    const locations = new Set();
    const required = new Set();
    dpgContext.hasApiVersionInClient = true;
    let hasApiVersionInOperation = true;
    for (const op of listOperationsUnderRLCClient(client)) {
        hasApiVersionInOperation = false;
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        const params = route.parameters.parameters.filter((p) => (p.type === "query" || p.type === "path") &&
            isApiVersion(dpgContext, p.param));
        params.map((p) => {
            const type = getSchemaForType(dpgContext, p.param.type, {
                usage: [SchemaContext.Exception, SchemaContext.Input],
                needRef: false,
                relevantProperty: p.param
            });
            required.add(!p.param.optional);
            if (p.type === "query" || p.type === "path") {
                locations.add(p.type);
            }
            const typeString = JSON.stringify(trimUsage(type));
            apiVersionTypes.add(typeString);
        });
        if (apiVersionTypes.size > 1 || !dpgContext.hasApiVersionInClient) {
            break;
        }
        if (params.length === 1) {
            hasApiVersionInOperation = true;
        }
        if (!hasApiVersionInOperation) {
            dpgContext.hasApiVersionInClient = false;
        }
    }
    // If no api-version parameter defined return directly
    if (apiVersionTypes.size === 0 || !dpgContext.hasApiVersionInClient) {
        return;
    }
    const detail = {
        definedPosition: locations.size > 1 ? "duplicate" : locations.values().next().value,
        isCrossedVersion: apiVersionTypes.size > 1,
        defaultValue: undefined, // We won't prompt the query versions into client one
        required: required.values().next().value
    };
    return detail;
}
