// @azure-tools/typespec-ts/dist/src/transform/transformHelperFunctionDetails.js
import { getHttpOperationWithCache } from "@azure-tools/typespec-client-generator-core";
import { extractPageDetails, getSpecialSerializeInfo, hasPagingOperations, hasPollingOperations } from "../utils/operationUtil.js";
import { getCollectionFormat } from "../utils/modelUtils.js";
import { listOperationsUnderRLCClient } from "../utils/clientUtils.js";
export function transformHelperFunctionDetails(client, dpgContext, flavor) {
    const serializeInfo = extractSpecialSerializeInfo(client, dpgContext);
    // Disable paging and long running for non-Azure clients.
    if (flavor !== "azure") {
        return {
            hasLongRunning: false,
            hasPaging: false,
            ...serializeInfo
        };
    }
    const annotationDetails = {
        hasLongRunning: hasPollingOperations(client, dpgContext)
    };
    const details = extractClientPageDetails(client, dpgContext);
    return {
        ...(details !== null && details !== void 0 ? details : {}),
        ...annotationDetails,
        ...serializeInfo
    };
}
function extractClientPageDetails(client, dpgContext) {
    var _a;
    const program = dpgContext.program;
    if (!hasPagingOperations(client, dpgContext)) {
        return;
    }
    const nextLinks = new Set(["nextLink"]);
    const itemNames = new Set(["value"]);
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        const pagedDetail = extractPageDetails(program, route);
        if (pagedDetail) {
            pagedDetail.itemNames.forEach((name) => itemNames.add(name));
            pagedDetail.nextLinkNames.forEach((name) => nextLinks.add(name));
        }
    }
    // If there are more than one options for nextLink and item names we need to generate a
    // more complex pagination helper.
    return {
        hasPaging: true,
        pageDetails: {
            itemNames: [...itemNames],
            nextLinkNames: [...nextLinks],
            isComplexPaging: nextLinks.size > 1 || itemNames.size > 1
        }
    };
}
function extractSpecialSerializeInfo(client, dpgContext) {
    let hasMultiCollection = false;
    let hasCsvCollection = false;
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        route.parameters.parameters.forEach((parameter) => {
            const format = getCollectionFormat(dpgContext, parameter);
            const serializeInfo = getSpecialSerializeInfo(dpgContext, parameter.type, format);
            hasMultiCollection = hasMultiCollection
                ? hasMultiCollection
                : serializeInfo.hasMultiCollection;
            hasCsvCollection = hasCsvCollection
                ? hasCsvCollection
                : serializeInfo.hasCsvCollection;
        });
    }
    return {
        hasMultiCollection,
        hasCsvCollection
    };
}
