// @azure-tools/typespec-ts/dist/src/transform/transformResponses.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { getLroLogicalResponseName, SchemaContext } from "@azure-tools/rlc-common";
import { getHttpOperationWithCache } from "@azure-tools/typespec-client-generator-core";
import { getDoc, isVoidType } from "@typespec/compiler";
import { getBinaryType, getImportedModelName, getSchemaForType, getTypeName } from "../utils/modelUtils.js";
import { getOperationGroupName, getOperationLroOverload, getOperationName, getOperationStatuscode, isBinaryPayload, sortedOperationResponses } from "../utils/operationUtil.js";
import { listOperationsUnderRLCClient } from "../utils/clientUtils.js";
export function transformToResponseTypes(client, dpgContext, importDetails) {
    var _a;
    const rlcResponses = [];
    const inputImportedSet = new Set();
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        transformToResponseTypesForRoute(route);
    }
    if (inputImportedSet.size > 0) {
        importDetails.response.importsSet = inputImportedSet;
    }
    function transformToResponseTypesForRoute(route) {
        var _a;
        const rlcOperationUnit = {
            operationGroup: getOperationGroupName(dpgContext, route),
            operationName: getOperationName(dpgContext, route.operation),
            path: route.path,
            responses: []
        };
        const defaultSchemas = [], nonDefaultSchemas = [];
        for (const resp of sortedOperationResponses(route.responses)) {
            const statusCode = getOperationStatuscode(resp);
            const rlcResponseUnit = {
                statusCode,
                description: resp.description
            };
            // transform header
            const headers = transformHeaders(dpgContext, resp, inputImportedSet);
            // transform body
            const [body, schemas] = (_a = transformBody(dpgContext, resp, inputImportedSet)) !== null && _a !== void 0 ? _a : [undefined, []];
            rlcOperationUnit.responses.push({
                ...rlcResponseUnit,
                headers,
                body
            });
            if (statusCode === "default") {
                defaultSchemas.push(...schemas);
            }
            else {
                nonDefaultSchemas.push(...schemas);
            }
        }
        const lroLogicalResponse = transformLroLogicalResponse(dpgContext, route, getOperationGroupName(dpgContext, route), rlcOperationUnit.responses);
        if (lroLogicalResponse) {
            rlcOperationUnit.responses.push(lroLogicalResponse);
        }
        rlcResponses.push(rlcOperationUnit);
    }
    return rlcResponses;
}
/**
 * Return undefined if no valid header param
 * @param response response detail
 * @returns rlc header schema
 */
function transformHeaders(dpgContext, response, importedModels) {
    var _a, _b, _c, _d;
    if (!response.responses.length) {
        return;
    }
    const rlcHeaders = new Map();
    // Current RLC client can't represent different headers per content type.
    // So we merge headers here, and report any duplicates.
    // It may be possible in principle to not error for identically declared
    // headers.
    for (const data of response.responses) {
        const headers = (_a = data === null || data === void 0 ? void 0 : data.headers) !== null && _a !== void 0 ? _a : {};
        if ((_b = data.body) === null || _b === void 0 ? void 0 : _b.contentTypeProperty) {
            headers["content-type"] = (_c = data.body) === null || _c === void 0 ? void 0 : _c.contentTypeProperty;
        }
        if (!headers || !Object.keys(headers).length) {
            continue;
        }
        for (const [key, value] of Object.entries(headers)) {
            if (!value) {
                continue;
            }
            const typeSchema = getSchemaForType(dpgContext, value.type, {
                usage: [SchemaContext.Output],
                relevantProperty: value
            });
            const type = getTypeName(typeSchema, [SchemaContext.Output]);
            (_d = getImportedModelName(typeSchema, [SchemaContext.Output])) === null || _d === void 0 ? void 0 : _d.forEach(importedModels.add, importedModels);
            const header = {
                name: `"${key.toLowerCase()}"`,
                type,
                required: !(value === null || value === void 0 ? void 0 : value.optional),
                description: getDoc(dpgContext.program, value)
            };
            rlcHeaders.set(header.name, header);
        }
    }
    return rlcHeaders.size ? Array.from(rlcHeaders.values()) : undefined;
}
function transformBody(dpgContext, response, importedModels) {
    var _a;
    if (!response.responses.length) {
        return;
    }
    // Currently RLC response only have one header and body defined
    // So we'll union all body shapes together with "|"
    const typeSet = new Set();
    const descriptions = new Set();
    let fromCore = false;
    const schemas = [];
    for (const data of response.responses) {
        const body = data === null || data === void 0 ? void 0 : data.body;
        if (!body || isVoidType(body.type)) {
            continue;
        }
        const hasBinaryContent = body.contentTypes.some((contentType) => isBinaryPayload(dpgContext, body.type, contentType));
        if (hasBinaryContent) {
            typeSet.add(getBinaryType([SchemaContext.Output]));
            descriptions.add("Value may contain any sequence of octets");
            continue;
        }
        const bodySchema = getSchemaForType(dpgContext, body.type, {
            usage: [SchemaContext.Output]
        });
        fromCore = (_a = bodySchema.fromCore) !== null && _a !== void 0 ? _a : false;
        const bodyType = getTypeName(bodySchema);
        const importedNames = getImportedModelName(bodySchema);
        if (importedNames && !fromCore) {
            importedNames
                .filter((name) => {
                return name !== "any";
            })
                .forEach(importedModels.add, importedModels);
        }
        typeSet.add(bodyType);
        schemas.push(bodySchema);
    }
    if (!typeSet.size) {
        return;
    }
    return [
        {
            name: "body",
            type: [...typeSet].join("|"),
            description: [...descriptions].join("\n\n"),
            fromCore
        },
        schemas
    ];
}
function transformLroLogicalResponse(dpgContext, route, operationGroupName, existingResponses) {
    const operationLroOverload = getOperationLroOverload(dpgContext.program, route, undefined, existingResponses);
    if (!operationLroOverload) {
        return;
    }
    const sortedResponses = existingResponses
        .filter((r) => r.statusCode.startsWith("20"))
        .sort((r1, r2) => (r1.statusCode > r2.statusCode ? 1 : -1));
    const successResp = sortedResponses[0];
    const logicalLROResponse = {
        statusCode: "200",
        description: `The final response for long-running ${route.operation.name} operation`,
        predefinedName: getLroLogicalResponseName(operationGroupName, getOperationName(dpgContext, route.operation)),
        body: successResp === null || successResp === void 0 ? void 0 : successResp.body
    };
    return logicalLROResponse;
}
