// @azure-tools/typespec-ts/dist/src/transform/transformSchemas.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { getServers } from "@typespec/http";
import { extractMediaTypes } from "../utils/mediaTypes.js";
import { getHttpOperationWithCache } from "@azure-tools/typespec-client-generator-core";
import { getBodyType, getDefaultService, getSchemaForType, includeDerivedModel, isAzureCoreErrorType, trimUsage } from "../utils/modelUtils.js";
import { SchemaContext } from "@azure-tools/rlc-common";
import { useContext } from "../contextManager.js";
import { listOperationsUnderRLCClient } from "../utils/clientUtils.js";
export function transformSchemas(client, dpgContext) {
    var _a;
    const program = dpgContext.program;
    const metatree = useContext("rlcMetaTree");
    const schemas = new Map();
    const schemaMap = new Map();
    const usageMap = new Map();
    const requestBodySet = new Set();
    const contentTypeMap = new Map();
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        transformSchemaForRoute(route);
    }
    function transformSchemaForRoute(route) {
        var _a, _b, _c, _d;
        if (route.parameters) {
            for (const param of route.parameters.parameters) {
                getGeneratedModels(param.param, SchemaContext.Input);
            }
        }
        const bodyModel = getBodyType(route);
        if (bodyModel &&
            (bodyModel.kind === "Model" ||
                bodyModel.kind === "Union" ||
                bodyModel.kind === "Enum")) {
            requestBodySet.add(bodyModel);
            const contentTypes = extractMediaTypes((_b = (_a = route.parameters.body) === null || _a === void 0 ? void 0 : _a.contentTypes) !== null && _b !== void 0 ? _b : []);
            if (contentTypes.length > 0) {
                contentTypeMap.set(bodyModel, contentTypes);
            }
            getGeneratedModels(bodyModel, SchemaContext.Input);
        }
        for (const resp of route.responses) {
            if (isAzureCoreErrorType(dpgContext.program, resp.type)) {
                continue;
            }
            for (const resps of resp.responses) {
                const headers = resps === null || resps === void 0 ? void 0 : resps.headers;
                if (headers && Object.keys(headers).length) {
                    for (const value of Object.values(headers)) {
                        getGeneratedModels(value, SchemaContext.Output);
                    }
                }
                if ((_c = resps.body) === null || _c === void 0 ? void 0 : _c.contentTypeProperty) {
                    getGeneratedModels(resps.body.contentTypeProperty, SchemaContext.Output);
                }
                const respModel = (_d = resps === null || resps === void 0 ? void 0 : resps.body) === null || _d === void 0 ? void 0 : _d.type;
                if (!respModel) {
                    continue;
                }
                getGeneratedModels(respModel, SchemaContext.Output);
            }
        }
    }
    function transformHostParameters() {
        var _a, _b, _c, _d, _e;
        const serviceNs = (_b = getDefaultService(program, (_a = dpgContext.rlcOptions) === null || _a === void 0 ? void 0 : _a.isModularLibrary)) === null || _b === void 0 ? void 0 : _b.type;
        if (serviceNs) {
            const host = getServers(program, serviceNs);
            if (host && (host === null || host === void 0 ? void 0 : host[0]) && ((_c = host === null || host === void 0 ? void 0 : host[0]) === null || _c === void 0 ? void 0 : _c.parameters)) {
                for (const key of host[0].parameters.keys()) {
                    const type = (_e = (_d = host === null || host === void 0 ? void 0 : host[0]) === null || _d === void 0 ? void 0 : _d.parameters.get(key)) === null || _e === void 0 ? void 0 : _e.type;
                    if (!type) {
                        continue;
                    }
                    getGeneratedModels(type, SchemaContext.Input);
                }
            }
        }
    }
    transformHostParameters();
    usageMap.forEach((context, tspModel) => {
        var _a;
        const model = getSchemaForType(dpgContext, tspModel, {
            usage: context,
            isRequestBody: requestBodySet.has(tspModel),
            isParentRequestBody: false,
            mediaTypes: contentTypeMap.get(tspModel)
        });
        if (model) {
            model.usage = context;
        }
        metatree.set(tspModel, { rlcType: model });
        if (model.name === "") {
            return;
        }
        const pureModel = JSON.stringify(trimUsage(model));
        schemaMap.set(pureModel, model);
        const usageSet = new Set(((_a = schemas.get(pureModel)) !== null && _a !== void 0 ? _a : []).concat(context));
        schemas.set(pureModel, Array.from(usageSet));
    });
    function setModelMap(type, schemaContext) {
        if (usageMap.get(type)) {
            const context = usageMap.get(type);
            if (context && context.indexOf(schemaContext) === -1) {
                context.push(schemaContext);
                usageMap.set(type, context);
            }
        }
        else {
            usageMap.set(type, [schemaContext]);
        }
    }
    function getGeneratedModels(model, context) {
        var _a, _b, _c, _d, _e, _f;
        if (model.kind === "Model") {
            setModelMap(model, context);
            const indexer = model.indexer;
            if ((indexer === null || indexer === void 0 ? void 0 : indexer.value) &&
                (!usageMap.get(indexer === null || indexer === void 0 ? void 0 : indexer.value) ||
                    !((_a = usageMap.get(indexer === null || indexer === void 0 ? void 0 : indexer.value)) === null || _a === void 0 ? void 0 : _a.includes(context)))) {
                getGeneratedModels(indexer.value, context);
            }
            for (const prop of model.properties) {
                const [, propType] = prop;
                if ((_b = usageMap.get(propType.type)) === null || _b === void 0 ? void 0 : _b.includes(context)) {
                    continue;
                }
                if (propType.type.kind === "Model") {
                    if (isAzureCoreErrorType(dpgContext.program, propType.type)) {
                        continue;
                    }
                    getGeneratedModels(propType.type, context);
                }
                else if (propType.type.kind === "Union") {
                    const variants = Array.from(propType.type.variants.values());
                    for (const variant of variants) {
                        if ((variant.type.kind === "Model" ||
                            variant.type.kind === "Union" ||
                            variant.type.kind === "Enum") &&
                            (!usageMap.get(variant.type) ||
                                !((_c = usageMap.get(variant.type)) === null || _c === void 0 ? void 0 : _c.includes(context)))) {
                            getGeneratedModels(variant.type, context);
                        }
                    }
                    // build type details for named union
                    if (!propType.type.expression) {
                        setModelMap(propType.type, context);
                    }
                }
                else if (propType.type.kind === "Enum") {
                    getGeneratedModels(propType.type, context);
                }
            }
            const baseModel = model.baseModel;
            if (baseModel &&
                baseModel.kind === "Model" &&
                (!usageMap.get(baseModel) ||
                    !((_d = usageMap.get(baseModel)) === null || _d === void 0 ? void 0 : _d.includes(context)))) {
                getGeneratedModels(baseModel, context);
            }
            const derivedModels = model.derivedModels.filter((dm) => {
                return includeDerivedModel(dm);
            });
            // getSchemaOrRef on all children to push them into components.schemas
            for (const child of derivedModels) {
                if (child.kind === "Model" &&
                    (!usageMap.get(child) || !((_e = usageMap.get(child)) === null || _e === void 0 ? void 0 : _e.includes(context)))) {
                    getGeneratedModels(child, context);
                }
            }
        }
        else if (model.kind === "Union") {
            const variants = Array.from(model.variants.values());
            for (const variant of variants) {
                if ((variant.type.kind === "Model" || variant.type.kind === "Union") &&
                    (!usageMap.get(variant.type) ||
                        !((_f = usageMap.get(variant.type)) === null || _f === void 0 ? void 0 : _f.includes(context)))) {
                    getGeneratedModels(variant.type, context);
                }
            }
            // build type details for named union
            if (!model.expression) {
                setModelMap(model, context);
            }
        }
        else if (model.kind === "ModelProperty") {
            getGeneratedModels(model.type, context);
        }
        else if (model.kind === "Enum") {
            setModelMap(model, context);
        }
    }
    const allSchemas = Array.from(schemas, function (item) {
        return { ...schemaMap.get(item[0]), usage: item[1] };
    });
    return allSchemas;
}
