// @azure-tools/typespec-ts/dist/src/transform/transformPaths.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { SchemaContext, getParameterTypeName, getResponseTypeName } from "@azure-tools/rlc-common";
import { getHttpOperationWithCache, isApiVersion } from "@azure-tools/typespec-client-generator-core";
import { extractOperationLroDetail, getOperationGroupName, getOperationName, getOperationResponseTypes, getOperationStatuscode, getOperationSuccessStatus, isPagingOperation, sortedOperationResponses } from "../utils/operationUtil.js";
import { getImportedModelName, getSchemaForType, isBodyRequired } from "../utils/modelUtils.js";
import { getDoc } from "@typespec/compiler";
import { getParameterSerializationInfo } from "../utils/parameterUtils.js";
import { listOperationsUnderRLCClient } from "../utils/clientUtils.js";
export function transformPaths(client, dpgContext, importDetails) {
    var _a;
    const pathParamsImportedSet = new Set();
    const paths = {};
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        transformOperation(dpgContext, route, paths, pathParamsImportedSet);
    }
    if (pathParamsImportedSet.size > 0) {
        importDetails.rlcClientDefinition.importsSet = pathParamsImportedSet;
    }
    return paths;
}
function transformOperation(dpgContext, route, paths, importSet) {
    var _a, _b, _c, _d, _e, _f;
    const program = dpgContext.program;
    const respNames = [];
    const operationGroupName = getOperationGroupName(dpgContext, route);
    for (const resp of sortedOperationResponses(route.responses)) {
        const respName = getResponseTypeName(operationGroupName, getOperationName(dpgContext, route.operation), getOperationStatuscode(resp));
        respNames.push(respName);
    }
    const responseTypes = getOperationResponseTypes(dpgContext, route);
    const method = {
        description: (_a = getDoc(program, route.operation)) !== null && _a !== void 0 ? _a : "",
        hasOptionalOptions: !hasRequiredOptions(dpgContext, route.parameters),
        optionsName: getParameterTypeName(operationGroupName, getOperationName(dpgContext, route.operation)),
        responseTypes,
        returnType: respNames.join(" | "),
        successStatus: getOperationSuccessStatus(route),
        operationName: getOperationName(dpgContext, route.operation),
        operationHelperDetail: {
            lroDetails: extractOperationLroDetail(dpgContext, route, responseTypes, operationGroupName),
            isPaging: isPagingOperation(dpgContext, route)
        }
    };
    if (paths[route.path] !== undefined &&
        !((_b = paths[route.path]) === null || _b === void 0 ? void 0 : _b.methods[route.verb])) {
        paths[route.path].methods[route.verb] = [method];
    }
    else if ((_c = paths[route.path]) === null || _c === void 0 ? void 0 : _c.methods[route.verb]) {
        (_e = (_d = paths[route.path]) === null || _d === void 0 ? void 0 : _d.methods[route.verb]) === null || _e === void 0 ? void 0 : _e.push(method);
    }
    else {
        paths[route.path] = {
            description: (_f = getDoc(program, route.operation)) !== null && _f !== void 0 ? _f : "",
            name: escapeCoreName(getOperationName(dpgContext, route.operation) || "Client"),
            pathParameters: route.parameters.parameters
                .filter((p) => p.type === "path")
                .map((p) => {
                var _a, _b, _c;
                const schemaUsage = [SchemaContext.Input, SchemaContext.Exception];
                const options = {
                    usage: schemaUsage,
                    needRef: false,
                    relevantProperty: p.param
                };
                const schema = p.param.sourceProperty
                    ? getSchemaForType(dpgContext, (_a = p.param.sourceProperty) === null || _a === void 0 ? void 0 : _a.type, options)
                    : getSchemaForType(dpgContext, p.param.type, options);
                const importedNames = (_b = getImportedModelName(schema, schemaUsage)) !== null && _b !== void 0 ? _b : [];
                importedNames.forEach(importSet.add, importSet);
                const serializationType = getParameterSerializationInfo(dpgContext, p, schema, operationGroupName, method.operationName);
                return {
                    name: p.name,
                    type: serializationType.typeName,
                    description: (_c = getDoc(program, p.param)) !== null && _c !== void 0 ? _c : "",
                    wrapperType: serializationType.wrapperType
                };
            }),
            operationGroupName: getOperationGroupName(dpgContext, route),
            methods: {
                [route.verb]: [method]
            }
        };
    }
}
function escapeCoreName(name) {
    if (["client", "streamablemethod"].indexOf(name.toLowerCase()) > -1) {
        return "_" + name;
    }
    return name;
}
function hasRequiredOptions(dpgContext, routeParameters) {
    const isRequiredBodyParam = isBodyRequired(routeParameters);
    const containsRequiredNonBodyParam = routeParameters.parameters
        .filter((parameter) => ["query", "header"].includes(parameter.type))
        .filter((parameter) => !isApiVersion(dpgContext, parameter.param))
        .filter((parameter) => !!parameter.param)
        .some((parameter) => parameter.param.optional === false);
    return isRequiredBodyParam || containsRequiredNonBodyParam;
}
