// @azure-tools/typespec-ts/dist/src/transform/transformParameters.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { SchemaContext } from "@azure-tools/rlc-common";
import { KnownMediaType, extractMediaTypes, hasMediaType, isMediaTypeJsonMergePatch } from "../utils/mediaTypes.js";
import { getHttpOperationWithCache, isApiVersion } from "@azure-tools/typespec-client-generator-core";
import { NoTarget, isVoidType } from "@typespec/compiler";
import { getBodyType, getCollectionFormat, getFormattedPropertyDoc, getImportedModelName, getSchemaForType, getTypeName, isArrayType, isBodyRequired } from "../utils/modelUtils.js";
import { getOperationGroupName, getOperationName, getSpecialSerializeInfo } from "../utils/operationUtil.js";
import { getParameterSerializationInfo } from "../utils/parameterUtils.js";
import { reportDiagnostic } from "../lib.js";
import { listOperationsUnderRLCClient } from "../utils/clientUtils.js";
export function transformToParameterTypes(client, dpgContext, importDetails, apiVersionInfo) {
    var _a;
    const rlcParameters = [];
    const outputImportedSet = new Set();
    for (const op of listOperationsUnderRLCClient(client)) {
        const route = getHttpOperationWithCache(dpgContext, op);
        // ignore overload base operation
        if (route.overloads && ((_a = route.overloads) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            continue;
        }
        transformToParameterTypesForRoute(route);
    }
    if (outputImportedSet.size > 0) {
        importDetails.parameter.importsSet = outputImportedSet;
    }
    function transformToParameterTypesForRoute(route) {
        const parameters = route.parameters;
        const rlcParameter = {
            operationGroup: getOperationGroupName(dpgContext, route),
            operationName: getOperationName(dpgContext, route.operation),
            parameters: []
        };
        const options = {
            apiVersionInfo,
            operationGroupName: rlcParameter.operationGroup,
            operationName: rlcParameter.operationName,
            importModels: outputImportedSet
        };
        // transform query param
        const queryParams = transformQueryParameters(dpgContext, parameters, options);
        // transform path param
        const pathParams = transformPathParameters(dpgContext, parameters, options);
        // TODO: support cookie parameters, https://github.com/Azure/autorest.typescript/issues/2898
        transformCookieParameters(dpgContext, parameters);
        // transform header param including content-type
        const headerParams = transformHeaderParameters(dpgContext, parameters, options);
        // transform body
        const bodyType = getBodyType(route);
        let bodyParameter = undefined;
        if (bodyType) {
            bodyParameter = transformBodyParameters(dpgContext, parameters, headerParams, outputImportedSet, bodyType);
        }
        rlcParameter.parameters.push({
            parameters: [...queryParams, ...pathParams, ...headerParams],
            body: bodyParameter
        });
        rlcParameters.push(rlcParameter);
    }
    return rlcParameters;
}
function getParameterMetadata(dpgContext, paramType, parameter, options) {
    var _a, _b, _c;
    const program = dpgContext.program;
    const importedModels = (_a = options.importModels) !== null && _a !== void 0 ? _a : new Set();
    const schemaContext = [SchemaContext.Exception, SchemaContext.Input];
    const schema = getSchemaForType(dpgContext, parameter.param.type, {
        usage: schemaContext,
        needRef: false,
        relevantProperty: parameter.param
    });
    const name = getParameterName(parameter.name);
    let description = (_b = getFormattedPropertyDoc(program, parameter.param, schema)) !== null && _b !== void 0 ? _b : "";
    const format = getCollectionFormat(dpgContext, parameter);
    if (isArrayType(schema) && format) {
        const serializeInfo = getSpecialSerializeInfo(dpgContext, parameter.type, format);
        if (serializeInfo.hasMultiCollection || serializeInfo.hasCsvCollection) {
            description += `${description ? "\n" : ""}This parameter could be formatted as ${serializeInfo.collectionInfo.join(", ")} collection string, we provide ${serializeInfo.descriptions.join(", ")} from serializeHelper.ts to help${serializeInfo.hasMultiCollection
                ? ", you will probably need to set skipUrlEncoding as true when sending the request"
                : ""}.`;
        }
        if (format === "tsv") {
            description += `${description ? "\n" : ""}This parameter could be formatted as tsv collection string.`;
        }
    }
    (_c = getImportedModelName(schema, schemaContext)) === null || _c === void 0 ? void 0 : _c.forEach(importedModels.add, importedModels);
    const serializationType = getParameterSerializationInfo(dpgContext, parameter, schema, options.operationGroupName, options.operationName);
    return {
        type: paramType,
        name,
        param: {
            name,
            type: serializationType.typeName,
            typeName: serializationType.typeName,
            required: !parameter.param.optional,
            description,
            wrapperType: serializationType.wrapperType
        }
    };
}
function getParameterName(name) {
    if (name && name.toLowerCase() === "content-type") {
        return "contentType";
    }
    return `"${name}"`;
}
function transformCookieParameters(dpgContext, parameters) {
    // TODO: support cookie parameters, https://github.com/Azure/autorest.typescript/issues/2898
    parameters.parameters
        .filter((p) => p.type === "cookie")
        .forEach((p) => {
        reportDiagnostic(dpgContext.program, {
            code: "parameter-type-not-supported",
            format: {
                paramName: p.name,
                paramType: p.type
            },
            target: NoTarget
        });
    });
}
function transformQueryParameters(dpgContext, parameters, options) {
    const queryParameters = parameters.parameters.filter((p) => {
        var _a;
        return p.type === "query" &&
            !(isApiVersion(dpgContext, p.param) &&
                ((_a = options.apiVersionInfo) === null || _a === void 0 ? void 0 : _a.definedPosition) === "query");
    });
    if (!queryParameters.length) {
        return [];
    }
    return queryParameters.map((qp) => getParameterMetadata(dpgContext, "query", qp, options));
}
/**
 * Only support to take the global path parameter as path parameter
 * @returns
 */
function transformPathParameters(dpgContext, parameters, options) {
    // build wrapper path parameters
    const pathParameters = parameters.parameters.filter((p) => p.type === "path");
    if (!pathParameters.length) {
        return [];
    }
    // only need to build path parameters for wrapper type
    const params = pathParameters
        .map((qp) => getParameterMetadata(dpgContext, "path", qp, options))
        .filter((p) => p.param.wrapperType);
    return params;
}
export function transformHeaderParameters(dpgContext, parameters, options) {
    const headerParameters = parameters.parameters.filter((p) => p.type === "header");
    if (!headerParameters.length) {
        return [];
    }
    return headerParameters.map((qp) => getParameterMetadata(dpgContext, "header", qp, options));
}
function transformBodyParameters(dpgContext, parameters, headers, importedModels, inputBodyType) {
    var _a;
    const bodyType = parameters.body && inputBodyType ? inputBodyType : (_a = parameters.body) === null || _a === void 0 ? void 0 : _a.type;
    if (!bodyType || isVoidType(bodyType)) {
        return;
    }
    return transformRequestBody(dpgContext, bodyType, parameters, importedModels, headers);
}
function transformRequestBody(dpgContext, bodyType, parameters, importedModels, headers) {
    var _a, _b;
    const contentTypes = extractMediaTypes((_b = (_a = parameters.body) === null || _a === void 0 ? void 0 : _a.contentTypes) !== null && _b !== void 0 ? _b : []);
    const schema = getSchemaForType(dpgContext, bodyType, {
        mediaTypes: contentTypes,
        isRequestBody: true,
        usage: [SchemaContext.Input, SchemaContext.Exception]
    });
    const descriptions = getBodyDescriptions(dpgContext, schema, parameters);
    const type = getRequestBodyType(schema, importedModels, headers);
    return {
        isPartialBody: false,
        body: [
            {
                properties: schema.properties,
                typeName: schema.name,
                name: "body",
                type,
                required: isBodyRequired(parameters),
                description: descriptions.join("\n\n"),
                isMultipartBody: hasMediaType(KnownMediaType.MultipartFormData, contentTypes) &&
                    contentTypes.length === 1,
                oriSchema: schema
            }
        ]
    };
}
function getRequestBodyType(bodySchema, importedModels, headers) {
    var _a;
    const schemaUsage = [SchemaContext.Input, SchemaContext.Exception];
    const importedNames = (_a = getImportedModelName(bodySchema, schemaUsage)) !== null && _a !== void 0 ? _a : [];
    importedNames.forEach(importedModels.add, importedModels);
    let typeName = getTypeName(bodySchema, schemaUsage);
    const contentTypes = headers === null || headers === void 0 ? void 0 : headers.filter((h) => h.name === "contentType").map((h) => h.param.type);
    const hasMergeAndPatchType = isMediaTypeJsonMergePatch(contentTypes !== null && contentTypes !== void 0 ? contentTypes : []);
    if (hasMergeAndPatchType &&
        Boolean(bodySchema.name) &&
        bodySchema.properties) {
        typeName = `${typeName}ResourceMergeAndPatch`;
    }
    return typeName;
}
function getBodyDescriptions(dpgContext, bodySchema, parameters) {
    var _a;
    const description = ((_a = parameters.body) === null || _a === void 0 ? void 0 : _a.property)
        ? getFormattedPropertyDoc(dpgContext.program, parameters.body.property, bodySchema)
        : "";
    return description ? [description] : [];
}
