// @azure-tools/rlc-common/dist-esm/metadata/buildWarpConfig.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { isAzureMonorepoPackage } from "../helpers/packageUtil.js";
export const WarpConfigTemplate = `# warp.config.yml — build configuration

exports:
{{exports}}

targets:
  - name: browser
    tsconfig: "../../../tsconfig.src.browser.json"
    polyfillSuffix: "-browser"

  - name: react-native
    tsconfig: "../../../tsconfig.src.react-native.json"
    polyfillSuffix: "-react-native"

  - name: esm
    condition: import
    tsconfig: "../../../tsconfig.src.esm.json"

  - name: commonjs
    condition: require
    tsconfig: "../../../tsconfig.src.cjs.json"
`;
/**
 * Builds a warp.config.yml file for Azure SDK monorepo packages.
 * Only generated when azureSdkForJs is true.
 */
export function buildWarpConfig(model, { exports } = {}) {
    if (!isAzureMonorepoPackage(model)) {
        return;
    }
    if (model.options?.moduleKind !== "esm") {
        return;
    }
    const allExports = {
        "./package.json": "./package.json",
        ".": "./src/index.ts",
        ...exports
    };
    const exportsContent = Object.entries(allExports)
        .map(([key, value]) => `  ${JSON.stringify(key)}: ${JSON.stringify(value)}`)
        .join("\n");
    const content = WarpConfigTemplate.replace("{{exports}}", exportsContent);
    return {
        path: "warp.config.yml",
        content
    };
}
